{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6868296f-0d69-4e6c-96b3-5be538ed259c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dce47ec4-8d0c-439f-8fcd-c963232600d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 308211.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AB is discontinuous at line 309029.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 309052.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 309063.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AF is discontinuous at line 309068.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AG is discontinuous at line 309076.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AH is discontinuous at line 309079.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 309080.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AO is discontinuous at line 309083.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AP is discontinuous at line 309084.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AQ is discontinuous at line 309087.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AR is discontinuous at line 309090.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AU is discontinuous at line 309092.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AV is discontinuous at line 309096.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AW is discontinuous at line 309100.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AX is discontinuous at line 309104.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 309105.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AZ is discontinuous at line 309107.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A0 is discontinuous at line 309108.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A1 is discontinuous at line 309113.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A2 is discontinuous at line 309115.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A4 is discontinuous at line 309116.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A5 is discontinuous at line 309118.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A6 is discontinuous at line 309122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A7 is discontinuous at line 309125.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A8 is discontinuous at line 309131.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A9 is discontinuous at line 309133.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BA is discontinuous at line 309135.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 309348.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BD is discontinuous at line 309349.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 309358.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BF is discontinuous at line 309359.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BK is discontinuous at line 309360.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 309361.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BM is discontinuous at line 309363.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BN is discontinuous at line 309364.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BT is discontinuous at line 309367.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BV is discontinuous at line 309368.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BW is discontinuous at line 309369.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BX is discontinuous at line 309371.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BY is discontinuous at line 309386.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BZ is discontinuous at line 309388.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 309417.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CB is discontinuous at line 310081.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 310094.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 310098.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CF is discontinuous at line 310104.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CG is discontinuous at line 310110.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 310111.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CO is discontinuous at line 310112.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CP is discontinuous at line 310114.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CQ is discontinuous at line 310115.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CR is discontinuous at line 310119.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CU is discontinuous at line 310121.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CV is discontinuous at line 310122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 310124.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C0 is discontinuous at line 310126.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C3 is discontinuous at line 310128.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C4 is discontinuous at line 310129.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C5 is discontinuous at line 310130.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C6 is discontinuous at line 310132.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C7 is discontinuous at line 310133.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C8 is discontinuous at line 310134.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C9 is discontinuous at line 310135.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DA is discontinuous at line 310136.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DD is discontinuous at line 310304.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 310313.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DF is discontinuous at line 310315.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DJ is discontinuous at line 310316.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DK is discontinuous at line 310317.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DN is discontinuous at line 310319.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DT is discontinuous at line 310320.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DW is discontinuous at line 310321.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DX is discontinuous at line 310322.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DZ is discontinuous at line 310323.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 310352.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AB is discontinuous at line 311760.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311796.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 311811.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AF is discontinuous at line 311830.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AG is discontinuous at line 311837.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AH is discontinuous at line 311840.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311841.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AO is discontinuous at line 311843.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AP is discontinuous at line 311844.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AQ is discontinuous at line 311859.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AR is discontinuous at line 311863.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AS is discontinuous at line 311865.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AT is discontinuous at line 311866.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AU is discontinuous at line 311868.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AV is discontinuous at line 311873.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AW is discontinuous at line 311875.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AX is discontinuous at line 311877.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AZ is discontinuous at line 311880.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A0 is discontinuous at line 311881.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A1 is discontinuous at line 311887.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A3 is discontinuous at line 311888.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A5 is discontinuous at line 311889.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A6 is discontinuous at line 311892.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A7 is discontinuous at line 311894.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A8 is discontinuous at line 311898.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BA is discontinuous at line 311908.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BD is discontinuous at line 312120.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 312122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 312124.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BM is discontinuous at line 312125.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BV is discontinuous at line 312126.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BW is discontinuous at line 312128.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BX is discontinuous at line 312131.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BY is discontinuous at line 312139.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BZ is discontinuous at line 312140.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 312142.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CB is discontinuous at line 313127.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 313136.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 313150.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CF is discontinuous at line 313163.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 313170.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CP is discontinuous at line 313172.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CQ is discontinuous at line 313182.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CR is discontinuous at line 313183.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CT is discontinuous at line 313184.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CU is discontinuous at line 313187.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CV is discontinuous at line 313189.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 313190.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C0 is discontinuous at line 313191.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C1 is discontinuous at line 313197.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C3 is discontinuous at line 313199.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C6 is discontinuous at line 313201.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C7 is discontinuous at line 313202.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C8 is discontinuous at line 313203.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DA is discontinuous at line 313206.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 313361.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DJ is discontinuous at line 313365.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DK is discontinuous at line 313366.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DL is discontinuous at line 313368.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DW is discontinuous at line 313369.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DX is discontinuous at line 313371.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4WPO\", \"4wpo.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "49178a24-f554-4695-b9f1-53c8910d0ef3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entity ID: 1, Chains: AA,CA, Molecule: 23S Ribosomal RNA\n",
      "Entity ID: 2, Chains: AB,CB, Molecule: 5S Ribosomal RNA\n",
      "Entity ID: 3, Chains: AC,CC, Molecule: 50S ribosomal protein L1\n",
      "Entity ID: 4, Chains: AD,CD, Molecule: 50S ribosomal protein L2\n",
      "Entity ID: 5, Chains: AE,CE, Molecule: 50S ribosomal protein L3\n",
      "Entity ID: 6, Chains: AF,CF, Molecule: 50S ribosomal protein L4\n",
      "Entity ID: 7, Chains: AG,CG, Molecule: 50S ribosomal protein L5\n",
      "Entity ID: 8, Chains: AH,CH, Molecule: 50S ribosomal protein L6\n",
      "Entity ID: 9, Chains: AK,CK, Molecule: 50S ribosomal protein L10\n",
      "Entity ID: 10, Chains: AL,CL, Molecule: 50S ribosomal protein L11\n",
      "Entity ID: 11, Chains: AN,CN, Molecule: 50S ribosomal protein L13\n",
      "Entity ID: 12, Chains: AO,CO, Molecule: 50S ribosomal protein L14\n",
      "Entity ID: 13, Chains: AP,CP, Molecule: 50S ribosomal protein L15\n",
      "Entity ID: 14, Chains: AQ,CQ, Molecule: 50S ribosomal protein L16\n",
      "Entity ID: 15, Chains: AR,CR, Molecule: 50S ribosomal protein L17\n",
      "Entity ID: 16, Chains: AS,CS, Molecule: 50S ribosomal protein L18\n",
      "Entity ID: 17, Chains: AT,CT, Molecule: 50S ribosomal protein L19\n",
      "Entity ID: 18, Chains: AU,CU, Molecule: 50S ribosomal protein L20\n",
      "Entity ID: 19, Chains: AV,CV, Molecule: 50S ribosomal protein L21\n",
      "Entity ID: 20, Chains: AW,CW, Molecule: 50S ribosomal protein L22\n",
      "Entity ID: 21, Chains: AX,CX, Molecule: 50S ribosomal protein L23\n",
      "Entity ID: 22, Chains: AY,CY, Molecule: 50S ribosomal protein L24\n",
      "Entity ID: 23, Chains: AZ,CZ, Molecule: 50S ribosomal protein L25\n",
      "Entity ID: 24, Chains: A0,C0, Molecule: 50S ribosomal protein L27\n",
      "Entity ID: 25, Chains: A1,C1, Molecule: 50S ribosomal protein L28\n",
      "Entity ID: 26, Chains: A2,C2, Molecule: 50S ribosomal protein L29\n",
      "Entity ID: 27, Chains: A3,C3, Molecule: 50S ribosomal protein L30\n",
      "Entity ID: 28, Chains: A4,C4, Molecule: 50S ribosomal protein L31\n",
      "Entity ID: 29, Chains: A5,C5, Molecule: 50S ribosomal protein L32\n",
      "Entity ID: 30, Chains: A6,C6, Molecule: 50S ribosomal protein L33\n",
      "Entity ID: 31, Chains: A7,C7, Molecule: 50S ribosomal protein L34\n",
      "Entity ID: 32, Chains: A8,C8, Molecule: 50S ribosomal protein L35\n",
      "Entity ID: 33, Chains: A9,C9, Molecule: 50S ribosomal protein L36\n",
      "Entity ID: 34, Chains: BA,DA, Molecule: 16S Ribosomal RNA\n",
      "Entity ID: 35, Chains: BB,DB, Molecule: 30S ribosomal protein S2\n",
      "Entity ID: 36, Chains: BC,DC, Molecule: 30S ribosomal protein S3\n",
      "Entity ID: 37, Chains: BD,DD, Molecule: 30S ribosomal protein S4\n",
      "Entity ID: 38, Chains: BE,DE, Molecule: 30S ribosomal protein S5\n",
      "Entity ID: 39, Chains: BF,DF, Molecule: 30S ribosomal protein S6\n",
      "Entity ID: 40, Chains: BG,DG, Molecule: 30S ribosomal protein S7\n",
      "Entity ID: 41, Chains: BH,DH, Molecule: 30S ribosomal protein S8\n",
      "Entity ID: 42, Chains: BI,DI, Molecule: 30S ribosomal protein S9\n",
      "Entity ID: 43, Chains: BJ,DJ, Molecule: 30S ribosomal protein S10\n",
      "Entity ID: 44, Chains: BK,DK, Molecule: 30S ribosomal protein S11\n",
      "Entity ID: 45, Chains: BL,DL, Molecule: 30S ribosomal protein S12\n",
      "Entity ID: 46, Chains: BM,DM, Molecule: 30S ribosomal protein S13\n",
      "Entity ID: 47, Chains: BN,DN, Molecule: 30S ribosomal protein S14 type Z\n",
      "Entity ID: 48, Chains: BO,DO, Molecule: 30S ribosomal protein S15\n",
      "Entity ID: 49, Chains: BP,DP, Molecule: 30S ribosomal protein S16\n",
      "Entity ID: 50, Chains: BQ,DQ, Molecule: 30S ribosomal protein S17\n",
      "Entity ID: 51, Chains: BR,DR, Molecule: 30S ribosomal protein S18\n",
      "Entity ID: 52, Chains: BS,DS, Molecule: 30S ribosomal protein S19\n",
      "Entity ID: 53, Chains: BT,DT, Molecule: 30S ribosomal protein S20\n",
      "Entity ID: 54, Chains: BU,DU, Molecule: 30S ribosomal protein Thx\n",
      "Entity ID: 55, Chains: BV,DV, Molecule: mRNA\n",
      "Entity ID: 56, Chains: BW,DW, Molecule: A-site tRNA\n",
      "Entity ID: 57, Chains: BX,DX, Molecule: P-site tRNA\n",
      "Entity ID: 58, Chains: BY,DY, Molecule: E-site tRNA\n",
      "Entity ID: 59, Chains: BZ,DZ, Molecule: 50S ribosomal protein L9,Elongation factor G\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import PDBList\n",
    "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
    "\n",
    "mmcif_dict = MMCIF2Dict(\"4wpo.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a296d2c7-7918-40f4-bf0a-47b8c605c85b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 308211.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AB is discontinuous at line 309029.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 309052.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 309063.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AF is discontinuous at line 309068.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AG is discontinuous at line 309076.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AH is discontinuous at line 309079.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 309080.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AO is discontinuous at line 309083.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AP is discontinuous at line 309084.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AQ is discontinuous at line 309087.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AR is discontinuous at line 309090.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AU is discontinuous at line 309092.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AV is discontinuous at line 309096.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AW is discontinuous at line 309100.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AX is discontinuous at line 309104.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 309105.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AZ is discontinuous at line 309107.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A0 is discontinuous at line 309108.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A1 is discontinuous at line 309113.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A2 is discontinuous at line 309115.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A4 is discontinuous at line 309116.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A5 is discontinuous at line 309118.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A6 is discontinuous at line 309122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A7 is discontinuous at line 309125.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A8 is discontinuous at line 309131.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A9 is discontinuous at line 309133.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BA is discontinuous at line 309135.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 309348.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BD is discontinuous at line 309349.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 309358.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BF is discontinuous at line 309359.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BK is discontinuous at line 309360.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 309361.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BM is discontinuous at line 309363.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BN is discontinuous at line 309364.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BT is discontinuous at line 309367.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BV is discontinuous at line 309368.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BW is discontinuous at line 309369.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BX is discontinuous at line 309371.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BY is discontinuous at line 309386.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BZ is discontinuous at line 309388.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 309417.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CB is discontinuous at line 310081.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 310094.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 310098.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CF is discontinuous at line 310104.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CG is discontinuous at line 310110.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 310111.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CO is discontinuous at line 310112.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CP is discontinuous at line 310114.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CQ is discontinuous at line 310115.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CR is discontinuous at line 310119.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CU is discontinuous at line 310121.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CV is discontinuous at line 310122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 310124.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C0 is discontinuous at line 310126.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C3 is discontinuous at line 310128.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C4 is discontinuous at line 310129.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C5 is discontinuous at line 310130.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C6 is discontinuous at line 310132.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C7 is discontinuous at line 310133.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C8 is discontinuous at line 310134.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C9 is discontinuous at line 310135.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DA is discontinuous at line 310136.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DD is discontinuous at line 310304.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 310313.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DF is discontinuous at line 310315.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DJ is discontinuous at line 310316.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DK is discontinuous at line 310317.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DN is discontinuous at line 310319.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DT is discontinuous at line 310320.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DW is discontinuous at line 310321.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DX is discontinuous at line 310322.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DZ is discontinuous at line 310323.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 310352.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AB is discontinuous at line 311760.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311796.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 311811.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AF is discontinuous at line 311830.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AG is discontinuous at line 311837.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AH is discontinuous at line 311840.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311841.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AO is discontinuous at line 311843.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AP is discontinuous at line 311844.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AQ is discontinuous at line 311859.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AR is discontinuous at line 311863.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AS is discontinuous at line 311865.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AT is discontinuous at line 311866.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AU is discontinuous at line 311868.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AV is discontinuous at line 311873.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AW is discontinuous at line 311875.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AX is discontinuous at line 311877.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AZ is discontinuous at line 311880.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A0 is discontinuous at line 311881.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A1 is discontinuous at line 311887.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A3 is discontinuous at line 311888.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A5 is discontinuous at line 311889.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A6 is discontinuous at line 311892.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A7 is discontinuous at line 311894.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A8 is discontinuous at line 311898.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BA is discontinuous at line 311908.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BD is discontinuous at line 312120.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 312122.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 312124.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BM is discontinuous at line 312125.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BV is discontinuous at line 312126.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BW is discontinuous at line 312128.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BX is discontinuous at line 312131.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BY is discontinuous at line 312139.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain BZ is discontinuous at line 312140.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 312142.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CB is discontinuous at line 313127.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 313136.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 313150.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CF is discontinuous at line 313163.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 313170.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CP is discontinuous at line 313172.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CQ is discontinuous at line 313182.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CR is discontinuous at line 313183.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CT is discontinuous at line 313184.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CU is discontinuous at line 313187.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CV is discontinuous at line 313189.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 313190.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C0 is discontinuous at line 313191.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C1 is discontinuous at line 313197.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C3 is discontinuous at line 313199.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C6 is discontinuous at line 313201.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C7 is discontinuous at line 313202.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C8 is discontinuous at line 313203.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DA is discontinuous at line 313206.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 313361.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DJ is discontinuous at line 313365.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DK is discontinuous at line 313366.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DL is discontinuous at line 313368.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DW is discontinuous at line 313369.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain DX is discontinuous at line 313371.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO, Select\n",
    "\n",
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4WPO\", \"4wpo.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G chains\n",
    "class EF_G_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G chains (GB and ND)\n",
    "io.save(\"4WPO_EF-G.cif\", EF_G_ChainSelect([\"BZ\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "28316515-3b6a-424c-8ea4-0e17bd639d1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "23e3b418a77d406a9ad6d9a2031958b5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "faa59867362540d2b53770f5279be345",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0933f22e8424deea419d1d20d8ac1f0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntSlider(value=50, description='My Slider')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "widgets.IntSlider()\n",
    "view = nv.show_file(\"4WPO_EF-G.cif\")\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')  \n",
    "view.add_representation('stick', selection='540 and PRO', color='black')\n",
    "display(view)\n",
    "display(slider)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f014d79c-34fc-44cc-8889-4cc386a40dfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Chain_BZ\n",
      "MKVILLEPLENLGDVGQVVDVKPGYARNYLLPRGLAVLATESNLKALEARIRAQAKRLAERKAEAERLKEILENDLKRLRNIGIAAHIDAGKTTTTERILYYTGRAAVTTCFWKDHRINIIDTPGHVDFTIEVERSMRVLDGAIVVFDSSQGVEPQSETVWRQAEKYKVPRIAFANKMDKTGADLWLVIRTMQERLGARPVVMQLPIGREDTFSGIIDVLRMKAYTYGNDLGTDIREIPIPEEYLDQAREYHEKLVEVAADFDENIMLKYLEGEEPTEEELVAAIRKGTIDLKITPVFLGSALKNKGVQLLLDAVVDYLPSPLDIPPIKGTTPEGEVVEIHPDPNGPLAALAFKIMADPYVGRLTFIRVYSGTLTSGSYVYNTTKGRKERVARLLRMHANHREEVEELKAGDLGAVVGLKETITGDTLVGEDAPRVILESIEVPEPVIDVAIEPKTKADQEKLSQALARLAEEDPTFRVSTHPETGQTIISGMGELHLEIIVDRLKREFKVDANVGKPQVAYRETITKPVDVEGKFIRQTGGRGQYGHVKIKVEPLPRGSGFEFVNAIVGGVIPKEYIPAVQKGIEEAMQSGPLIGFPVVDIKVTLYDGSYHEVDSSEMAFKIAGSMAIKEAVQKGDPVILEPIMRVEVTTPEEYMGDVIGDLNARRGQILGMEPRGNAQVIRAFVPLAEMFGYATDLRSKTQGRGSFVMFFDHYQEVPKQVQEKLIK\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, PPBuilder\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Parse the structure\n",
    "parser = MMCIFParser(QUIET=True)\n",
    "structure = parser.get_structure(\"4WPO\", \"4WPO_EF-G.cif\")\n",
    "\n",
    "ppb = PPBuilder()\n",
    "records = []\n",
    "\n",
    "# Extract sequences for each chain\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        seq = \"\"\n",
    "        for pp in ppb.build_peptides(chain):\n",
    "            seq += str(pp.get_sequence())\n",
    "\n",
    "        if seq:\n",
    "            record = SeqRecord(Seq(seq), id=f\"Chain_{chain.id}\", description=\"\")\n",
    "            records.append(record)\n",
    "\n",
    "# Print the sequences in FASTA format\n",
    "for record in records:\n",
    "    print(f\">{record.id}\\n{record.seq}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e556671-2d7f-4a00-b660-20ca4f298452",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311412.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311413.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 311414.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 311480.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain B9 is discontinuous at line 311481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 311482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 311483.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 311484.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 311550.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D9 is discontinuous at line 311551.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4499fba2-6459-4592-8730-71b672514ca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entity ID: 1, Chains: AA,CA, Molecule: 16S ribosomal RNA\n",
      "Entity ID: 2, Chains: AB,CB, Molecule: 30S RIBOSOMAL PROTEIN S2\n",
      "Entity ID: 3, Chains: AC,CC, Molecule: 30S RIBOSOMAL PROTEIN S3\n",
      "Entity ID: 4, Chains: AD,CD, Molecule: 30S RIBOSOMAL PROTEIN S4\n",
      "Entity ID: 5, Chains: AE,CE, Molecule: 30S RIBOSOMAL PROTEIN S5\n",
      "Entity ID: 6, Chains: AF,CF, Molecule: 30S RIBOSOMAL PROTEIN S6\n",
      "Entity ID: 7, Chains: AG,CG, Molecule: 30S RIBOSOMAL PROTEIN S7\n",
      "Entity ID: 8, Chains: AH,CH, Molecule: 30S RIBOSOMAL PROTEIN S8\n",
      "Entity ID: 9, Chains: AI,CI, Molecule: 30S RIBOSOMAL PROTEIN S9\n",
      "Entity ID: 10, Chains: AJ,CJ, Molecule: 30S RIBOSOMAL PROTEIN S10\n",
      "Entity ID: 11, Chains: AK,CK, Molecule: 30S RIBOSOMAL PROTEIN S11\n",
      "Entity ID: 12, Chains: AL,CL, Molecule: 30S RIBOSOMAL PROTEIN S12\n",
      "Entity ID: 13, Chains: AM,CM, Molecule: 30S RIBOSOMAL PROTEIN S13\n",
      "Entity ID: 14, Chains: AN,CN, Molecule: 30S RIBOSOMAL PROTEIN S14 TYPE Z\n",
      "Entity ID: 15, Chains: AO,CO, Molecule: 30S RIBOSOMAL PROTEIN S15\n",
      "Entity ID: 16, Chains: AP,CP, Molecule: 30S RIBOSOMAL PROTEIN S16\n",
      "Entity ID: 17, Chains: AQ,CQ, Molecule: 30S RIBOSOMAL PROTEIN S17\n",
      "Entity ID: 18, Chains: AR,CR, Molecule: 30S RIBOSOMAL PROTEIN S18\n",
      "Entity ID: 19, Chains: AS,CS, Molecule: 30S RIBOSOMAL PROTEIN S19\n",
      "Entity ID: 20, Chains: AT,CT, Molecule: 30S RIBOSOMAL PROTEIN S20\n",
      "Entity ID: 21, Chains: AU,CU, Molecule: 30S RIBOSOMAL PROTEIN THX\n",
      "Entity ID: 22, Chains: AV,AW,CV,CW, Molecule: E-SITE TRNA FMET OR P-SITE TRNA FMET (UNMODIFIED BASES EXCEPT FOR THYMINE 54)\n",
      "Entity ID: 23, Chains: AX,CX, Molecule: MRNA\n",
      "Entity ID: 24, Chains: AY,CY, Molecule: ELONGATION FACTOR G\n",
      "Entity ID: 25, Chains: B0,D0, Molecule: 50S RIBOSOMAL PROTEIN L27\n",
      "Entity ID: 26, Chains: B1,D1, Molecule: 50S RIBOSOMAL PROTEIN L28\n",
      "Entity ID: 27, Chains: B2,D2, Molecule: 50S RIBOSOMAL PROTEIN L29\n",
      "Entity ID: 28, Chains: B3,D3, Molecule: 50S RIBOSOMAL PROTEIN L30\n",
      "Entity ID: 29, Chains: B4,D4, Molecule: 50S RIBOSOMAL PROTEIN L31\n",
      "Entity ID: 30, Chains: B5,D5, Molecule: 50S RIBOSOMAL PROTEIN L32\n",
      "Entity ID: 31, Chains: B6,D6, Molecule: 50S RIBOSOMAL PROTEIN L33\n",
      "Entity ID: 32, Chains: B7,D7, Molecule: 50S RIBOSOMAL PROTEIN L34\n",
      "Entity ID: 33, Chains: B8,D8, Molecule: 50S RIBOSOMAL PROTEIN L35\n",
      "Entity ID: 34, Chains: B9,D9, Molecule: 50S RIBOSOMAL PROTEIN L36\n",
      "Entity ID: 35, Chains: BA,DA, Molecule: 23S RIBOSOMAL RNA\n",
      "Entity ID: 36, Chains: BB,DB, Molecule: 5S RIBOSOMAL RNA\n",
      "Entity ID: 37, Chains: BC,DC, Molecule: 50S RIBOSOMAL PROTEIN L1\n",
      "Entity ID: 38, Chains: BD,DD, Molecule: 50S RIBOSOMAL PROTEIN L2\n",
      "Entity ID: 39, Chains: BE,DE, Molecule: 50S RIBOSOMAL PROTEIN L3\n",
      "Entity ID: 40, Chains: BF,DF, Molecule: 50S RIBOSOMAL PROTEIN L4\n",
      "Entity ID: 41, Chains: BG,DG, Molecule: 50S RIBOSOMAL PROTEIN L5\n",
      "Entity ID: 42, Chains: BH,DH, Molecule: 50S RIBOSOMAL PROTEIN L6\n",
      "Entity ID: 43, Chains: BJ,DJ, Molecule: 50S RIBOSOMAL PROTEIN L10\n",
      "Entity ID: 44, Chains: BK,DK, Molecule: 50S RIBOSOMAL PROTEIN L11\n",
      "Entity ID: 45, Chains: BL,BM,Bl,Bm,DL,DM,Dl,Dm, Molecule: 50S RIBOSOMAL PROTEIN L12\n",
      "Entity ID: 46, Chains: BN,DN, Molecule: 50S RIBOSOMAL PROTEIN L13\n",
      "Entity ID: 47, Chains: BO,DO, Molecule: 50S RIBOSOMAL PROTEIN L14\n",
      "Entity ID: 48, Chains: BP,DP, Molecule: 50S RIBOSOMAL PROTEIN L15\n",
      "Entity ID: 49, Chains: BQ,DQ, Molecule: 50S RIBOSOMAL PROTEIN L16\n",
      "Entity ID: 50, Chains: BR,DR, Molecule: 50S RIBOSOMAL PROTEIN L17\n",
      "Entity ID: 51, Chains: BS,DS, Molecule: 50S RIBOSOMAL PROTEIN L18\n",
      "Entity ID: 52, Chains: BT,DT, Molecule: 50S RIBOSOMAL PROTEIN L19\n",
      "Entity ID: 53, Chains: BU,DU, Molecule: 50S RIBOSOMAL PROTEIN L20\n",
      "Entity ID: 54, Chains: BV,DV, Molecule: 50S RIBOSOMAL PROTEIN L21\n",
      "Entity ID: 55, Chains: BW,DW, Molecule: 50S RIBOSOMAL PROTEIN L22\n",
      "Entity ID: 56, Chains: BX,DX, Molecule: 50S RIBOSOMAL PROTEIN L23\n",
      "Entity ID: 57, Chains: BY,DY, Molecule: 50S RIBOSOMAL PROTEIN L24\n",
      "Entity ID: 58, Chains: BZ,DZ, Molecule: 50S RIBOSOMAL PROTEIN L25\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
    "\n",
    "mmcif_dict = MMCIF2Dict(\"4v5f.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f70cdaac-7e08-45a0-8552-1a370f519247",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311412.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311413.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 311414.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 311480.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain B9 is discontinuous at line 311481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 311482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 311483.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 311484.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 311550.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D9 is discontinuous at line 311551.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO, Select\n",
    "\n",
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G chains\n",
    "class EF_G_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G chains \n",
    "io.save(\"4v5f_EF-G.cif\", EF_G_ChainSelect([\"CY\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "346ca0e9-3d06-4408-9956-f7184f180a88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chain ID: CY\n",
      "Residue 4: LYS\n",
      "Residue 5: VAL\n",
      "Residue 6: GLU\n",
      "Residue 7: TYR\n",
      "Residue 8: ASP\n",
      "Residue 9: LEU\n",
      "Residue 10: LYS\n",
      "Residue 11: ARG\n",
      "Residue 12: LEU\n",
      "Residue 13: ARG\n",
      "Residue 14: ASN\n",
      "Residue 15: ILE\n",
      "Residue 16: GLY\n",
      "Residue 17: ILE\n",
      "Residue 18: ALA\n",
      "Residue 19: ALA\n",
      "Residue 20: HIS\n",
      "Residue 21: ILE\n",
      "Residue 22: ASP\n",
      "Residue 23: ALA\n",
      "Residue 24: GLY\n",
      "Residue 25: LYS\n",
      "Residue 26: THR\n",
      "Residue 27: THR\n",
      "Residue 28: THR\n",
      "Residue 29: THR\n",
      "Residue 30: GLU\n",
      "Residue 31: ARG\n",
      "Residue 32: ILE\n",
      "Residue 33: LEU\n",
      "Residue 34: TYR\n",
      "Residue 35: TYR\n",
      "Residue 36: THR\n",
      "Residue 37: GLY\n",
      "Residue 38: ARG\n",
      "Residue 39: ILE\n",
      "Residue 40: HIS\n",
      "Residue 41: LYS\n",
      "Residue 42: ILE\n",
      "Residue 43: GLY\n",
      "Residue 64: THR\n",
      "Residue 65: ILE\n",
      "Residue 66: THR\n",
      "Residue 67: ALA\n",
      "Residue 68: ALA\n",
      "Residue 69: VAL\n",
      "Residue 70: THR\n",
      "Residue 71: THR\n",
      "Residue 72: CYS\n",
      "Residue 73: PHE\n",
      "Residue 74: TRP\n",
      "Residue 75: LYS\n",
      "Residue 76: ASP\n",
      "Residue 77: HIS\n",
      "Residue 78: ARG\n",
      "Residue 79: ILE\n",
      "Residue 80: ASN\n",
      "Residue 81: ILE\n",
      "Residue 82: ILE\n",
      "Residue 83: ASP\n",
      "Residue 84: THR\n",
      "Residue 85: PRO\n",
      "Residue 86: GLY\n",
      "Residue 87: HIS\n",
      "Residue 88: VAL\n",
      "Residue 89: ASP\n",
      "Residue 90: PHE\n",
      "Residue 91: THR\n",
      "Residue 92: ILE\n",
      "Residue 93: GLU\n",
      "Residue 94: VAL\n",
      "Residue 95: GLU\n",
      "Residue 96: ARG\n",
      "Residue 97: SER\n",
      "Residue 98: MET\n",
      "Residue 99: ARG\n",
      "Residue 100: VAL\n",
      "Residue 101: LEU\n",
      "Residue 102: ASP\n",
      "Residue 103: GLY\n",
      "Residue 104: ALA\n",
      "Residue 105: ILE\n",
      "Residue 106: VAL\n",
      "Residue 107: VAL\n",
      "Residue 108: PHE\n",
      "Residue 109: ASP\n",
      "Residue 110: SER\n",
      "Residue 111: SER\n",
      "Residue 112: GLN\n",
      "Residue 113: GLY\n",
      "Residue 114: VAL\n",
      "Residue 115: GLU\n",
      "Residue 116: PRO\n",
      "Residue 117: GLN\n",
      "Residue 118: SER\n",
      "Residue 119: GLU\n",
      "Residue 120: THR\n",
      "Residue 121: VAL\n",
      "Residue 122: TRP\n",
      "Residue 123: ARG\n",
      "Residue 124: GLN\n",
      "Residue 125: ALA\n",
      "Residue 126: GLU\n",
      "Residue 127: LYS\n",
      "Residue 128: TYR\n",
      "Residue 129: LYS\n",
      "Residue 130: VAL\n",
      "Residue 131: PRO\n",
      "Residue 132: ARG\n",
      "Residue 133: ILE\n",
      "Residue 134: ALA\n",
      "Residue 135: PHE\n",
      "Residue 136: ALA\n",
      "Residue 137: ASN\n",
      "Residue 138: LYS\n",
      "Residue 139: MET\n",
      "Residue 140: ASP\n",
      "Residue 141: LYS\n",
      "Residue 142: THR\n",
      "Residue 143: GLY\n",
      "Residue 144: ALA\n",
      "Residue 145: ASP\n",
      "Residue 146: LEU\n",
      "Residue 147: TRP\n",
      "Residue 148: LEU\n",
      "Residue 149: VAL\n",
      "Residue 150: ILE\n",
      "Residue 151: ARG\n",
      "Residue 152: THR\n",
      "Residue 153: MET\n",
      "Residue 154: GLN\n",
      "Residue 155: GLU\n",
      "Residue 156: ARG\n",
      "Residue 157: LEU\n",
      "Residue 158: GLY\n",
      "Residue 159: ALA\n",
      "Residue 160: ARG\n",
      "Residue 161: PRO\n",
      "Residue 162: VAL\n",
      "Residue 163: VAL\n",
      "Residue 164: MET\n",
      "Residue 165: GLN\n",
      "Residue 166: LEU\n",
      "Residue 167: PRO\n",
      "Residue 168: ILE\n",
      "Residue 169: GLY\n",
      "Residue 170: ARG\n",
      "Residue 171: GLU\n",
      "Residue 172: ASP\n",
      "Residue 173: THR\n",
      "Residue 174: PHE\n",
      "Residue 175: SER\n",
      "Residue 176: GLY\n",
      "Residue 177: ILE\n",
      "Residue 178: ILE\n",
      "Residue 179: ASP\n",
      "Residue 180: VAL\n",
      "Residue 181: LEU\n",
      "Residue 182: ARG\n",
      "Residue 183: MET\n",
      "Residue 184: LYS\n",
      "Residue 185: ALA\n",
      "Residue 186: TYR\n",
      "Residue 187: THR\n",
      "Residue 188: TYR\n",
      "Residue 189: GLY\n",
      "Residue 190: ASN\n",
      "Residue 191: ASP\n",
      "Residue 192: LEU\n",
      "Residue 193: GLY\n",
      "Residue 194: THR\n",
      "Residue 195: ASP\n",
      "Residue 196: ILE\n",
      "Residue 197: ARG\n",
      "Residue 198: GLU\n",
      "Residue 199: ILE\n",
      "Residue 200: PRO\n",
      "Residue 201: ILE\n",
      "Residue 202: PRO\n",
      "Residue 203: GLU\n",
      "Residue 204: GLU\n",
      "Residue 205: TYR\n",
      "Residue 206: LEU\n",
      "Residue 207: ASP\n",
      "Residue 208: GLN\n",
      "Residue 209: ALA\n",
      "Residue 210: ARG\n",
      "Residue 211: GLU\n",
      "Residue 212: TYR\n",
      "Residue 213: HIS\n",
      "Residue 214: GLU\n",
      "Residue 215: LYS\n",
      "Residue 216: LEU\n",
      "Residue 217: VAL\n",
      "Residue 218: GLU\n",
      "Residue 219: VAL\n",
      "Residue 220: ALA\n",
      "Residue 221: ALA\n",
      "Residue 222: ASP\n",
      "Residue 223: PHE\n",
      "Residue 224: ASP\n",
      "Residue 225: GLU\n",
      "Residue 226: ASN\n",
      "Residue 227: ILE\n",
      "Residue 228: MET\n",
      "Residue 229: LEU\n",
      "Residue 230: LYS\n",
      "Residue 231: TYR\n",
      "Residue 232: LEU\n",
      "Residue 233: GLU\n",
      "Residue 234: GLY\n",
      "Residue 235: GLU\n",
      "Residue 236: GLU\n",
      "Residue 237: PRO\n",
      "Residue 238: THR\n",
      "Residue 239: GLU\n",
      "Residue 240: GLU\n",
      "Residue 241: GLU\n",
      "Residue 242: LEU\n",
      "Residue 243: VAL\n",
      "Residue 244: ALA\n",
      "Residue 245: ALA\n",
      "Residue 246: ILE\n",
      "Residue 247: ARG\n",
      "Residue 248: LYS\n",
      "Residue 249: GLY\n",
      "Residue 250: THR\n",
      "Residue 251: ILE\n",
      "Residue 252: ASP\n",
      "Residue 253: LEU\n",
      "Residue 254: LYS\n",
      "Residue 255: ILE\n",
      "Residue 256: THR\n",
      "Residue 257: PRO\n",
      "Residue 258: VAL\n",
      "Residue 259: PHE\n",
      "Residue 260: LEU\n",
      "Residue 261: GLY\n",
      "Residue 262: SER\n",
      "Residue 263: ALA\n",
      "Residue 264: LEU\n",
      "Residue 265: LYS\n",
      "Residue 266: ASN\n",
      "Residue 267: LYS\n",
      "Residue 268: GLY\n",
      "Residue 269: VAL\n",
      "Residue 270: GLN\n",
      "Residue 271: LEU\n",
      "Residue 272: LEU\n",
      "Residue 273: LEU\n",
      "Residue 274: ASP\n",
      "Residue 275: ALA\n",
      "Residue 276: VAL\n",
      "Residue 277: VAL\n",
      "Residue 278: ASP\n",
      "Residue 279: TYR\n",
      "Residue 280: LEU\n",
      "Residue 281: PRO\n",
      "Residue 282: SER\n",
      "Residue 283: PRO\n",
      "Residue 284: LEU\n",
      "Residue 285: ASP\n",
      "Residue 286: ILE\n",
      "Residue 287: PRO\n",
      "Residue 288: PRO\n",
      "Residue 289: ILE\n",
      "Residue 290: LYS\n",
      "Residue 291: GLY\n",
      "Residue 292: THR\n",
      "Residue 293: THR\n",
      "Residue 294: PRO\n",
      "Residue 295: GLU\n",
      "Residue 296: GLY\n",
      "Residue 297: GLU\n",
      "Residue 298: VAL\n",
      "Residue 299: VAL\n",
      "Residue 300: GLU\n",
      "Residue 301: ILE\n",
      "Residue 302: HIS\n",
      "Residue 303: PRO\n",
      "Residue 304: ASP\n",
      "Residue 305: PRO\n",
      "Residue 306: ASN\n",
      "Residue 307: GLY\n",
      "Residue 308: PRO\n",
      "Residue 309: LEU\n",
      "Residue 310: ALA\n",
      "Residue 311: ALA\n",
      "Residue 312: LEU\n",
      "Residue 313: ALA\n",
      "Residue 314: PHE\n",
      "Residue 315: LYS\n",
      "Residue 316: ILE\n",
      "Residue 317: MET\n",
      "Residue 318: ALA\n",
      "Residue 319: ASP\n",
      "Residue 320: PRO\n",
      "Residue 321: TYR\n",
      "Residue 322: VAL\n",
      "Residue 323: GLY\n",
      "Residue 324: ARG\n",
      "Residue 325: LEU\n",
      "Residue 326: THR\n",
      "Residue 327: PHE\n",
      "Residue 328: ILE\n",
      "Residue 329: ARG\n",
      "Residue 330: VAL\n",
      "Residue 331: TYR\n",
      "Residue 332: SER\n",
      "Residue 333: GLY\n",
      "Residue 334: THR\n",
      "Residue 335: LEU\n",
      "Residue 336: THR\n",
      "Residue 337: SER\n",
      "Residue 338: GLY\n",
      "Residue 339: SER\n",
      "Residue 340: TYR\n",
      "Residue 341: VAL\n",
      "Residue 342: TYR\n",
      "Residue 343: ASN\n",
      "Residue 344: THR\n",
      "Residue 345: THR\n",
      "Residue 346: LYS\n",
      "Residue 347: GLY\n",
      "Residue 348: ARG\n",
      "Residue 349: LYS\n",
      "Residue 350: GLU\n",
      "Residue 351: ARG\n",
      "Residue 352: VAL\n",
      "Residue 353: ALA\n",
      "Residue 354: ARG\n",
      "Residue 355: LEU\n",
      "Residue 356: LEU\n",
      "Residue 357: ARG\n",
      "Residue 358: MET\n",
      "Residue 359: HIS\n",
      "Residue 360: ALA\n",
      "Residue 361: ASN\n",
      "Residue 362: HIS\n",
      "Residue 363: ARG\n",
      "Residue 364: GLU\n",
      "Residue 365: GLU\n",
      "Residue 366: VAL\n",
      "Residue 367: GLU\n",
      "Residue 368: GLU\n",
      "Residue 369: LEU\n",
      "Residue 370: LYS\n",
      "Residue 371: ALA\n",
      "Residue 372: GLY\n",
      "Residue 373: ASP\n",
      "Residue 374: LEU\n",
      "Residue 375: GLY\n",
      "Residue 376: ALA\n",
      "Residue 377: VAL\n",
      "Residue 378: VAL\n",
      "Residue 379: GLY\n",
      "Residue 380: LEU\n",
      "Residue 381: LYS\n",
      "Residue 382: GLU\n",
      "Residue 383: THR\n",
      "Residue 384: ILE\n",
      "Residue 385: THR\n",
      "Residue 386: GLY\n",
      "Residue 387: ASP\n",
      "Residue 388: THR\n",
      "Residue 389: LEU\n",
      "Residue 390: VAL\n",
      "Residue 391: GLY\n",
      "Residue 392: GLU\n",
      "Residue 393: ASP\n",
      "Residue 394: ALA\n",
      "Residue 395: PRO\n",
      "Residue 396: ARG\n",
      "Residue 397: VAL\n",
      "Residue 398: ILE\n",
      "Residue 399: LEU\n",
      "Residue 400: GLU\n",
      "Residue 401: SER\n",
      "Residue 402: ILE\n",
      "Residue 403: GLU\n",
      "Residue 404: VAL\n",
      "Residue 405: PRO\n",
      "Residue 406: GLU\n",
      "Residue 407: PRO\n",
      "Residue 408: VAL\n",
      "Residue 409: ILE\n",
      "Residue 410: ASP\n",
      "Residue 411: VAL\n",
      "Residue 412: ALA\n",
      "Residue 413: ILE\n",
      "Residue 414: GLU\n",
      "Residue 415: PRO\n",
      "Residue 416: LYS\n",
      "Residue 417: THR\n",
      "Residue 418: LYS\n",
      "Residue 419: ALA\n",
      "Residue 420: ASP\n",
      "Residue 421: GLN\n",
      "Residue 422: GLU\n",
      "Residue 423: LYS\n",
      "Residue 424: LEU\n",
      "Residue 425: SER\n",
      "Residue 426: GLN\n",
      "Residue 427: ALA\n",
      "Residue 428: LEU\n",
      "Residue 429: ALA\n",
      "Residue 430: ARG\n",
      "Residue 431: LEU\n",
      "Residue 432: ALA\n",
      "Residue 433: GLU\n",
      "Residue 434: GLU\n",
      "Residue 435: ASP\n",
      "Residue 436: PRO\n",
      "Residue 437: THR\n",
      "Residue 438: PHE\n",
      "Residue 439: ARG\n",
      "Residue 440: VAL\n",
      "Residue 441: SER\n",
      "Residue 442: THR\n",
      "Residue 443: HIS\n",
      "Residue 444: PRO\n",
      "Residue 445: GLU\n",
      "Residue 446: THR\n",
      "Residue 447: GLY\n",
      "Residue 448: GLN\n",
      "Residue 449: THR\n",
      "Residue 450: ILE\n",
      "Residue 451: ILE\n",
      "Residue 452: SER\n",
      "Residue 453: GLY\n",
      "Residue 454: MET\n",
      "Residue 455: GLY\n",
      "Residue 456: GLU\n",
      "Residue 457: LEU\n",
      "Residue 458: HIS\n",
      "Residue 459: LEU\n",
      "Residue 460: GLU\n",
      "Residue 461: ILE\n",
      "Residue 462: ILE\n",
      "Residue 463: VAL\n",
      "Residue 464: ASP\n",
      "Residue 465: ARG\n",
      "Residue 466: LEU\n",
      "Residue 467: LYS\n",
      "Residue 468: ARG\n",
      "Residue 469: GLU\n",
      "Residue 470: PHE\n",
      "Residue 471: LYS\n",
      "Residue 472: VAL\n",
      "Residue 473: ASP\n",
      "Residue 474: ALA\n",
      "Residue 475: ASN\n",
      "Residue 476: VAL\n",
      "Residue 477: GLY\n",
      "Residue 478: LYS\n",
      "Residue 479: PRO\n",
      "Residue 480: GLN\n",
      "Residue 481: VAL\n",
      "Residue 482: ALA\n",
      "Residue 483: TYR\n",
      "Residue 484: ARG\n",
      "Residue 485: GLU\n",
      "Residue 486: THR\n",
      "Residue 487: ILE\n",
      "Residue 488: THR\n",
      "Residue 489: LYS\n",
      "Residue 490: PRO\n",
      "Residue 491: VAL\n",
      "Residue 492: ASP\n",
      "Residue 493: VAL\n",
      "Residue 494: GLU\n",
      "Residue 495: GLY\n",
      "Residue 496: LYS\n",
      "Residue 497: PHE\n",
      "Residue 498: ILE\n",
      "Residue 499: ARG\n",
      "Residue 500: GLN\n",
      "Residue 501: THR\n",
      "Residue 502: GLY\n",
      "Residue 503: GLY\n",
      "Residue 504: ARG\n",
      "Residue 505: GLY\n",
      "Residue 506: GLN\n",
      "Residue 507: TYR\n",
      "Residue 508: GLY\n",
      "Residue 509: HIS\n",
      "Residue 510: VAL\n",
      "Residue 511: LYS\n",
      "Residue 512: ILE\n",
      "Residue 513: LYS\n",
      "Residue 514: VAL\n",
      "Residue 515: GLU\n",
      "Residue 516: PRO\n",
      "Residue 517: LEU\n",
      "Residue 518: PRO\n",
      "Residue 519: ARG\n",
      "Residue 520: GLY\n",
      "Residue 521: SER\n",
      "Residue 522: GLY\n",
      "Residue 523: PHE\n",
      "Residue 524: GLU\n",
      "Residue 525: PHE\n",
      "Residue 526: VAL\n",
      "Residue 527: ASN\n",
      "Residue 528: ALA\n",
      "Residue 529: ILE\n",
      "Residue 530: VAL\n",
      "Residue 531: GLY\n",
      "Residue 532: GLY\n",
      "Residue 533: VAL\n",
      "Residue 534: ILE\n",
      "Residue 535: PRO\n",
      "Residue 536: LYS\n",
      "Residue 537: GLU\n",
      "Residue 538: TYR\n",
      "Residue 539: ILE\n",
      "Residue 540: PRO\n",
      "Residue 541: ALA\n",
      "Residue 542: VAL\n",
      "Residue 543: GLN\n",
      "Residue 544: LYS\n",
      "Residue 545: GLY\n",
      "Residue 546: ILE\n",
      "Residue 547: GLU\n",
      "Residue 548: GLU\n",
      "Residue 549: ALA\n",
      "Residue 550: MET\n",
      "Residue 551: GLN\n",
      "Residue 552: SER\n",
      "Residue 553: GLY\n",
      "Residue 554: PRO\n",
      "Residue 555: LEU\n",
      "Residue 556: ILE\n",
      "Residue 557: GLY\n",
      "Residue 558: PHE\n",
      "Residue 559: PRO\n",
      "Residue 560: VAL\n",
      "Residue 561: VAL\n",
      "Residue 562: ASP\n",
      "Residue 563: ILE\n",
      "Residue 564: LYS\n",
      "Residue 565: VAL\n",
      "Residue 566: THR\n",
      "Residue 567: LEU\n",
      "Residue 568: TYR\n",
      "Residue 569: ASP\n",
      "Residue 570: GLY\n",
      "Residue 571: SER\n",
      "Residue 572: TYR\n",
      "Residue 573: HIS\n",
      "Residue 574: GLU\n",
      "Residue 575: VAL\n",
      "Residue 576: ASP\n",
      "Residue 577: SER\n",
      "Residue 578: SER\n",
      "Residue 579: GLU\n",
      "Residue 580: MET\n",
      "Residue 581: ALA\n",
      "Residue 582: PHE\n",
      "Residue 583: LYS\n",
      "Residue 584: ILE\n",
      "Residue 585: ALA\n",
      "Residue 586: GLY\n",
      "Residue 587: SER\n",
      "Residue 588: MET\n",
      "Residue 589: ALA\n",
      "Residue 590: ILE\n",
      "Residue 591: LYS\n",
      "Residue 592: GLU\n",
      "Residue 593: ALA\n",
      "Residue 594: VAL\n",
      "Residue 595: GLN\n",
      "Residue 596: LYS\n",
      "Residue 597: GLY\n",
      "Residue 598: ASP\n",
      "Residue 599: PRO\n",
      "Residue 600: VAL\n",
      "Residue 601: ILE\n",
      "Residue 602: LEU\n",
      "Residue 603: GLU\n",
      "Residue 604: PRO\n",
      "Residue 605: ILE\n",
      "Residue 606: MET\n",
      "Residue 607: ARG\n",
      "Residue 608: VAL\n",
      "Residue 609: GLU\n",
      "Residue 610: VAL\n",
      "Residue 611: THR\n",
      "Residue 612: THR\n",
      "Residue 613: PRO\n",
      "Residue 614: GLU\n",
      "Residue 615: GLU\n",
      "Residue 616: TYR\n",
      "Residue 617: MET\n",
      "Residue 618: GLY\n",
      "Residue 619: ASP\n",
      "Residue 620: VAL\n",
      "Residue 621: ILE\n",
      "Residue 622: GLY\n",
      "Residue 623: ASP\n",
      "Residue 624: LEU\n",
      "Residue 625: ASN\n",
      "Residue 626: ALA\n",
      "Residue 627: ARG\n",
      "Residue 628: ARG\n",
      "Residue 629: GLY\n",
      "Residue 630: GLN\n",
      "Residue 631: ILE\n",
      "Residue 632: LEU\n",
      "Residue 633: GLY\n",
      "Residue 634: MET\n",
      "Residue 635: GLU\n",
      "Residue 636: PRO\n",
      "Residue 637: ARG\n",
      "Residue 638: GLY\n",
      "Residue 639: ASN\n",
      "Residue 640: ALA\n",
      "Residue 641: GLN\n",
      "Residue 642: VAL\n",
      "Residue 643: ILE\n",
      "Residue 644: ARG\n",
      "Residue 645: ALA\n",
      "Residue 646: PHE\n",
      "Residue 647: VAL\n",
      "Residue 648: PRO\n",
      "Residue 649: LEU\n",
      "Residue 650: ALA\n",
      "Residue 651: GLU\n",
      "Residue 652: MET\n",
      "Residue 653: PHE\n",
      "Residue 654: GLY\n",
      "Residue 655: TYR\n",
      "Residue 656: ALA\n",
      "Residue 657: THR\n",
      "Residue 658: ASP\n",
      "Residue 659: LEU\n",
      "Residue 660: ARG\n",
      "Residue 661: SER\n",
      "Residue 662: LYS\n",
      "Residue 663: THR\n",
      "Residue 664: GLN\n",
      "Residue 665: GLY\n",
      "Residue 666: ARG\n",
      "Residue 667: GLY\n",
      "Residue 668: SER\n",
      "Residue 669: PHE\n",
      "Residue 670: VAL\n",
      "Residue 671: MET\n",
      "Residue 672: PHE\n",
      "Residue 673: PHE\n",
      "Residue 674: ASP\n",
      "Residue 675: HIS\n",
      "Residue 676: TYR\n",
      "Residue 677: GLN\n",
      "Residue 678: GLU\n",
      "Residue 679: VAL\n",
      "Residue 680: PRO\n",
      "Residue 681: LYS\n",
      "Residue 682: GLN\n",
      "Residue 683: VAL\n",
      "Residue 684: GLN\n",
      "Residue 685: GLU\n",
      "Residue 686: LYS\n",
      "Residue 687: LEU\n",
      "Residue 688: ILE\n",
      "Residue 689: LYS\n",
      "Residue 690: GLY\n",
      "Residue 701: MG\n",
      "Residue 702: FUA\n",
      "Residue 703: GDP\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## for finding proline number in 4v5f structure:\n",
    "from Bio.PDB import MMCIFParser\n",
    "\n",
    "# Parse the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure('4V5F', '4v5f_EF-G.cif')\n",
    "\n",
    "# Iterate over models, chains, and residues to list amino acids and their residue numbers\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        print(f\"Chain ID: {chain.id}\")\n",
    "        for residue in chain:\n",
    "            # residue.id is a tuple like (' ', resseq, insertion_code)\n",
    "            resseq = residue.id[1]  # residue sequence number\n",
    "            resname = residue.get_resname()  # 3-letter amino acid code\n",
    "            print(f\"Residue {resseq}: {resname}\")\n",
    "        print()  # Blank line between chains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a45da778-b447-4047-985e-d1a43f0563c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "54bb8f91f90546e49dd0e872c881ee3b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "77ad35b125db49049d50682e426c25f3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntSlider(value=50, description='My Slider')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "view = nv.show_file(\"4v5f_EF-G.cif\")\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='silver')\n",
    "selection_string = \"648:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string)\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "display(view)\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "14cdd56a-1fac-4d7b-a560-362c7aa53b10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSD after superimposition: 33.220390795845105\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, Superimposer, MMCIFIO, Model, Structure\n",
    "\n",
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', '4wpo_EF-G.cif')\n",
    "structure2 = parser.get_structure('s2', '4v5f_EF-G.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'BZ'\n",
    "chain2_id = 'CY'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_aligned.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "79a17878-1b21-4043-99f0-a05375912dab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d3d6c2c6183643afaf3422ed48dd5398",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0e6afb3aad7042dea886e97507c31f47",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntSlider(value=50, description='My Slider')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "view = nv.show_file(\"combined_aligned.cif\")\n",
    "\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "\n",
    "view.clear_representations()\n",
    "\n",
    "# Cartoon for chain CY in silver\n",
    "view.add_representation('cartoon', selection=':CY', color='silver')\n",
    "\n",
    "# Cartoon for chain BZ in pink\n",
    "view.add_representation('cartoon', selection=':BZ', color='pink')\n",
    "\n",
    "# Highlight residue 648 in chain CY as red ball+stick + label\n",
    "selection_string = \"648:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n",
    "display(slider)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c669d3f6-cef0-49fc-94f0-7d41758e8e78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "916d81f782d84dfd8ddec154118336bb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli36099ef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')\n",
    "selection_string = \"659:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "375acbd3-ab8f-44da-b9d8-07d6d40f1dfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "03cd25ecc3814e3c88a802069763095f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli3609910xmicef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='siver')\n",
    "selection_string = \"659:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e88b87b0-6885-4d91-8efa-f87b567b297c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSD after superimposition: 0.5198646287552445\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, Superimposer, MMCIFIO, Model, Structure\n",
    "\n",
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', 'e.coli36099ef-g.cif')\n",
    "structure2 = parser.get_structure('s2', 'e.coli3609910xmicef-g.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'A'\n",
    "chain2_id = 'A'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_alignedAF3model.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3a9b1778-09d0-4927-8df8-d9a2c72b5b69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using model /0 and /1 - SUCCESS\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be21a4df386b4ff09b5d28939b853f15",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5a941e5b62b04d68a2bef20464e1783c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntSlider(value=50, description='My Slider')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "view = nv.show_file(\"combined_alignedAF3model.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Try different model selection syntaxes\n",
    "try:\n",
    "    view.add_representation('cartoon', selection='/0', color='pink')      # Model 0 with slash\n",
    "    view.add_representation('cartoon', selection='/1', color='silver')    # Model 1 with slash\n",
    "    print(\"Using model /0 and /1 - SUCCESS\")\n",
    "except Exception as e:\n",
    "    print(f\"Model selection failed: {e}\")\n",
    "    \n",
    "    # Fallback: try chain selection\n",
    "    try:\n",
    "        view.add_representation('cartoon', selection=':A', color='pink')\n",
    "        view.add_representation('cartoon', selection=':B', color='silver')\n",
    "        print(\"Using chain A and B - SUCCESS\")\n",
    "    except Exception as e2:\n",
    "        print(f\"Chain selection failed: {e2}\")\n",
    "        \n",
    "        # Final fallback: show everything with automatic coloring\n",
    "        view.add_representation('cartoon', color='chainid')  # Colors by chain automatically\n",
    "        print(\"Using automatic chain coloring - SUCCESS\")\n",
    "\n",
    "# Highlight residue\n",
    "selection_string = \"659:A\"\n",
    "try:\n",
    "    view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "    view.add_representation('label', \n",
    "                           selection=selection_string, \n",
    "                           labelType='format', \n",
    "                           labelFormat='%(resname)s%(resno)s', \n",
    "                           color='red', \n",
    "                           xOffset=1, \n",
    "                           fixedSize=True)\n",
    "except Exception as e:\n",
    "    print(f\"Residue highlighting failed: {e}\")\n",
    "\n",
    "# Force display\n",
    "display(view)\n",
    "\n",
    "# Slider\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c688434a-fd3d-4ebc-b2d9-e38e8112eba1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "20bafa062c4848a5bf115be3f5267140",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli36099ef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')\n",
    "selection_string = \"593:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c169bd5d-58fb-481b-a1ac-08bed7897dfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "697fccf19b274d0abfb3a90daffa4b92",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"ef-ge.colimg165510xmic.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='siver')\n",
    "selection_string = \"593:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "93ea78a5-34c0-4f20-ba58-6df81ed5d69e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSD after superimposition: 0.4638705762984491\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, Superimposer, MMCIFIO, Model, Structure\n",
    "\n",
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', 'e.coli36099ef-g.cif')\n",
    "structure2 = parser.get_structure('s2', 'ef-ge.colimg165510xmic.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'A'\n",
    "chain2_id = 'A'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_alignedAF3model2.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fafcc483-58f5-42c4-8113-6d583b691845",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using model /0 and /1 - SUCCESS\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "40e6465cb93042c4939089fa3059e3ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0cd54083e80f4ace84553eab8a5e1e63",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "IntSlider(value=50, description='My Slider')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "\n",
    "view = nv.show_file(\"combined_alignedAF3model2.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Try different model selection syntaxes\n",
    "try:\n",
    "    view.add_representation('cartoon', selection='/0', color='pink')      # Model 0 with slash\n",
    "    view.add_representation('cartoon', selection='/1', color='silver')    # Model 1 with slash\n",
    "    print(\"Using model /0 and /1 - SUCCESS\")\n",
    "except Exception as e:\n",
    "    print(f\"Model selection failed: {e}\")\n",
    "    \n",
    "    # Fallback: try chain selection\n",
    "    try:\n",
    "        view.add_representation('cartoon', selection=':A', color='pink')\n",
    "        view.add_representation('cartoon', selection=':B', color='silver')\n",
    "        print(\"Using chain A and B - SUCCESS\")\n",
    "    except Exception as e2:\n",
    "        print(f\"Chain selection failed: {e2}\")\n",
    "        \n",
    "        # Final fallback: show everything with automatic coloring\n",
    "        view.add_representation('cartoon', color='chainid')  # Colors by chain automatically\n",
    "        print(\"Using automatic chain coloring - SUCCESS\")\n",
    "\n",
    "# Highlight residue\n",
    "selection_string = \"593:A\"\n",
    "try:\n",
    "    view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "    view.add_representation('label', \n",
    "                           selection=selection_string, \n",
    "                           labelType='format', \n",
    "                           labelFormat='%(resname)s%(resno)s', \n",
    "                           color='red', \n",
    "                           xOffset=1, \n",
    "                           fixedSize=True)\n",
    "except Exception as e:\n",
    "    print(f\"Residue highlighting failed: {e}\")\n",
    "\n",
    "# Force display\n",
    "display(view)\n",
    "# Slider\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c89c24ca-890d-4d50-a8e7-1392385a24e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311412.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311413.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 311414.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 311480.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B9 is discontinuous at line 311481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 311482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 311483.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 311484.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 311550.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D9 is discontinuous at line 311551.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c899aeb5-d934-40d6-8ae9-0970d1ec6c65",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AD is discontinuous at line 311412.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 311413.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AY is discontinuous at line 311414.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 311480.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B9 is discontinuous at line 311481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CD is discontinuous at line 311482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 311483.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CY is discontinuous at line 311484.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 311550.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D9 is discontinuous at line 311551.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO, Select\n",
    "\n",
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G_S12 chains\n",
    "class EF_G_S12_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G_S12 chains \n",
    "io.save(\"4v5f_EF-G_S12.cif\", EF_G_S12_ChainSelect([\"CY\", \"CL\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "2f4ef6ff-a2c0-47bf-a130-5123255cbbaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chain ID: CL\n",
      "Residue 5: PRO\n",
      "Residue 6: THR\n",
      "Residue 7: ILE\n",
      "Residue 8: ASN\n",
      "Residue 9: GLN\n",
      "Residue 10: LEU\n",
      "Residue 11: VAL\n",
      "Residue 12: ARG\n",
      "Residue 13: LYS\n",
      "Residue 14: GLY\n",
      "Residue 15: ARG\n",
      "Residue 16: GLU\n",
      "Residue 17: LYS\n",
      "Residue 18: VAL\n",
      "Residue 19: ARG\n",
      "Residue 20: LYS\n",
      "Residue 21: LYS\n",
      "Residue 22: SER\n",
      "Residue 23: LYS\n",
      "Residue 24: VAL\n",
      "Residue 25: PRO\n",
      "Residue 26: ALA\n",
      "Residue 27: LEU\n",
      "Residue 28: LYS\n",
      "Residue 29: GLY\n",
      "Residue 30: ALA\n",
      "Residue 31: PRO\n",
      "Residue 32: PHE\n",
      "Residue 33: ARG\n",
      "Residue 34: ARG\n",
      "Residue 35: GLY\n",
      "Residue 36: VAL\n",
      "Residue 37: CYS\n",
      "Residue 38: THR\n",
      "Residue 39: VAL\n",
      "Residue 40: VAL\n",
      "Residue 41: ARG\n",
      "Residue 42: THR\n",
      "Residue 43: VAL\n",
      "Residue 44: THR\n",
      "Residue 45: PRO\n",
      "Residue 46: LYS\n",
      "Residue 47: LYS\n",
      "Residue 48: PRO\n",
      "Residue 49: ASN\n",
      "Residue 50: SER\n",
      "Residue 51: ALA\n",
      "Residue 52: LEU\n",
      "Residue 53: ARG\n",
      "Residue 54: LYS\n",
      "Residue 55: VAL\n",
      "Residue 56: ALA\n",
      "Residue 57: LYS\n",
      "Residue 58: VAL\n",
      "Residue 59: ARG\n",
      "Residue 60: LEU\n",
      "Residue 61: THR\n",
      "Residue 62: SER\n",
      "Residue 63: GLY\n",
      "Residue 64: TYR\n",
      "Residue 65: GLU\n",
      "Residue 66: VAL\n",
      "Residue 67: THR\n",
      "Residue 68: ALA\n",
      "Residue 69: TYR\n",
      "Residue 70: ILE\n",
      "Residue 71: PRO\n",
      "Residue 72: GLY\n",
      "Residue 73: GLU\n",
      "Residue 74: GLY\n",
      "Residue 75: HIS\n",
      "Residue 76: ASN\n",
      "Residue 77: LEU\n",
      "Residue 78: GLN\n",
      "Residue 79: GLU\n",
      "Residue 80: HIS\n",
      "Residue 81: SER\n",
      "Residue 82: VAL\n",
      "Residue 83: VAL\n",
      "Residue 84: LEU\n",
      "Residue 85: ILE\n",
      "Residue 86: ARG\n",
      "Residue 87: GLY\n",
      "Residue 88: GLY\n",
      "Residue 89: ARG\n",
      "Residue 90: VAL\n",
      "Residue 91: LYS\n",
      "Residue 92: ASP\n",
      "Residue 93: LEU\n",
      "Residue 94: PRO\n",
      "Residue 95: GLY\n",
      "Residue 96: VAL\n",
      "Residue 97: ARG\n",
      "Residue 98: TYR\n",
      "Residue 99: HIS\n",
      "Residue 100: ILE\n",
      "Residue 101: VAL\n",
      "Residue 102: ARG\n",
      "Residue 103: GLY\n",
      "Residue 104: VAL\n",
      "Residue 105: TYR\n",
      "Residue 106: ASP\n",
      "Residue 107: ALA\n",
      "Residue 108: ALA\n",
      "Residue 109: GLY\n",
      "Residue 110: VAL\n",
      "Residue 111: LYS\n",
      "Residue 112: ASP\n",
      "Residue 113: ARG\n",
      "Residue 114: LYS\n",
      "Residue 115: LYS\n",
      "Residue 116: SER\n",
      "Residue 117: ARG\n",
      "Residue 118: SER\n",
      "Residue 119: LYS\n",
      "Residue 120: TYR\n",
      "Residue 121: GLY\n",
      "Residue 122: THR\n",
      "Residue 123: LYS\n",
      "Residue 124: LYS\n",
      "Residue 125: PRO\n",
      "Residue 126: LYS\n",
      "Residue 127: GLU\n",
      "Residue 128: ALA\n",
      "Residue 129: ALA\n",
      "\n",
      "Chain ID: CY\n",
      "Residue 4: LYS\n",
      "Residue 5: VAL\n",
      "Residue 6: GLU\n",
      "Residue 7: TYR\n",
      "Residue 8: ASP\n",
      "Residue 9: LEU\n",
      "Residue 10: LYS\n",
      "Residue 11: ARG\n",
      "Residue 12: LEU\n",
      "Residue 13: ARG\n",
      "Residue 14: ASN\n",
      "Residue 15: ILE\n",
      "Residue 16: GLY\n",
      "Residue 17: ILE\n",
      "Residue 18: ALA\n",
      "Residue 19: ALA\n",
      "Residue 20: HIS\n",
      "Residue 21: ILE\n",
      "Residue 22: ASP\n",
      "Residue 23: ALA\n",
      "Residue 24: GLY\n",
      "Residue 25: LYS\n",
      "Residue 26: THR\n",
      "Residue 27: THR\n",
      "Residue 28: THR\n",
      "Residue 29: THR\n",
      "Residue 30: GLU\n",
      "Residue 31: ARG\n",
      "Residue 32: ILE\n",
      "Residue 33: LEU\n",
      "Residue 34: TYR\n",
      "Residue 35: TYR\n",
      "Residue 36: THR\n",
      "Residue 37: GLY\n",
      "Residue 38: ARG\n",
      "Residue 39: ILE\n",
      "Residue 40: HIS\n",
      "Residue 41: LYS\n",
      "Residue 42: ILE\n",
      "Residue 43: GLY\n",
      "Residue 64: THR\n",
      "Residue 65: ILE\n",
      "Residue 66: THR\n",
      "Residue 67: ALA\n",
      "Residue 68: ALA\n",
      "Residue 69: VAL\n",
      "Residue 70: THR\n",
      "Residue 71: THR\n",
      "Residue 72: CYS\n",
      "Residue 73: PHE\n",
      "Residue 74: TRP\n",
      "Residue 75: LYS\n",
      "Residue 76: ASP\n",
      "Residue 77: HIS\n",
      "Residue 78: ARG\n",
      "Residue 79: ILE\n",
      "Residue 80: ASN\n",
      "Residue 81: ILE\n",
      "Residue 82: ILE\n",
      "Residue 83: ASP\n",
      "Residue 84: THR\n",
      "Residue 85: PRO\n",
      "Residue 86: GLY\n",
      "Residue 87: HIS\n",
      "Residue 88: VAL\n",
      "Residue 89: ASP\n",
      "Residue 90: PHE\n",
      "Residue 91: THR\n",
      "Residue 92: ILE\n",
      "Residue 93: GLU\n",
      "Residue 94: VAL\n",
      "Residue 95: GLU\n",
      "Residue 96: ARG\n",
      "Residue 97: SER\n",
      "Residue 98: MET\n",
      "Residue 99: ARG\n",
      "Residue 100: VAL\n",
      "Residue 101: LEU\n",
      "Residue 102: ASP\n",
      "Residue 103: GLY\n",
      "Residue 104: ALA\n",
      "Residue 105: ILE\n",
      "Residue 106: VAL\n",
      "Residue 107: VAL\n",
      "Residue 108: PHE\n",
      "Residue 109: ASP\n",
      "Residue 110: SER\n",
      "Residue 111: SER\n",
      "Residue 112: GLN\n",
      "Residue 113: GLY\n",
      "Residue 114: VAL\n",
      "Residue 115: GLU\n",
      "Residue 116: PRO\n",
      "Residue 117: GLN\n",
      "Residue 118: SER\n",
      "Residue 119: GLU\n",
      "Residue 120: THR\n",
      "Residue 121: VAL\n",
      "Residue 122: TRP\n",
      "Residue 123: ARG\n",
      "Residue 124: GLN\n",
      "Residue 125: ALA\n",
      "Residue 126: GLU\n",
      "Residue 127: LYS\n",
      "Residue 128: TYR\n",
      "Residue 129: LYS\n",
      "Residue 130: VAL\n",
      "Residue 131: PRO\n",
      "Residue 132: ARG\n",
      "Residue 133: ILE\n",
      "Residue 134: ALA\n",
      "Residue 135: PHE\n",
      "Residue 136: ALA\n",
      "Residue 137: ASN\n",
      "Residue 138: LYS\n",
      "Residue 139: MET\n",
      "Residue 140: ASP\n",
      "Residue 141: LYS\n",
      "Residue 142: THR\n",
      "Residue 143: GLY\n",
      "Residue 144: ALA\n",
      "Residue 145: ASP\n",
      "Residue 146: LEU\n",
      "Residue 147: TRP\n",
      "Residue 148: LEU\n",
      "Residue 149: VAL\n",
      "Residue 150: ILE\n",
      "Residue 151: ARG\n",
      "Residue 152: THR\n",
      "Residue 153: MET\n",
      "Residue 154: GLN\n",
      "Residue 155: GLU\n",
      "Residue 156: ARG\n",
      "Residue 157: LEU\n",
      "Residue 158: GLY\n",
      "Residue 159: ALA\n",
      "Residue 160: ARG\n",
      "Residue 161: PRO\n",
      "Residue 162: VAL\n",
      "Residue 163: VAL\n",
      "Residue 164: MET\n",
      "Residue 165: GLN\n",
      "Residue 166: LEU\n",
      "Residue 167: PRO\n",
      "Residue 168: ILE\n",
      "Residue 169: GLY\n",
      "Residue 170: ARG\n",
      "Residue 171: GLU\n",
      "Residue 172: ASP\n",
      "Residue 173: THR\n",
      "Residue 174: PHE\n",
      "Residue 175: SER\n",
      "Residue 176: GLY\n",
      "Residue 177: ILE\n",
      "Residue 178: ILE\n",
      "Residue 179: ASP\n",
      "Residue 180: VAL\n",
      "Residue 181: LEU\n",
      "Residue 182: ARG\n",
      "Residue 183: MET\n",
      "Residue 184: LYS\n",
      "Residue 185: ALA\n",
      "Residue 186: TYR\n",
      "Residue 187: THR\n",
      "Residue 188: TYR\n",
      "Residue 189: GLY\n",
      "Residue 190: ASN\n",
      "Residue 191: ASP\n",
      "Residue 192: LEU\n",
      "Residue 193: GLY\n",
      "Residue 194: THR\n",
      "Residue 195: ASP\n",
      "Residue 196: ILE\n",
      "Residue 197: ARG\n",
      "Residue 198: GLU\n",
      "Residue 199: ILE\n",
      "Residue 200: PRO\n",
      "Residue 201: ILE\n",
      "Residue 202: PRO\n",
      "Residue 203: GLU\n",
      "Residue 204: GLU\n",
      "Residue 205: TYR\n",
      "Residue 206: LEU\n",
      "Residue 207: ASP\n",
      "Residue 208: GLN\n",
      "Residue 209: ALA\n",
      "Residue 210: ARG\n",
      "Residue 211: GLU\n",
      "Residue 212: TYR\n",
      "Residue 213: HIS\n",
      "Residue 214: GLU\n",
      "Residue 215: LYS\n",
      "Residue 216: LEU\n",
      "Residue 217: VAL\n",
      "Residue 218: GLU\n",
      "Residue 219: VAL\n",
      "Residue 220: ALA\n",
      "Residue 221: ALA\n",
      "Residue 222: ASP\n",
      "Residue 223: PHE\n",
      "Residue 224: ASP\n",
      "Residue 225: GLU\n",
      "Residue 226: ASN\n",
      "Residue 227: ILE\n",
      "Residue 228: MET\n",
      "Residue 229: LEU\n",
      "Residue 230: LYS\n",
      "Residue 231: TYR\n",
      "Residue 232: LEU\n",
      "Residue 233: GLU\n",
      "Residue 234: GLY\n",
      "Residue 235: GLU\n",
      "Residue 236: GLU\n",
      "Residue 237: PRO\n",
      "Residue 238: THR\n",
      "Residue 239: GLU\n",
      "Residue 240: GLU\n",
      "Residue 241: GLU\n",
      "Residue 242: LEU\n",
      "Residue 243: VAL\n",
      "Residue 244: ALA\n",
      "Residue 245: ALA\n",
      "Residue 246: ILE\n",
      "Residue 247: ARG\n",
      "Residue 248: LYS\n",
      "Residue 249: GLY\n",
      "Residue 250: THR\n",
      "Residue 251: ILE\n",
      "Residue 252: ASP\n",
      "Residue 253: LEU\n",
      "Residue 254: LYS\n",
      "Residue 255: ILE\n",
      "Residue 256: THR\n",
      "Residue 257: PRO\n",
      "Residue 258: VAL\n",
      "Residue 259: PHE\n",
      "Residue 260: LEU\n",
      "Residue 261: GLY\n",
      "Residue 262: SER\n",
      "Residue 263: ALA\n",
      "Residue 264: LEU\n",
      "Residue 265: LYS\n",
      "Residue 266: ASN\n",
      "Residue 267: LYS\n",
      "Residue 268: GLY\n",
      "Residue 269: VAL\n",
      "Residue 270: GLN\n",
      "Residue 271: LEU\n",
      "Residue 272: LEU\n",
      "Residue 273: LEU\n",
      "Residue 274: ASP\n",
      "Residue 275: ALA\n",
      "Residue 276: VAL\n",
      "Residue 277: VAL\n",
      "Residue 278: ASP\n",
      "Residue 279: TYR\n",
      "Residue 280: LEU\n",
      "Residue 281: PRO\n",
      "Residue 282: SER\n",
      "Residue 283: PRO\n",
      "Residue 284: LEU\n",
      "Residue 285: ASP\n",
      "Residue 286: ILE\n",
      "Residue 287: PRO\n",
      "Residue 288: PRO\n",
      "Residue 289: ILE\n",
      "Residue 290: LYS\n",
      "Residue 291: GLY\n",
      "Residue 292: THR\n",
      "Residue 293: THR\n",
      "Residue 294: PRO\n",
      "Residue 295: GLU\n",
      "Residue 296: GLY\n",
      "Residue 297: GLU\n",
      "Residue 298: VAL\n",
      "Residue 299: VAL\n",
      "Residue 300: GLU\n",
      "Residue 301: ILE\n",
      "Residue 302: HIS\n",
      "Residue 303: PRO\n",
      "Residue 304: ASP\n",
      "Residue 305: PRO\n",
      "Residue 306: ASN\n",
      "Residue 307: GLY\n",
      "Residue 308: PRO\n",
      "Residue 309: LEU\n",
      "Residue 310: ALA\n",
      "Residue 311: ALA\n",
      "Residue 312: LEU\n",
      "Residue 313: ALA\n",
      "Residue 314: PHE\n",
      "Residue 315: LYS\n",
      "Residue 316: ILE\n",
      "Residue 317: MET\n",
      "Residue 318: ALA\n",
      "Residue 319: ASP\n",
      "Residue 320: PRO\n",
      "Residue 321: TYR\n",
      "Residue 322: VAL\n",
      "Residue 323: GLY\n",
      "Residue 324: ARG\n",
      "Residue 325: LEU\n",
      "Residue 326: THR\n",
      "Residue 327: PHE\n",
      "Residue 328: ILE\n",
      "Residue 329: ARG\n",
      "Residue 330: VAL\n",
      "Residue 331: TYR\n",
      "Residue 332: SER\n",
      "Residue 333: GLY\n",
      "Residue 334: THR\n",
      "Residue 335: LEU\n",
      "Residue 336: THR\n",
      "Residue 337: SER\n",
      "Residue 338: GLY\n",
      "Residue 339: SER\n",
      "Residue 340: TYR\n",
      "Residue 341: VAL\n",
      "Residue 342: TYR\n",
      "Residue 343: ASN\n",
      "Residue 344: THR\n",
      "Residue 345: THR\n",
      "Residue 346: LYS\n",
      "Residue 347: GLY\n",
      "Residue 348: ARG\n",
      "Residue 349: LYS\n",
      "Residue 350: GLU\n",
      "Residue 351: ARG\n",
      "Residue 352: VAL\n",
      "Residue 353: ALA\n",
      "Residue 354: ARG\n",
      "Residue 355: LEU\n",
      "Residue 356: LEU\n",
      "Residue 357: ARG\n",
      "Residue 358: MET\n",
      "Residue 359: HIS\n",
      "Residue 360: ALA\n",
      "Residue 361: ASN\n",
      "Residue 362: HIS\n",
      "Residue 363: ARG\n",
      "Residue 364: GLU\n",
      "Residue 365: GLU\n",
      "Residue 366: VAL\n",
      "Residue 367: GLU\n",
      "Residue 368: GLU\n",
      "Residue 369: LEU\n",
      "Residue 370: LYS\n",
      "Residue 371: ALA\n",
      "Residue 372: GLY\n",
      "Residue 373: ASP\n",
      "Residue 374: LEU\n",
      "Residue 375: GLY\n",
      "Residue 376: ALA\n",
      "Residue 377: VAL\n",
      "Residue 378: VAL\n",
      "Residue 379: GLY\n",
      "Residue 380: LEU\n",
      "Residue 381: LYS\n",
      "Residue 382: GLU\n",
      "Residue 383: THR\n",
      "Residue 384: ILE\n",
      "Residue 385: THR\n",
      "Residue 386: GLY\n",
      "Residue 387: ASP\n",
      "Residue 388: THR\n",
      "Residue 389: LEU\n",
      "Residue 390: VAL\n",
      "Residue 391: GLY\n",
      "Residue 392: GLU\n",
      "Residue 393: ASP\n",
      "Residue 394: ALA\n",
      "Residue 395: PRO\n",
      "Residue 396: ARG\n",
      "Residue 397: VAL\n",
      "Residue 398: ILE\n",
      "Residue 399: LEU\n",
      "Residue 400: GLU\n",
      "Residue 401: SER\n",
      "Residue 402: ILE\n",
      "Residue 403: GLU\n",
      "Residue 404: VAL\n",
      "Residue 405: PRO\n",
      "Residue 406: GLU\n",
      "Residue 407: PRO\n",
      "Residue 408: VAL\n",
      "Residue 409: ILE\n",
      "Residue 410: ASP\n",
      "Residue 411: VAL\n",
      "Residue 412: ALA\n",
      "Residue 413: ILE\n",
      "Residue 414: GLU\n",
      "Residue 415: PRO\n",
      "Residue 416: LYS\n",
      "Residue 417: THR\n",
      "Residue 418: LYS\n",
      "Residue 419: ALA\n",
      "Residue 420: ASP\n",
      "Residue 421: GLN\n",
      "Residue 422: GLU\n",
      "Residue 423: LYS\n",
      "Residue 424: LEU\n",
      "Residue 425: SER\n",
      "Residue 426: GLN\n",
      "Residue 427: ALA\n",
      "Residue 428: LEU\n",
      "Residue 429: ALA\n",
      "Residue 430: ARG\n",
      "Residue 431: LEU\n",
      "Residue 432: ALA\n",
      "Residue 433: GLU\n",
      "Residue 434: GLU\n",
      "Residue 435: ASP\n",
      "Residue 436: PRO\n",
      "Residue 437: THR\n",
      "Residue 438: PHE\n",
      "Residue 439: ARG\n",
      "Residue 440: VAL\n",
      "Residue 441: SER\n",
      "Residue 442: THR\n",
      "Residue 443: HIS\n",
      "Residue 444: PRO\n",
      "Residue 445: GLU\n",
      "Residue 446: THR\n",
      "Residue 447: GLY\n",
      "Residue 448: GLN\n",
      "Residue 449: THR\n",
      "Residue 450: ILE\n",
      "Residue 451: ILE\n",
      "Residue 452: SER\n",
      "Residue 453: GLY\n",
      "Residue 454: MET\n",
      "Residue 455: GLY\n",
      "Residue 456: GLU\n",
      "Residue 457: LEU\n",
      "Residue 458: HIS\n",
      "Residue 459: LEU\n",
      "Residue 460: GLU\n",
      "Residue 461: ILE\n",
      "Residue 462: ILE\n",
      "Residue 463: VAL\n",
      "Residue 464: ASP\n",
      "Residue 465: ARG\n",
      "Residue 466: LEU\n",
      "Residue 467: LYS\n",
      "Residue 468: ARG\n",
      "Residue 469: GLU\n",
      "Residue 470: PHE\n",
      "Residue 471: LYS\n",
      "Residue 472: VAL\n",
      "Residue 473: ASP\n",
      "Residue 474: ALA\n",
      "Residue 475: ASN\n",
      "Residue 476: VAL\n",
      "Residue 477: GLY\n",
      "Residue 478: LYS\n",
      "Residue 479: PRO\n",
      "Residue 480: GLN\n",
      "Residue 481: VAL\n",
      "Residue 482: ALA\n",
      "Residue 483: TYR\n",
      "Residue 484: ARG\n",
      "Residue 485: GLU\n",
      "Residue 486: THR\n",
      "Residue 487: ILE\n",
      "Residue 488: THR\n",
      "Residue 489: LYS\n",
      "Residue 490: PRO\n",
      "Residue 491: VAL\n",
      "Residue 492: ASP\n",
      "Residue 493: VAL\n",
      "Residue 494: GLU\n",
      "Residue 495: GLY\n",
      "Residue 496: LYS\n",
      "Residue 497: PHE\n",
      "Residue 498: ILE\n",
      "Residue 499: ARG\n",
      "Residue 500: GLN\n",
      "Residue 501: THR\n",
      "Residue 502: GLY\n",
      "Residue 503: GLY\n",
      "Residue 504: ARG\n",
      "Residue 505: GLY\n",
      "Residue 506: GLN\n",
      "Residue 507: TYR\n",
      "Residue 508: GLY\n",
      "Residue 509: HIS\n",
      "Residue 510: VAL\n",
      "Residue 511: LYS\n",
      "Residue 512: ILE\n",
      "Residue 513: LYS\n",
      "Residue 514: VAL\n",
      "Residue 515: GLU\n",
      "Residue 516: PRO\n",
      "Residue 517: LEU\n",
      "Residue 518: PRO\n",
      "Residue 519: ARG\n",
      "Residue 520: GLY\n",
      "Residue 521: SER\n",
      "Residue 522: GLY\n",
      "Residue 523: PHE\n",
      "Residue 524: GLU\n",
      "Residue 525: PHE\n",
      "Residue 526: VAL\n",
      "Residue 527: ASN\n",
      "Residue 528: ALA\n",
      "Residue 529: ILE\n",
      "Residue 530: VAL\n",
      "Residue 531: GLY\n",
      "Residue 532: GLY\n",
      "Residue 533: VAL\n",
      "Residue 534: ILE\n",
      "Residue 535: PRO\n",
      "Residue 536: LYS\n",
      "Residue 537: GLU\n",
      "Residue 538: TYR\n",
      "Residue 539: ILE\n",
      "Residue 540: PRO\n",
      "Residue 541: ALA\n",
      "Residue 542: VAL\n",
      "Residue 543: GLN\n",
      "Residue 544: LYS\n",
      "Residue 545: GLY\n",
      "Residue 546: ILE\n",
      "Residue 547: GLU\n",
      "Residue 548: GLU\n",
      "Residue 549: ALA\n",
      "Residue 550: MET\n",
      "Residue 551: GLN\n",
      "Residue 552: SER\n",
      "Residue 553: GLY\n",
      "Residue 554: PRO\n",
      "Residue 555: LEU\n",
      "Residue 556: ILE\n",
      "Residue 557: GLY\n",
      "Residue 558: PHE\n",
      "Residue 559: PRO\n",
      "Residue 560: VAL\n",
      "Residue 561: VAL\n",
      "Residue 562: ASP\n",
      "Residue 563: ILE\n",
      "Residue 564: LYS\n",
      "Residue 565: VAL\n",
      "Residue 566: THR\n",
      "Residue 567: LEU\n",
      "Residue 568: TYR\n",
      "Residue 569: ASP\n",
      "Residue 570: GLY\n",
      "Residue 571: SER\n",
      "Residue 572: TYR\n",
      "Residue 573: HIS\n",
      "Residue 574: GLU\n",
      "Residue 575: VAL\n",
      "Residue 576: ASP\n",
      "Residue 577: SER\n",
      "Residue 578: SER\n",
      "Residue 579: GLU\n",
      "Residue 580: MET\n",
      "Residue 581: ALA\n",
      "Residue 582: PHE\n",
      "Residue 583: LYS\n",
      "Residue 584: ILE\n",
      "Residue 585: ALA\n",
      "Residue 586: GLY\n",
      "Residue 587: SER\n",
      "Residue 588: MET\n",
      "Residue 589: ALA\n",
      "Residue 590: ILE\n",
      "Residue 591: LYS\n",
      "Residue 592: GLU\n",
      "Residue 593: ALA\n",
      "Residue 594: VAL\n",
      "Residue 595: GLN\n",
      "Residue 596: LYS\n",
      "Residue 597: GLY\n",
      "Residue 598: ASP\n",
      "Residue 599: PRO\n",
      "Residue 600: VAL\n",
      "Residue 601: ILE\n",
      "Residue 602: LEU\n",
      "Residue 603: GLU\n",
      "Residue 604: PRO\n",
      "Residue 605: ILE\n",
      "Residue 606: MET\n",
      "Residue 607: ARG\n",
      "Residue 608: VAL\n",
      "Residue 609: GLU\n",
      "Residue 610: VAL\n",
      "Residue 611: THR\n",
      "Residue 612: THR\n",
      "Residue 613: PRO\n",
      "Residue 614: GLU\n",
      "Residue 615: GLU\n",
      "Residue 616: TYR\n",
      "Residue 617: MET\n",
      "Residue 618: GLY\n",
      "Residue 619: ASP\n",
      "Residue 620: VAL\n",
      "Residue 621: ILE\n",
      "Residue 622: GLY\n",
      "Residue 623: ASP\n",
      "Residue 624: LEU\n",
      "Residue 625: ASN\n",
      "Residue 626: ALA\n",
      "Residue 627: ARG\n",
      "Residue 628: ARG\n",
      "Residue 629: GLY\n",
      "Residue 630: GLN\n",
      "Residue 631: ILE\n",
      "Residue 632: LEU\n",
      "Residue 633: GLY\n",
      "Residue 634: MET\n",
      "Residue 635: GLU\n",
      "Residue 636: PRO\n",
      "Residue 637: ARG\n",
      "Residue 638: GLY\n",
      "Residue 639: ASN\n",
      "Residue 640: ALA\n",
      "Residue 641: GLN\n",
      "Residue 642: VAL\n",
      "Residue 643: ILE\n",
      "Residue 644: ARG\n",
      "Residue 645: ALA\n",
      "Residue 646: PHE\n",
      "Residue 647: VAL\n",
      "Residue 648: PRO\n",
      "Residue 649: LEU\n",
      "Residue 650: ALA\n",
      "Residue 651: GLU\n",
      "Residue 652: MET\n",
      "Residue 653: PHE\n",
      "Residue 654: GLY\n",
      "Residue 655: TYR\n",
      "Residue 656: ALA\n",
      "Residue 657: THR\n",
      "Residue 658: ASP\n",
      "Residue 659: LEU\n",
      "Residue 660: ARG\n",
      "Residue 661: SER\n",
      "Residue 662: LYS\n",
      "Residue 663: THR\n",
      "Residue 664: GLN\n",
      "Residue 665: GLY\n",
      "Residue 666: ARG\n",
      "Residue 667: GLY\n",
      "Residue 668: SER\n",
      "Residue 669: PHE\n",
      "Residue 670: VAL\n",
      "Residue 671: MET\n",
      "Residue 672: PHE\n",
      "Residue 673: PHE\n",
      "Residue 674: ASP\n",
      "Residue 675: HIS\n",
      "Residue 676: TYR\n",
      "Residue 677: GLN\n",
      "Residue 678: GLU\n",
      "Residue 679: VAL\n",
      "Residue 680: PRO\n",
      "Residue 681: LYS\n",
      "Residue 682: GLN\n",
      "Residue 683: VAL\n",
      "Residue 684: GLN\n",
      "Residue 685: GLU\n",
      "Residue 686: LYS\n",
      "Residue 687: LEU\n",
      "Residue 688: ILE\n",
      "Residue 689: LYS\n",
      "Residue 690: GLY\n",
      "Residue 701: MG\n",
      "Residue 702: FUA\n",
      "Residue 703: GDP\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## for finding proline number in 4v5f structure:\n",
    "from Bio.PDB import MMCIFParser\n",
    "\n",
    "# Parse the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure('4V5F', '4v5f_EF-G_S12.cif')\n",
    "\n",
    "# Iterate over models, chains, and residues to list amino acids and their residue numbers\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        print(f\"Chain ID: {chain.id}\")\n",
    "        for residue in chain:\n",
    "            # residue.id is a tuple like (' ', resseq, insertion_code)\n",
    "            resseq = residue.id[1]  # residue sequence number\n",
    "            resname = residue.get_resname()  # 3-letter amino acid code\n",
    "            print(f\"Residue {resseq}: {resname}\")\n",
    "        print()  # Blank line between chains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "83a092dd-f2d1-4bd4-997e-b7d5989cdbc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 282032.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 282185.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 282326.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 282327.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 282481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DB is discontinuous at line 282482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 282624.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 282625.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 282917.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AK is discontinuous at line 282918.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AL is discontinuous at line 282919.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 282921.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AT is discontinuous at line 282923.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 282925.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BC is discontinuous at line 283417.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 283424.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B2 is discontinuous at line 283427.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 283428.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BH is discontinuous at line 283431.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 283432.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 283729.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CK is discontinuous at line 283731.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CL is discontinuous at line 283732.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 283734.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CT is discontinuous at line 283738.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DB is discontinuous at line 283740.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DC is discontinuous at line 284242.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 284246.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DL is discontinuous at line 284248.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v53\", \"4v53.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e7357c2f-d395-47d5-ae9f-06a753422cc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entity ID: 1, Chains: AA,CA, Molecule: 16S rRNA\n",
      "Entity ID: 2, Chains: AC,CC, Molecule: 30S ribosomal protein S3\n",
      "Entity ID: 3, Chains: AD,CD, Molecule: 30S ribosomal protein S4\n",
      "Entity ID: 4, Chains: AE,CE, Molecule: 30S ribosomal protein S5\n",
      "Entity ID: 5, Chains: AF,CF, Molecule: 30S ribosomal protein S6\n",
      "Entity ID: 6, Chains: AG,CG, Molecule: 30S ribosomal protein S7\n",
      "Entity ID: 7, Chains: AH,CH, Molecule: 30S ribosomal protein S8\n",
      "Entity ID: 8, Chains: AI,CI, Molecule: 30S ribosomal protein S9\n",
      "Entity ID: 9, Chains: AJ,CJ, Molecule: 30S ribosomal protein S10\n",
      "Entity ID: 10, Chains: AK,CK, Molecule: 30S ribosomal protein S11\n",
      "Entity ID: 11, Chains: AL,CL, Molecule: 30S ribosomal protein S12\n",
      "Entity ID: 12, Chains: AM,CM, Molecule: 30S ribosomal protein S13\n",
      "Entity ID: 13, Chains: AN,CN, Molecule: 30S ribosomal protein S14\n",
      "Entity ID: 14, Chains: AO,CO, Molecule: 30S ribosomal protein S15\n",
      "Entity ID: 15, Chains: AP,CP, Molecule: 30S ribosomal protein S16\n",
      "Entity ID: 16, Chains: AQ,CQ, Molecule: 30S ribosomal protein S17\n",
      "Entity ID: 17, Chains: AR,CR, Molecule: 30S ribosomal protein S18\n",
      "Entity ID: 18, Chains: AS,CS, Molecule: 30S ribosomal protein S19\n",
      "Entity ID: 19, Chains: AT,CT, Molecule: 30S ribosomal protein S20\n",
      "Entity ID: 20, Chains: AB,CB, Molecule: 30S ribosomal protein S2\n",
      "Entity ID: 21, Chains: AU,CU, Molecule: 30S ribosomal protein S21\n",
      "Entity ID: 22, Chains: BA,DA, Molecule: 5S rRNA\n",
      "Entity ID: 23, Chains: BB,DB, Molecule: 23S rRNA\n",
      "Entity ID: 24, Chains: BI,DI, Molecule: 50S ribosomal protein L11\n",
      "Entity ID: 25, Chains: BC,DC, Molecule: 50S ribosomal protein L2\n",
      "Entity ID: 26, Chains: BD,DD, Molecule: 50S ribosomal protein L3\n",
      "Entity ID: 27, Chains: BK,DK, Molecule: 50S ribosomal protein L14\n",
      "Entity ID: 28, Chains: BP,DP, Molecule: 50S ribosomal protein L19\n",
      "Entity ID: 29, Chains: BE,DE, Molecule: 50S ribosomal protein L4\n",
      "Entity ID: 30, Chains: BY,DY, Molecule: 50S ribosomal protein L30\n",
      "Entity ID: 31, Chains: B0,D0, Molecule: 50S ribosomal protein L32\n",
      "Entity ID: 32, Chains: B4,D4, Molecule: 50S ribosomal protein L36\n",
      "Entity ID: 33, Chains: B1,D1, Molecule: 50S ribosomal protein L33\n",
      "Entity ID: 34, Chains: B3,D3, Molecule: 50S ribosomal protein L35\n",
      "Entity ID: 35, Chains: BV,DV, Molecule: 50S ribosomal protein L25\n",
      "Entity ID: 36, Chains: B2,D2, Molecule: 50S ribosomal protein L34\n",
      "Entity ID: 37, Chains: BL,DL, Molecule: 50S ribosomal protein L15\n",
      "Entity ID: 38, Chains: BM,DM, Molecule: 50S ribosomal protein L16\n",
      "Entity ID: 39, Chains: BX,DX, Molecule: 50S ribosomal protein L29\n",
      "Entity ID: 40, Chains: BH,DH, Molecule: 50S ribosomal protein L9\n",
      "Entity ID: 41, Chains: BJ,DJ, Molecule: 50S ribosomal protein L13\n",
      "Entity ID: 42, Chains: BN,DN, Molecule: 50S ribosomal protein L17\n",
      "Entity ID: 43, Chains: BO,DO, Molecule: 50S ribosomal protein L18\n",
      "Entity ID: 44, Chains: BQ,DQ, Molecule: 50S ribosomal protein L20\n",
      "Entity ID: 45, Chains: BS,DS, Molecule: 50S ribosomal protein L22\n",
      "Entity ID: 46, Chains: BU,DU, Molecule: 50S ribosomal protein L24\n",
      "Entity ID: 47, Chains: BF,DF, Molecule: 50S ribosomal protein L5\n",
      "Entity ID: 48, Chains: BG,DG, Molecule: 50S ribosomal protein L6\n",
      "Entity ID: 49, Chains: BR,DR, Molecule: 50S ribosomal protein L21\n",
      "Entity ID: 50, Chains: BT,DT, Molecule: 50S ribosomal protein L23\n",
      "Entity ID: 51, Chains: BZ,DZ, Molecule: 50S ribosomal protein L28\n",
      "Entity ID: 52, Chains: BW,DW, Molecule: 50S ribosomal protein L27\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import PDBList\n",
    "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
    "\n",
    "mmcif_dict = MMCIF2Dict(\"4v53.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "416f5d92-73e9-48f9-8ea1-4752b6d0717e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 282032.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 282185.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B4 is discontinuous at line 282326.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 282327.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 282481.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DB is discontinuous at line 282482.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D4 is discontinuous at line 282624.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AA is discontinuous at line 282625.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AE is discontinuous at line 282917.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AK is discontinuous at line 282918.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AL is discontinuous at line 282919.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AN is discontinuous at line 282921.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain AT is discontinuous at line 282923.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BB is discontinuous at line 282925.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BC is discontinuous at line 283417.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BE is discontinuous at line 283424.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B2 is discontinuous at line 283427.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BL is discontinuous at line 283428.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain BH is discontinuous at line 283431.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CA is discontinuous at line 283432.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CE is discontinuous at line 283729.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CK is discontinuous at line 283731.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CL is discontinuous at line 283732.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CN is discontinuous at line 283734.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain CT is discontinuous at line 283738.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DB is discontinuous at line 283740.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DC is discontinuous at line 284242.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DE is discontinuous at line 284246.\n",
      "  warnings.warn(\n",
      "C:\\Users\\razon\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain DL is discontinuous at line 284248.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO, Select\n",
    "\n",
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v53\", \"4v53.cif\")\n",
    "\n",
    "# Custom Select class to keep only S12 chains\n",
    "class S12_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only S12 chains \n",
    "io.save(\"4v53_S12.cif\", S12_ChainSelect([\"CL\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "13e7dec5-3744-43b4-a802-15ae6caac1f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSD after superimposition: 20.623224232638357\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import FastMMCIFParser\n",
    "\n",
    "parser = FastMMCIFParser(QUIET=True)\n",
    "structure1 = parser.get_structure('s1', '4v5f_EF-G_S12.cif')\n",
    "structure2 = parser.get_structure('s2', '4v53_S12.cif')\n",
    "\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = [\"CY\", \"CL\"]  \n",
    "chain2_id = \"CL\"\n",
    "\n",
    "atoms1 = [res[\"CA\"] \n",
    "          for cid in chain1_id \n",
    "          for res in model1[cid] \n",
    "          if \"CA\" in res]\n",
    "\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"alignedS12+EF-G.cif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "157da6d9-6d33-4108-a3d5-bcb9b8b89bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename overlapping chains in model2 to avoid coloring clashes\n",
    "existing_chains = {chain.id for chain in model1}\n",
    "for chain in model2:\n",
    "    if chain.id in existing_chains:\n",
    "        chain.id = chain.id + \"_2\"  # e.g., \"CL\" -> \"CL_2\"\n",
    "\n",
    "# Create combined structure\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "combined_structure.add(model1)\n",
    "model2.id = 1  # second model\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"alignedS12+EF-G_renamed.cif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8d6f185c-4841-4257-b59c-700e7569d150",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8eae2278c82d4ce0b304b666df1f249e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "view = nv.show_file(\"alignedS12+EF-G_renamed.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "#color by chain names\n",
    "view.add_representation(\"cartoon\", selection=\":CY\", color=\"pink\")\n",
    "view.add_representation(\"cartoon\", selection=\":CL\", color=\"blue\")\n",
    "view.add_representation(\"cartoon\", selection=\":CL_2\", color=\"silver\")\n",
    "\n",
    "from IPython.display import display\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "874fe6c3-47ae-4e19-926c-ef7c6ddf498f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RMSD after superimposition: 0.0005234103095913419\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import MMCIFParser, MMCIFIO, Superimposer, Structure\n",
    "\n",
    "# Load the previously superimposed structure\n",
    "parser = MMCIFParser(QUIET=True)\n",
    "aligned_structure = parser.get_structure('aligned', 'alignedS12+EF-G_renamed.cif')\n",
    "model2 = aligned_structure[0]  # first model (contains CL_2)\n",
    "\n",
    "# Load the new structure\n",
    "new_structure = parser.get_structure('4V53', '4V53.cif')\n",
    "model_new = new_structure[0]  # first model of 4V53\n",
    "\n",
    "# Select chains to use for superimposition\n",
    "chain_old = \"CL_2\"  # chain from previously superimposed structure\n",
    "chain_new = \"CL\"    # chain in 4V53\n",
    "\n",
    "# Extract CA atoms for alignment\n",
    "atoms_old = [res[\"CA\"] for res in model2[chain_old] if \"CA\" in res]\n",
    "atoms_new = [res[\"CA\"] for res in model_new[chain_new] if \"CA\" in res]\n",
    "\n",
    "# Match lengths\n",
    "min_len = min(len(atoms_old), len(atoms_new))\n",
    "atoms_old = atoms_old[:min_len]\n",
    "atoms_new = atoms_new[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms_old, atoms_new)\n",
    "sup.apply(model_new.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Rename chain in new model if needed to avoid duplicates\n",
    "if chain_new in [c.id for c in model2]:\n",
    "    for chain in model_new:\n",
    "        if chain.id == chain_new:\n",
    "            chain.id = chain_new + \"_new\"\n",
    "\n",
    "# Assign unique model IDs before combining\n",
    "model2.id = 0\n",
    "model_new.id = 1\n",
    "\n",
    "# Combine structures into one\n",
    "combined_structure = Structure.Structure(\"super_aligned_2\")\n",
    "combined_structure.add(model2)\n",
    "combined_structure.add(model_new)\n",
    "\n",
    "# Save combined structure\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"superimposed_combined.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "076eb53f-7f5b-4faa-a48a-5aeabfc396bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5a7dbdcd79c54388b2d48eab1204ac67",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from IPython.display import display\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# Load the CIF file directly\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "\n",
    "# Clear default representations if you want custom colors\n",
    "view.clear_representations()\n",
    "\n",
    "# Add cartoon representations for chains\n",
    "view.add_representation(\"cartoon\", selection=\":CY\", color=\"pink\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Parse the CIF file\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Extract chain IDs for model 1\n",
    "model_index = 0  # Python index\n",
    "chains_model1 = [chain.id for chain in structure[model_index]]\n",
    "\n",
    "# Assign colors\n",
    "colors = [plt.cm.tab20(i % 20) for i in range(len(chains_model1))]\n",
    "\n",
    "# Load into NGLView directly\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Add cartoon representation for each chain with colors\n",
    "for chain_id, color in zip(chains_model1, colors):\n",
    "    view.add_representation(\"cartoon\", selection=f\":{chain_id}\", color=color)\n",
    "\n",
    "from IPython.display import display\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "bb9350e9-bbc7-4bd0-8c33-5113a94593f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found ligands/hetero atoms:\n",
      "Model 0, Chain CY: MG ('H_MG', 701, ' ')\n",
      "Model 0, Chain CY: FUA ('H_FUA', 702, ' ')\n",
      "Model 0, Chain CY: GDP ('H_GDP', 703, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2001, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2002, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2003, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2004, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2005, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2006, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2007, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2008, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2009, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2010, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2011, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2012, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2013, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2014, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2015, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2016, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2017, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2018, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2019, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2020, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2021, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2022, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2023, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2024, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2025, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2026, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2027, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2028, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2029, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2030, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2031, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2032, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2033, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2034, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2035, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2036, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2037, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2038, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2039, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2040, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2041, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2042, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2043, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2044, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2045, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2046, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2047, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2048, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2049, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2050, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2051, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2052, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2053, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2054, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2055, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2056, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2057, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2058, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2059, ' ')\n",
      "Model 0, Chain AA: MG ('H_MG', 2060, ' ')\n",
      "Model 0, Chain AA: LLL ('H_LLL', 2061, ' ')\n",
      "Model 0, Chain AA: LLL ('H_LLL', 2062, ' ')\n",
      "Model 0, Chain AA: LLL ('H_LLL', 2063, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3001, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3002, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3003, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3004, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3005, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3006, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3007, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3008, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3009, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3010, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3011, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3012, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3013, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3014, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3015, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3016, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3017, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3018, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3019, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3020, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3021, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3022, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3023, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3024, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3025, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3026, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3027, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3028, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3029, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3030, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3031, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3032, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3033, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3034, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3035, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3036, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3037, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3038, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3039, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3040, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3041, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3042, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3043, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3044, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3045, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3046, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3047, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3048, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3049, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3050, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3051, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3052, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3053, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3054, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3055, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3056, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3057, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3058, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3059, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3060, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3061, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3062, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3063, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3064, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3065, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3066, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3067, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3068, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3069, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3070, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3071, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3072, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3073, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3074, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3075, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3076, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3077, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3078, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3079, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3080, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3081, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3082, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3083, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3084, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3085, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3086, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3087, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3088, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3089, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3090, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3091, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3092, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3093, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3094, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3095, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3096, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3097, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3098, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3099, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3100, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3101, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3102, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3103, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3104, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3105, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3106, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3107, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3108, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3109, ' ')\n",
      "Model 0, Chain BB: MG ('H_MG', 3110, ' ')\n",
      "Model 0, Chain BB: LLL ('H_LLL', 3111, ' ')\n",
      "Model 0, Chain B4: ZN ('H_ZN', 101, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2001, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2002, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2003, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2004, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2005, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2006, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2007, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2008, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2009, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2010, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2011, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2012, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2013, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2014, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2015, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2016, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2017, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2018, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2019, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2020, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2021, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2022, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2023, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2024, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2025, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2026, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2027, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2028, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2029, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2030, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2031, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2032, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2033, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2034, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2035, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2036, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2037, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2038, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2039, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2040, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2041, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2042, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2043, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2044, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2045, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2046, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2047, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2048, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2049, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2050, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2051, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2052, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2053, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2054, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2055, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2056, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2057, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2058, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2059, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2060, ' ')\n",
      "Model 0, Chain CA: MG ('H_MG', 2061, ' ')\n",
      "Model 0, Chain CA: LLL ('H_LLL', 2062, ' ')\n",
      "Model 0, Chain CA: LLL ('H_LLL', 2063, ' ')\n",
      "Model 0, Chain CA: LLL ('H_LLL', 2064, ' ')\n",
      "Model 0, Chain CN: MG ('H_MG', 201, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3001, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3002, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3003, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3004, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3005, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3006, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3007, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3008, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3009, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3010, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3011, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3012, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3013, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3014, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3015, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3016, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3017, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3018, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3019, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3020, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3021, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3022, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3023, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3024, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3025, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3026, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3027, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3028, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3029, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3030, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3031, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3032, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3033, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3034, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3035, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3036, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3037, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3038, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3039, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3040, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3041, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3042, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3043, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3044, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3045, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3046, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3047, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3048, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3049, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3050, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3051, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3052, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3053, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3054, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3055, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3056, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3057, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3058, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3059, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3060, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3061, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3062, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3063, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3064, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3065, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3066, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3067, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3068, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3069, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3070, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3071, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3072, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3073, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3074, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3075, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3076, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3077, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3078, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3079, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3080, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3081, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3082, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3083, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3084, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3085, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3086, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3087, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3088, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3089, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3090, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3091, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3092, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3093, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3094, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3095, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3096, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3097, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3098, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3099, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3100, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3101, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3102, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3103, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3104, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3105, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3106, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3107, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3108, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3109, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3110, ' ')\n",
      "Model 0, Chain DB: MG ('H_MG', 3111, ' ')\n",
      "Model 0, Chain DB: LLL ('H_LLL', 3112, ' ')\n",
      "Model 0, Chain D4: ZN ('H_ZN', 101, ' ')\n",
      "\n",
      "Detailed residue information:\n",
      "Model 0:\n",
      "  Chain CL:\n",
      "    No hetero residues found\n",
      "  Chain CY:\n",
      "    MG ('H_MG', 701, ' ')\n",
      "    FUA ('H_FUA', 702, ' ')\n",
      "    GDP ('H_GDP', 703, ' ')\n",
      "  Chain CL_2:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "  Chain AA:\n",
      "    MG ('H_MG', 2001, ' ')\n",
      "    MG ('H_MG', 2002, ' ')\n",
      "    MG ('H_MG', 2003, ' ')\n",
      "    MG ('H_MG', 2004, ' ')\n",
      "    MG ('H_MG', 2005, ' ')\n",
      "    MG ('H_MG', 2006, ' ')\n",
      "    MG ('H_MG', 2007, ' ')\n",
      "    MG ('H_MG', 2008, ' ')\n",
      "    MG ('H_MG', 2009, ' ')\n",
      "    MG ('H_MG', 2010, ' ')\n",
      "    MG ('H_MG', 2011, ' ')\n",
      "    MG ('H_MG', 2012, ' ')\n",
      "    MG ('H_MG', 2013, ' ')\n",
      "    MG ('H_MG', 2014, ' ')\n",
      "    MG ('H_MG', 2015, ' ')\n",
      "    MG ('H_MG', 2016, ' ')\n",
      "    MG ('H_MG', 2017, ' ')\n",
      "    MG ('H_MG', 2018, ' ')\n",
      "    MG ('H_MG', 2019, ' ')\n",
      "    MG ('H_MG', 2020, ' ')\n",
      "    MG ('H_MG', 2021, ' ')\n",
      "    MG ('H_MG', 2022, ' ')\n",
      "    MG ('H_MG', 2023, ' ')\n",
      "    MG ('H_MG', 2024, ' ')\n",
      "    MG ('H_MG', 2025, ' ')\n",
      "    MG ('H_MG', 2026, ' ')\n",
      "    MG ('H_MG', 2027, ' ')\n",
      "    MG ('H_MG', 2028, ' ')\n",
      "    MG ('H_MG', 2029, ' ')\n",
      "    MG ('H_MG', 2030, ' ')\n",
      "    MG ('H_MG', 2031, ' ')\n",
      "    MG ('H_MG', 2032, ' ')\n",
      "    MG ('H_MG', 2033, ' ')\n",
      "    MG ('H_MG', 2034, ' ')\n",
      "    MG ('H_MG', 2035, ' ')\n",
      "    MG ('H_MG', 2036, ' ')\n",
      "    MG ('H_MG', 2037, ' ')\n",
      "    MG ('H_MG', 2038, ' ')\n",
      "    MG ('H_MG', 2039, ' ')\n",
      "    MG ('H_MG', 2040, ' ')\n",
      "    MG ('H_MG', 2041, ' ')\n",
      "    MG ('H_MG', 2042, ' ')\n",
      "    MG ('H_MG', 2043, ' ')\n",
      "    MG ('H_MG', 2044, ' ')\n",
      "    MG ('H_MG', 2045, ' ')\n",
      "    MG ('H_MG', 2046, ' ')\n",
      "    MG ('H_MG', 2047, ' ')\n",
      "    MG ('H_MG', 2048, ' ')\n",
      "    MG ('H_MG', 2049, ' ')\n",
      "    MG ('H_MG', 2050, ' ')\n",
      "    MG ('H_MG', 2051, ' ')\n",
      "    MG ('H_MG', 2052, ' ')\n",
      "    MG ('H_MG', 2053, ' ')\n",
      "    MG ('H_MG', 2054, ' ')\n",
      "    MG ('H_MG', 2055, ' ')\n",
      "    MG ('H_MG', 2056, ' ')\n",
      "    MG ('H_MG', 2057, ' ')\n",
      "    MG ('H_MG', 2058, ' ')\n",
      "    MG ('H_MG', 2059, ' ')\n",
      "    MG ('H_MG', 2060, ' ')\n",
      "    LLL ('H_LLL', 2061, ' ')\n",
      "    LLL ('H_LLL', 2062, ' ')\n",
      "    LLL ('H_LLL', 2063, ' ')\n",
      "    HOH ('W', 2101, ' ')\n",
      "    HOH ('W', 2102, ' ')\n",
      "    HOH ('W', 2103, ' ')\n",
      "    HOH ('W', 2104, ' ')\n",
      "    HOH ('W', 2105, ' ')\n",
      "    HOH ('W', 2106, ' ')\n",
      "    HOH ('W', 2107, ' ')\n",
      "    HOH ('W', 2108, ' ')\n",
      "    HOH ('W', 2109, ' ')\n",
      "    HOH ('W', 2110, ' ')\n",
      "    HOH ('W', 2111, ' ')\n",
      "    HOH ('W', 2112, ' ')\n",
      "    HOH ('W', 2113, ' ')\n",
      "    HOH ('W', 2114, ' ')\n",
      "    HOH ('W', 2115, ' ')\n",
      "    HOH ('W', 2116, ' ')\n",
      "    HOH ('W', 2117, ' ')\n",
      "    HOH ('W', 2118, ' ')\n",
      "    HOH ('W', 2119, ' ')\n",
      "    HOH ('W', 2120, ' ')\n",
      "    HOH ('W', 2121, ' ')\n",
      "    HOH ('W', 2122, ' ')\n",
      "    HOH ('W', 2123, ' ')\n",
      "    HOH ('W', 2124, ' ')\n",
      "    HOH ('W', 2125, ' ')\n",
      "    HOH ('W', 2126, ' ')\n",
      "    HOH ('W', 2127, ' ')\n",
      "    HOH ('W', 2128, ' ')\n",
      "    HOH ('W', 2129, ' ')\n",
      "    HOH ('W', 2130, ' ')\n",
      "    HOH ('W', 2131, ' ')\n",
      "    HOH ('W', 2132, ' ')\n",
      "    HOH ('W', 2133, ' ')\n",
      "    HOH ('W', 2134, ' ')\n",
      "    HOH ('W', 2135, ' ')\n",
      "    HOH ('W', 2136, ' ')\n",
      "    HOH ('W', 2137, ' ')\n",
      "    HOH ('W', 2138, ' ')\n",
      "    HOH ('W', 2139, ' ')\n",
      "    HOH ('W', 2140, ' ')\n",
      "    HOH ('W', 2141, ' ')\n",
      "    HOH ('W', 2142, ' ')\n",
      "    HOH ('W', 2143, ' ')\n",
      "    HOH ('W', 2144, ' ')\n",
      "    HOH ('W', 2145, ' ')\n",
      "    HOH ('W', 2146, ' ')\n",
      "    HOH ('W', 2147, ' ')\n",
      "    HOH ('W', 2148, ' ')\n",
      "    HOH ('W', 2149, ' ')\n",
      "    HOH ('W', 2150, ' ')\n",
      "    HOH ('W', 2151, ' ')\n",
      "    HOH ('W', 2152, ' ')\n",
      "    HOH ('W', 2153, ' ')\n",
      "    HOH ('W', 2154, ' ')\n",
      "    HOH ('W', 2155, ' ')\n",
      "    HOH ('W', 2156, ' ')\n",
      "    HOH ('W', 2157, ' ')\n",
      "    HOH ('W', 2158, ' ')\n",
      "    HOH ('W', 2159, ' ')\n",
      "    HOH ('W', 2160, ' ')\n",
      "    HOH ('W', 2161, ' ')\n",
      "    HOH ('W', 2162, ' ')\n",
      "    HOH ('W', 2163, ' ')\n",
      "    HOH ('W', 2164, ' ')\n",
      "    HOH ('W', 2165, ' ')\n",
      "    HOH ('W', 2166, ' ')\n",
      "    HOH ('W', 2167, ' ')\n",
      "    HOH ('W', 2168, ' ')\n",
      "    HOH ('W', 2169, ' ')\n",
      "    HOH ('W', 2170, ' ')\n",
      "    HOH ('W', 2171, ' ')\n",
      "    HOH ('W', 2172, ' ')\n",
      "    HOH ('W', 2173, ' ')\n",
      "    HOH ('W', 2174, ' ')\n",
      "    HOH ('W', 2175, ' ')\n",
      "    HOH ('W', 2176, ' ')\n",
      "    HOH ('W', 2177, ' ')\n",
      "    HOH ('W', 2178, ' ')\n",
      "    HOH ('W', 2179, ' ')\n",
      "    HOH ('W', 2180, ' ')\n",
      "    HOH ('W', 2181, ' ')\n",
      "    HOH ('W', 2182, ' ')\n",
      "    HOH ('W', 2183, ' ')\n",
      "    HOH ('W', 2184, ' ')\n",
      "    HOH ('W', 2185, ' ')\n",
      "    HOH ('W', 2186, ' ')\n",
      "    HOH ('W', 2187, ' ')\n",
      "    HOH ('W', 2188, ' ')\n",
      "    HOH ('W', 2189, ' ')\n",
      "    HOH ('W', 2190, ' ')\n",
      "    HOH ('W', 2191, ' ')\n",
      "    HOH ('W', 2192, ' ')\n",
      "    HOH ('W', 2193, ' ')\n",
      "    HOH ('W', 2194, ' ')\n",
      "    HOH ('W', 2195, ' ')\n",
      "    HOH ('W', 2196, ' ')\n",
      "    HOH ('W', 2197, ' ')\n",
      "    HOH ('W', 2198, ' ')\n",
      "    HOH ('W', 2199, ' ')\n",
      "    HOH ('W', 2200, ' ')\n",
      "    HOH ('W', 2201, ' ')\n",
      "    HOH ('W', 2202, ' ')\n",
      "    HOH ('W', 2203, ' ')\n",
      "    HOH ('W', 2204, ' ')\n",
      "    HOH ('W', 2205, ' ')\n",
      "    HOH ('W', 2206, ' ')\n",
      "    HOH ('W', 2207, ' ')\n",
      "    HOH ('W', 2208, ' ')\n",
      "    HOH ('W', 2209, ' ')\n",
      "    HOH ('W', 2210, ' ')\n",
      "    HOH ('W', 2211, ' ')\n",
      "    HOH ('W', 2212, ' ')\n",
      "    HOH ('W', 2213, ' ')\n",
      "    HOH ('W', 2214, ' ')\n",
      "    HOH ('W', 2215, ' ')\n",
      "    HOH ('W', 2216, ' ')\n",
      "    HOH ('W', 2217, ' ')\n",
      "    HOH ('W', 2218, ' ')\n",
      "    HOH ('W', 2219, ' ')\n",
      "    HOH ('W', 2220, ' ')\n",
      "    HOH ('W', 2221, ' ')\n",
      "    HOH ('W', 2222, ' ')\n",
      "    HOH ('W', 2223, ' ')\n",
      "    HOH ('W', 2224, ' ')\n",
      "    HOH ('W', 2225, ' ')\n",
      "    HOH ('W', 2226, ' ')\n",
      "    HOH ('W', 2227, ' ')\n",
      "    HOH ('W', 2228, ' ')\n",
      "    HOH ('W', 2229, ' ')\n",
      "    HOH ('W', 2230, ' ')\n",
      "    HOH ('W', 2231, ' ')\n",
      "    HOH ('W', 2232, ' ')\n",
      "    HOH ('W', 2233, ' ')\n",
      "    HOH ('W', 2234, ' ')\n",
      "    HOH ('W', 2235, ' ')\n",
      "    HOH ('W', 2236, ' ')\n",
      "    HOH ('W', 2237, ' ')\n",
      "    HOH ('W', 2238, ' ')\n",
      "    HOH ('W', 2239, ' ')\n",
      "    HOH ('W', 2240, ' ')\n",
      "    HOH ('W', 2241, ' ')\n",
      "    HOH ('W', 2242, ' ')\n",
      "    HOH ('W', 2243, ' ')\n",
      "    HOH ('W', 2244, ' ')\n",
      "    HOH ('W', 2245, ' ')\n",
      "    HOH ('W', 2246, ' ')\n",
      "    HOH ('W', 2247, ' ')\n",
      "    HOH ('W', 2248, ' ')\n",
      "    HOH ('W', 2249, ' ')\n",
      "    HOH ('W', 2250, ' ')\n",
      "    HOH ('W', 2251, ' ')\n",
      "    HOH ('W', 2252, ' ')\n",
      "    HOH ('W', 2253, ' ')\n",
      "    HOH ('W', 2254, ' ')\n",
      "    HOH ('W', 2255, ' ')\n",
      "    HOH ('W', 2256, ' ')\n",
      "    HOH ('W', 2257, ' ')\n",
      "    HOH ('W', 2258, ' ')\n",
      "    HOH ('W', 2259, ' ')\n",
      "    HOH ('W', 2260, ' ')\n",
      "    HOH ('W', 2261, ' ')\n",
      "    HOH ('W', 2262, ' ')\n",
      "    HOH ('W', 2263, ' ')\n",
      "    HOH ('W', 2264, ' ')\n",
      "    HOH ('W', 2265, ' ')\n",
      "    HOH ('W', 2266, ' ')\n",
      "    HOH ('W', 2267, ' ')\n",
      "    HOH ('W', 2268, ' ')\n",
      "    HOH ('W', 2269, ' ')\n",
      "    HOH ('W', 2270, ' ')\n",
      "    HOH ('W', 2271, ' ')\n",
      "    HOH ('W', 2272, ' ')\n",
      "    HOH ('W', 2273, ' ')\n",
      "    HOH ('W', 2274, ' ')\n",
      "    HOH ('W', 2275, ' ')\n",
      "    HOH ('W', 2276, ' ')\n",
      "    HOH ('W', 2277, ' ')\n",
      "    HOH ('W', 2278, ' ')\n",
      "    HOH ('W', 2279, ' ')\n",
      "    HOH ('W', 2280, ' ')\n",
      "    HOH ('W', 2281, ' ')\n",
      "    HOH ('W', 2282, ' ')\n",
      "    HOH ('W', 2283, ' ')\n",
      "    HOH ('W', 2284, ' ')\n",
      "    HOH ('W', 2285, ' ')\n",
      "    HOH ('W', 2286, ' ')\n",
      "    HOH ('W', 2287, ' ')\n",
      "    HOH ('W', 2288, ' ')\n",
      "    HOH ('W', 2289, ' ')\n",
      "    HOH ('W', 2290, ' ')\n",
      "    HOH ('W', 2291, ' ')\n",
      "    HOH ('W', 2292, ' ')\n",
      "    HOH ('W', 2293, ' ')\n",
      "    HOH ('W', 2294, ' ')\n",
      "    HOH ('W', 2295, ' ')\n",
      "    HOH ('W', 2296, ' ')\n",
      "    HOH ('W', 2297, ' ')\n",
      "    HOH ('W', 2298, ' ')\n",
      "    HOH ('W', 2299, ' ')\n",
      "    HOH ('W', 2300, ' ')\n",
      "    HOH ('W', 2301, ' ')\n",
      "    HOH ('W', 2302, ' ')\n",
      "    HOH ('W', 2303, ' ')\n",
      "    HOH ('W', 2304, ' ')\n",
      "    HOH ('W', 2305, ' ')\n",
      "    HOH ('W', 2306, ' ')\n",
      "    HOH ('W', 2307, ' ')\n",
      "    HOH ('W', 2308, ' ')\n",
      "    HOH ('W', 2309, ' ')\n",
      "    HOH ('W', 2310, ' ')\n",
      "    HOH ('W', 2311, ' ')\n",
      "    HOH ('W', 2312, ' ')\n",
      "    HOH ('W', 2313, ' ')\n",
      "    HOH ('W', 2314, ' ')\n",
      "    HOH ('W', 2315, ' ')\n",
      "    HOH ('W', 2316, ' ')\n",
      "    HOH ('W', 2317, ' ')\n",
      "    HOH ('W', 2318, ' ')\n",
      "    HOH ('W', 2319, ' ')\n",
      "    HOH ('W', 2320, ' ')\n",
      "    HOH ('W', 2321, ' ')\n",
      "    HOH ('W', 2322, ' ')\n",
      "    HOH ('W', 2323, ' ')\n",
      "    HOH ('W', 2324, ' ')\n",
      "    HOH ('W', 2325, ' ')\n",
      "    HOH ('W', 2326, ' ')\n",
      "    HOH ('W', 2327, ' ')\n",
      "    HOH ('W', 2328, ' ')\n",
      "    HOH ('W', 2329, ' ')\n",
      "    HOH ('W', 2330, ' ')\n",
      "    HOH ('W', 2331, ' ')\n",
      "    HOH ('W', 2332, ' ')\n",
      "    HOH ('W', 2333, ' ')\n",
      "    HOH ('W', 2334, ' ')\n",
      "    HOH ('W', 2335, ' ')\n",
      "    HOH ('W', 2336, ' ')\n",
      "    HOH ('W', 2337, ' ')\n",
      "    HOH ('W', 2338, ' ')\n",
      "    HOH ('W', 2339, ' ')\n",
      "    HOH ('W', 2340, ' ')\n",
      "    HOH ('W', 2341, ' ')\n",
      "    HOH ('W', 2342, ' ')\n",
      "    HOH ('W', 2343, ' ')\n",
      "    HOH ('W', 2344, ' ')\n",
      "    HOH ('W', 2345, ' ')\n",
      "    HOH ('W', 2346, ' ')\n",
      "    HOH ('W', 2347, ' ')\n",
      "    HOH ('W', 2348, ' ')\n",
      "    HOH ('W', 2349, ' ')\n",
      "    HOH ('W', 2350, ' ')\n",
      "    HOH ('W', 2351, ' ')\n",
      "    HOH ('W', 2352, ' ')\n",
      "    HOH ('W', 2353, ' ')\n",
      "    HOH ('W', 2354, ' ')\n",
      "    HOH ('W', 2355, ' ')\n",
      "    HOH ('W', 2356, ' ')\n",
      "    HOH ('W', 2357, ' ')\n",
      "    HOH ('W', 2358, ' ')\n",
      "    HOH ('W', 2359, ' ')\n",
      "    HOH ('W', 2360, ' ')\n",
      "    HOH ('W', 2361, ' ')\n",
      "    HOH ('W', 2362, ' ')\n",
      "    HOH ('W', 2363, ' ')\n",
      "    HOH ('W', 2364, ' ')\n",
      "    HOH ('W', 2365, ' ')\n",
      "    HOH ('W', 2366, ' ')\n",
      "    HOH ('W', 2367, ' ')\n",
      "    HOH ('W', 2368, ' ')\n",
      "    HOH ('W', 2369, ' ')\n",
      "    HOH ('W', 2370, ' ')\n",
      "    HOH ('W', 2371, ' ')\n",
      "    HOH ('W', 2372, ' ')\n",
      "    HOH ('W', 2373, ' ')\n",
      "    HOH ('W', 2374, ' ')\n",
      "    HOH ('W', 2375, ' ')\n",
      "    HOH ('W', 2376, ' ')\n",
      "    HOH ('W', 2377, ' ')\n",
      "    HOH ('W', 2378, ' ')\n",
      "    HOH ('W', 2379, ' ')\n",
      "    HOH ('W', 2380, ' ')\n",
      "    HOH ('W', 2381, ' ')\n",
      "    HOH ('W', 2382, ' ')\n",
      "    HOH ('W', 2383, ' ')\n",
      "    HOH ('W', 2384, ' ')\n",
      "    HOH ('W', 2385, ' ')\n",
      "    HOH ('W', 2386, ' ')\n",
      "    HOH ('W', 2387, ' ')\n",
      "    HOH ('W', 2388, ' ')\n",
      "    HOH ('W', 2389, ' ')\n",
      "    HOH ('W', 2390, ' ')\n",
      "    HOH ('W', 2391, ' ')\n",
      "    HOH ('W', 2392, ' ')\n",
      "  Chain AC:\n",
      "    No hetero residues found\n",
      "  Chain AD:\n",
      "    No hetero residues found\n",
      "  Chain AE:\n",
      "    HOH ('W', 201, ' ')\n",
      "  Chain AF:\n",
      "    No hetero residues found\n",
      "  Chain AG:\n",
      "    No hetero residues found\n",
      "  Chain AH:\n",
      "    No hetero residues found\n",
      "  Chain AI:\n",
      "    No hetero residues found\n",
      "  Chain AJ:\n",
      "    No hetero residues found\n",
      "  Chain AK:\n",
      "    HOH ('W', 201, ' ')\n",
      "  Chain AL:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "  Chain AM:\n",
      "    No hetero residues found\n",
      "  Chain AN:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "  Chain AO:\n",
      "    No hetero residues found\n",
      "  Chain AP:\n",
      "    No hetero residues found\n",
      "  Chain AQ:\n",
      "    No hetero residues found\n",
      "  Chain AR:\n",
      "    No hetero residues found\n",
      "  Chain AS:\n",
      "    No hetero residues found\n",
      "  Chain AT:\n",
      "    HOH ('W', 101, ' ')\n",
      "    HOH ('W', 102, ' ')\n",
      "  Chain AB:\n",
      "    No hetero residues found\n",
      "  Chain AU:\n",
      "    No hetero residues found\n",
      "  Chain BA:\n",
      "    No hetero residues found\n",
      "  Chain BB:\n",
      "    MG ('H_MG', 3001, ' ')\n",
      "    MG ('H_MG', 3002, ' ')\n",
      "    MG ('H_MG', 3003, ' ')\n",
      "    MG ('H_MG', 3004, ' ')\n",
      "    MG ('H_MG', 3005, ' ')\n",
      "    MG ('H_MG', 3006, ' ')\n",
      "    MG ('H_MG', 3007, ' ')\n",
      "    MG ('H_MG', 3008, ' ')\n",
      "    MG ('H_MG', 3009, ' ')\n",
      "    MG ('H_MG', 3010, ' ')\n",
      "    MG ('H_MG', 3011, ' ')\n",
      "    MG ('H_MG', 3012, ' ')\n",
      "    MG ('H_MG', 3013, ' ')\n",
      "    MG ('H_MG', 3014, ' ')\n",
      "    MG ('H_MG', 3015, ' ')\n",
      "    MG ('H_MG', 3016, ' ')\n",
      "    MG ('H_MG', 3017, ' ')\n",
      "    MG ('H_MG', 3018, ' ')\n",
      "    MG ('H_MG', 3019, ' ')\n",
      "    MG ('H_MG', 3020, ' ')\n",
      "    MG ('H_MG', 3021, ' ')\n",
      "    MG ('H_MG', 3022, ' ')\n",
      "    MG ('H_MG', 3023, ' ')\n",
      "    MG ('H_MG', 3024, ' ')\n",
      "    MG ('H_MG', 3025, ' ')\n",
      "    MG ('H_MG', 3026, ' ')\n",
      "    MG ('H_MG', 3027, ' ')\n",
      "    MG ('H_MG', 3028, ' ')\n",
      "    MG ('H_MG', 3029, ' ')\n",
      "    MG ('H_MG', 3030, ' ')\n",
      "    MG ('H_MG', 3031, ' ')\n",
      "    MG ('H_MG', 3032, ' ')\n",
      "    MG ('H_MG', 3033, ' ')\n",
      "    MG ('H_MG', 3034, ' ')\n",
      "    MG ('H_MG', 3035, ' ')\n",
      "    MG ('H_MG', 3036, ' ')\n",
      "    MG ('H_MG', 3037, ' ')\n",
      "    MG ('H_MG', 3038, ' ')\n",
      "    MG ('H_MG', 3039, ' ')\n",
      "    MG ('H_MG', 3040, ' ')\n",
      "    MG ('H_MG', 3041, ' ')\n",
      "    MG ('H_MG', 3042, ' ')\n",
      "    MG ('H_MG', 3043, ' ')\n",
      "    MG ('H_MG', 3044, ' ')\n",
      "    MG ('H_MG', 3045, ' ')\n",
      "    MG ('H_MG', 3046, ' ')\n",
      "    MG ('H_MG', 3047, ' ')\n",
      "    MG ('H_MG', 3048, ' ')\n",
      "    MG ('H_MG', 3049, ' ')\n",
      "    MG ('H_MG', 3050, ' ')\n",
      "    MG ('H_MG', 3051, ' ')\n",
      "    MG ('H_MG', 3052, ' ')\n",
      "    MG ('H_MG', 3053, ' ')\n",
      "    MG ('H_MG', 3054, ' ')\n",
      "    MG ('H_MG', 3055, ' ')\n",
      "    MG ('H_MG', 3056, ' ')\n",
      "    MG ('H_MG', 3057, ' ')\n",
      "    MG ('H_MG', 3058, ' ')\n",
      "    MG ('H_MG', 3059, ' ')\n",
      "    MG ('H_MG', 3060, ' ')\n",
      "    MG ('H_MG', 3061, ' ')\n",
      "    MG ('H_MG', 3062, ' ')\n",
      "    MG ('H_MG', 3063, ' ')\n",
      "    MG ('H_MG', 3064, ' ')\n",
      "    MG ('H_MG', 3065, ' ')\n",
      "    MG ('H_MG', 3066, ' ')\n",
      "    MG ('H_MG', 3067, ' ')\n",
      "    MG ('H_MG', 3068, ' ')\n",
      "    MG ('H_MG', 3069, ' ')\n",
      "    MG ('H_MG', 3070, ' ')\n",
      "    MG ('H_MG', 3071, ' ')\n",
      "    MG ('H_MG', 3072, ' ')\n",
      "    MG ('H_MG', 3073, ' ')\n",
      "    MG ('H_MG', 3074, ' ')\n",
      "    MG ('H_MG', 3075, ' ')\n",
      "    MG ('H_MG', 3076, ' ')\n",
      "    MG ('H_MG', 3077, ' ')\n",
      "    MG ('H_MG', 3078, ' ')\n",
      "    MG ('H_MG', 3079, ' ')\n",
      "    MG ('H_MG', 3080, ' ')\n",
      "    MG ('H_MG', 3081, ' ')\n",
      "    MG ('H_MG', 3082, ' ')\n",
      "    MG ('H_MG', 3083, ' ')\n",
      "    MG ('H_MG', 3084, ' ')\n",
      "    MG ('H_MG', 3085, ' ')\n",
      "    MG ('H_MG', 3086, ' ')\n",
      "    MG ('H_MG', 3087, ' ')\n",
      "    MG ('H_MG', 3088, ' ')\n",
      "    MG ('H_MG', 3089, ' ')\n",
      "    MG ('H_MG', 3090, ' ')\n",
      "    MG ('H_MG', 3091, ' ')\n",
      "    MG ('H_MG', 3092, ' ')\n",
      "    MG ('H_MG', 3093, ' ')\n",
      "    MG ('H_MG', 3094, ' ')\n",
      "    MG ('H_MG', 3095, ' ')\n",
      "    MG ('H_MG', 3096, ' ')\n",
      "    MG ('H_MG', 3097, ' ')\n",
      "    MG ('H_MG', 3098, ' ')\n",
      "    MG ('H_MG', 3099, ' ')\n",
      "    MG ('H_MG', 3100, ' ')\n",
      "    MG ('H_MG', 3101, ' ')\n",
      "    MG ('H_MG', 3102, ' ')\n",
      "    MG ('H_MG', 3103, ' ')\n",
      "    MG ('H_MG', 3104, ' ')\n",
      "    MG ('H_MG', 3105, ' ')\n",
      "    MG ('H_MG', 3106, ' ')\n",
      "    MG ('H_MG', 3107, ' ')\n",
      "    MG ('H_MG', 3108, ' ')\n",
      "    MG ('H_MG', 3109, ' ')\n",
      "    MG ('H_MG', 3110, ' ')\n",
      "    LLL ('H_LLL', 3111, ' ')\n",
      "    HOH ('W', 3201, ' ')\n",
      "    HOH ('W', 3202, ' ')\n",
      "    HOH ('W', 3203, ' ')\n",
      "    HOH ('W', 3204, ' ')\n",
      "    HOH ('W', 3205, ' ')\n",
      "    HOH ('W', 3206, ' ')\n",
      "    HOH ('W', 3207, ' ')\n",
      "    HOH ('W', 3208, ' ')\n",
      "    HOH ('W', 3209, ' ')\n",
      "    HOH ('W', 3210, ' ')\n",
      "    HOH ('W', 3211, ' ')\n",
      "    HOH ('W', 3212, ' ')\n",
      "    HOH ('W', 3213, ' ')\n",
      "    HOH ('W', 3214, ' ')\n",
      "    HOH ('W', 3215, ' ')\n",
      "    HOH ('W', 3216, ' ')\n",
      "    HOH ('W', 3217, ' ')\n",
      "    HOH ('W', 3218, ' ')\n",
      "    HOH ('W', 3219, ' ')\n",
      "    HOH ('W', 3220, ' ')\n",
      "    HOH ('W', 3221, ' ')\n",
      "    HOH ('W', 3222, ' ')\n",
      "    HOH ('W', 3223, ' ')\n",
      "    HOH ('W', 3224, ' ')\n",
      "    HOH ('W', 3225, ' ')\n",
      "    HOH ('W', 3226, ' ')\n",
      "    HOH ('W', 3227, ' ')\n",
      "    HOH ('W', 3228, ' ')\n",
      "    HOH ('W', 3229, ' ')\n",
      "    HOH ('W', 3230, ' ')\n",
      "    HOH ('W', 3231, ' ')\n",
      "    HOH ('W', 3232, ' ')\n",
      "    HOH ('W', 3233, ' ')\n",
      "    HOH ('W', 3234, ' ')\n",
      "    HOH ('W', 3235, ' ')\n",
      "    HOH ('W', 3236, ' ')\n",
      "    HOH ('W', 3237, ' ')\n",
      "    HOH ('W', 3238, ' ')\n",
      "    HOH ('W', 3239, ' ')\n",
      "    HOH ('W', 3240, ' ')\n",
      "    HOH ('W', 3241, ' ')\n",
      "    HOH ('W', 3242, ' ')\n",
      "    HOH ('W', 3243, ' ')\n",
      "    HOH ('W', 3244, ' ')\n",
      "    HOH ('W', 3245, ' ')\n",
      "    HOH ('W', 3246, ' ')\n",
      "    HOH ('W', 3247, ' ')\n",
      "    HOH ('W', 3248, ' ')\n",
      "    HOH ('W', 3249, ' ')\n",
      "    HOH ('W', 3250, ' ')\n",
      "    HOH ('W', 3251, ' ')\n",
      "    HOH ('W', 3252, ' ')\n",
      "    HOH ('W', 3253, ' ')\n",
      "    HOH ('W', 3254, ' ')\n",
      "    HOH ('W', 3255, ' ')\n",
      "    HOH ('W', 3256, ' ')\n",
      "    HOH ('W', 3257, ' ')\n",
      "    HOH ('W', 3258, ' ')\n",
      "    HOH ('W', 3259, ' ')\n",
      "    HOH ('W', 3260, ' ')\n",
      "    HOH ('W', 3261, ' ')\n",
      "    HOH ('W', 3262, ' ')\n",
      "    HOH ('W', 3263, ' ')\n",
      "    HOH ('W', 3264, ' ')\n",
      "    HOH ('W', 3265, ' ')\n",
      "    HOH ('W', 3266, ' ')\n",
      "    HOH ('W', 3267, ' ')\n",
      "    HOH ('W', 3268, ' ')\n",
      "    HOH ('W', 3269, ' ')\n",
      "    HOH ('W', 3270, ' ')\n",
      "    HOH ('W', 3271, ' ')\n",
      "    HOH ('W', 3272, ' ')\n",
      "    HOH ('W', 3273, ' ')\n",
      "    HOH ('W', 3274, ' ')\n",
      "    HOH ('W', 3275, ' ')\n",
      "    HOH ('W', 3276, ' ')\n",
      "    HOH ('W', 3277, ' ')\n",
      "    HOH ('W', 3278, ' ')\n",
      "    HOH ('W', 3279, ' ')\n",
      "    HOH ('W', 3280, ' ')\n",
      "    HOH ('W', 3281, ' ')\n",
      "    HOH ('W', 3282, ' ')\n",
      "    HOH ('W', 3283, ' ')\n",
      "    HOH ('W', 3284, ' ')\n",
      "    HOH ('W', 3285, ' ')\n",
      "    HOH ('W', 3286, ' ')\n",
      "    HOH ('W', 3287, ' ')\n",
      "    HOH ('W', 3288, ' ')\n",
      "    HOH ('W', 3289, ' ')\n",
      "    HOH ('W', 3290, ' ')\n",
      "    HOH ('W', 3291, ' ')\n",
      "    HOH ('W', 3292, ' ')\n",
      "    HOH ('W', 3293, ' ')\n",
      "    HOH ('W', 3294, ' ')\n",
      "    HOH ('W', 3295, ' ')\n",
      "    HOH ('W', 3296, ' ')\n",
      "    HOH ('W', 3297, ' ')\n",
      "    HOH ('W', 3298, ' ')\n",
      "    HOH ('W', 3299, ' ')\n",
      "    HOH ('W', 3300, ' ')\n",
      "    HOH ('W', 3301, ' ')\n",
      "    HOH ('W', 3302, ' ')\n",
      "    HOH ('W', 3303, ' ')\n",
      "    HOH ('W', 3304, ' ')\n",
      "    HOH ('W', 3305, ' ')\n",
      "    HOH ('W', 3306, ' ')\n",
      "    HOH ('W', 3307, ' ')\n",
      "    HOH ('W', 3308, ' ')\n",
      "    HOH ('W', 3309, ' ')\n",
      "    HOH ('W', 3310, ' ')\n",
      "    HOH ('W', 3311, ' ')\n",
      "    HOH ('W', 3312, ' ')\n",
      "    HOH ('W', 3313, ' ')\n",
      "    HOH ('W', 3314, ' ')\n",
      "    HOH ('W', 3315, ' ')\n",
      "    HOH ('W', 3316, ' ')\n",
      "    HOH ('W', 3317, ' ')\n",
      "    HOH ('W', 3318, ' ')\n",
      "    HOH ('W', 3319, ' ')\n",
      "    HOH ('W', 3320, ' ')\n",
      "    HOH ('W', 3321, ' ')\n",
      "    HOH ('W', 3322, ' ')\n",
      "    HOH ('W', 3323, ' ')\n",
      "    HOH ('W', 3324, ' ')\n",
      "    HOH ('W', 3325, ' ')\n",
      "    HOH ('W', 3326, ' ')\n",
      "    HOH ('W', 3327, ' ')\n",
      "    HOH ('W', 3328, ' ')\n",
      "    HOH ('W', 3329, ' ')\n",
      "    HOH ('W', 3330, ' ')\n",
      "    HOH ('W', 3331, ' ')\n",
      "    HOH ('W', 3332, ' ')\n",
      "    HOH ('W', 3333, ' ')\n",
      "    HOH ('W', 3334, ' ')\n",
      "    HOH ('W', 3335, ' ')\n",
      "    HOH ('W', 3336, ' ')\n",
      "    HOH ('W', 3337, ' ')\n",
      "    HOH ('W', 3338, ' ')\n",
      "    HOH ('W', 3339, ' ')\n",
      "    HOH ('W', 3340, ' ')\n",
      "    HOH ('W', 3341, ' ')\n",
      "    HOH ('W', 3342, ' ')\n",
      "    HOH ('W', 3343, ' ')\n",
      "    HOH ('W', 3344, ' ')\n",
      "    HOH ('W', 3345, ' ')\n",
      "    HOH ('W', 3346, ' ')\n",
      "    HOH ('W', 3347, ' ')\n",
      "    HOH ('W', 3348, ' ')\n",
      "    HOH ('W', 3349, ' ')\n",
      "    HOH ('W', 3350, ' ')\n",
      "    HOH ('W', 3351, ' ')\n",
      "    HOH ('W', 3352, ' ')\n",
      "    HOH ('W', 3353, ' ')\n",
      "    HOH ('W', 3354, ' ')\n",
      "    HOH ('W', 3355, ' ')\n",
      "    HOH ('W', 3356, ' ')\n",
      "    HOH ('W', 3357, ' ')\n",
      "    HOH ('W', 3358, ' ')\n",
      "    HOH ('W', 3359, ' ')\n",
      "    HOH ('W', 3360, ' ')\n",
      "    HOH ('W', 3361, ' ')\n",
      "    HOH ('W', 3362, ' ')\n",
      "    HOH ('W', 3363, ' ')\n",
      "    HOH ('W', 3364, ' ')\n",
      "    HOH ('W', 3365, ' ')\n",
      "    HOH ('W', 3366, ' ')\n",
      "    HOH ('W', 3367, ' ')\n",
      "    HOH ('W', 3368, ' ')\n",
      "    HOH ('W', 3369, ' ')\n",
      "    HOH ('W', 3370, ' ')\n",
      "    HOH ('W', 3371, ' ')\n",
      "    HOH ('W', 3372, ' ')\n",
      "    HOH ('W', 3373, ' ')\n",
      "    HOH ('W', 3374, ' ')\n",
      "    HOH ('W', 3375, ' ')\n",
      "    HOH ('W', 3376, ' ')\n",
      "    HOH ('W', 3377, ' ')\n",
      "    HOH ('W', 3378, ' ')\n",
      "    HOH ('W', 3379, ' ')\n",
      "    HOH ('W', 3380, ' ')\n",
      "    HOH ('W', 3381, ' ')\n",
      "    HOH ('W', 3382, ' ')\n",
      "    HOH ('W', 3383, ' ')\n",
      "    HOH ('W', 3384, ' ')\n",
      "    HOH ('W', 3385, ' ')\n",
      "    HOH ('W', 3386, ' ')\n",
      "    HOH ('W', 3387, ' ')\n",
      "    HOH ('W', 3388, ' ')\n",
      "    HOH ('W', 3389, ' ')\n",
      "    HOH ('W', 3390, ' ')\n",
      "    HOH ('W', 3391, ' ')\n",
      "    HOH ('W', 3392, ' ')\n",
      "    HOH ('W', 3393, ' ')\n",
      "    HOH ('W', 3394, ' ')\n",
      "    HOH ('W', 3395, ' ')\n",
      "    HOH ('W', 3396, ' ')\n",
      "    HOH ('W', 3397, ' ')\n",
      "    HOH ('W', 3398, ' ')\n",
      "    HOH ('W', 3399, ' ')\n",
      "    HOH ('W', 3400, ' ')\n",
      "    HOH ('W', 3401, ' ')\n",
      "    HOH ('W', 3402, ' ')\n",
      "    HOH ('W', 3403, ' ')\n",
      "    HOH ('W', 3404, ' ')\n",
      "    HOH ('W', 3405, ' ')\n",
      "    HOH ('W', 3406, ' ')\n",
      "    HOH ('W', 3407, ' ')\n",
      "    HOH ('W', 3408, ' ')\n",
      "    HOH ('W', 3409, ' ')\n",
      "    HOH ('W', 3410, ' ')\n",
      "    HOH ('W', 3411, ' ')\n",
      "    HOH ('W', 3412, ' ')\n",
      "    HOH ('W', 3413, ' ')\n",
      "    HOH ('W', 3414, ' ')\n",
      "    HOH ('W', 3415, ' ')\n",
      "    HOH ('W', 3416, ' ')\n",
      "    HOH ('W', 3417, ' ')\n",
      "    HOH ('W', 3418, ' ')\n",
      "    HOH ('W', 3419, ' ')\n",
      "    HOH ('W', 3420, ' ')\n",
      "    HOH ('W', 3421, ' ')\n",
      "    HOH ('W', 3422, ' ')\n",
      "    HOH ('W', 3423, ' ')\n",
      "    HOH ('W', 3424, ' ')\n",
      "    HOH ('W', 3425, ' ')\n",
      "    HOH ('W', 3426, ' ')\n",
      "    HOH ('W', 3427, ' ')\n",
      "    HOH ('W', 3428, ' ')\n",
      "    HOH ('W', 3429, ' ')\n",
      "    HOH ('W', 3430, ' ')\n",
      "    HOH ('W', 3431, ' ')\n",
      "    HOH ('W', 3432, ' ')\n",
      "    HOH ('W', 3433, ' ')\n",
      "    HOH ('W', 3434, ' ')\n",
      "    HOH ('W', 3435, ' ')\n",
      "    HOH ('W', 3436, ' ')\n",
      "    HOH ('W', 3437, ' ')\n",
      "    HOH ('W', 3438, ' ')\n",
      "    HOH ('W', 3439, ' ')\n",
      "    HOH ('W', 3440, ' ')\n",
      "    HOH ('W', 3441, ' ')\n",
      "    HOH ('W', 3442, ' ')\n",
      "    HOH ('W', 3443, ' ')\n",
      "    HOH ('W', 3444, ' ')\n",
      "    HOH ('W', 3445, ' ')\n",
      "    HOH ('W', 3446, ' ')\n",
      "    HOH ('W', 3447, ' ')\n",
      "    HOH ('W', 3448, ' ')\n",
      "    HOH ('W', 3449, ' ')\n",
      "    HOH ('W', 3450, ' ')\n",
      "    HOH ('W', 3451, ' ')\n",
      "    HOH ('W', 3452, ' ')\n",
      "    HOH ('W', 3453, ' ')\n",
      "    HOH ('W', 3454, ' ')\n",
      "    HOH ('W', 3455, ' ')\n",
      "    HOH ('W', 3456, ' ')\n",
      "    HOH ('W', 3457, ' ')\n",
      "    HOH ('W', 3458, ' ')\n",
      "    HOH ('W', 3459, ' ')\n",
      "    HOH ('W', 3460, ' ')\n",
      "    HOH ('W', 3461, ' ')\n",
      "    HOH ('W', 3462, ' ')\n",
      "    HOH ('W', 3463, ' ')\n",
      "    HOH ('W', 3464, ' ')\n",
      "    HOH ('W', 3465, ' ')\n",
      "    HOH ('W', 3466, ' ')\n",
      "    HOH ('W', 3467, ' ')\n",
      "    HOH ('W', 3468, ' ')\n",
      "    HOH ('W', 3469, ' ')\n",
      "    HOH ('W', 3470, ' ')\n",
      "    HOH ('W', 3471, ' ')\n",
      "    HOH ('W', 3472, ' ')\n",
      "    HOH ('W', 3473, ' ')\n",
      "    HOH ('W', 3474, ' ')\n",
      "    HOH ('W', 3475, ' ')\n",
      "    HOH ('W', 3476, ' ')\n",
      "    HOH ('W', 3477, ' ')\n",
      "    HOH ('W', 3478, ' ')\n",
      "    HOH ('W', 3479, ' ')\n",
      "    HOH ('W', 3480, ' ')\n",
      "    HOH ('W', 3481, ' ')\n",
      "    HOH ('W', 3482, ' ')\n",
      "    HOH ('W', 3483, ' ')\n",
      "    HOH ('W', 3484, ' ')\n",
      "    HOH ('W', 3485, ' ')\n",
      "    HOH ('W', 3486, ' ')\n",
      "    HOH ('W', 3487, ' ')\n",
      "    HOH ('W', 3488, ' ')\n",
      "    HOH ('W', 3489, ' ')\n",
      "    HOH ('W', 3490, ' ')\n",
      "    HOH ('W', 3491, ' ')\n",
      "    HOH ('W', 3492, ' ')\n",
      "    HOH ('W', 3493, ' ')\n",
      "    HOH ('W', 3494, ' ')\n",
      "    HOH ('W', 3495, ' ')\n",
      "    HOH ('W', 3496, ' ')\n",
      "    HOH ('W', 3497, ' ')\n",
      "    HOH ('W', 3498, ' ')\n",
      "    HOH ('W', 3499, ' ')\n",
      "    HOH ('W', 3500, ' ')\n",
      "    HOH ('W', 3501, ' ')\n",
      "    HOH ('W', 3502, ' ')\n",
      "    HOH ('W', 3503, ' ')\n",
      "    HOH ('W', 3504, ' ')\n",
      "    HOH ('W', 3505, ' ')\n",
      "    HOH ('W', 3506, ' ')\n",
      "    HOH ('W', 3507, ' ')\n",
      "    HOH ('W', 3508, ' ')\n",
      "    HOH ('W', 3509, ' ')\n",
      "    HOH ('W', 3510, ' ')\n",
      "    HOH ('W', 3511, ' ')\n",
      "    HOH ('W', 3512, ' ')\n",
      "    HOH ('W', 3513, ' ')\n",
      "    HOH ('W', 3514, ' ')\n",
      "    HOH ('W', 3515, ' ')\n",
      "    HOH ('W', 3516, ' ')\n",
      "    HOH ('W', 3517, ' ')\n",
      "    HOH ('W', 3518, ' ')\n",
      "    HOH ('W', 3519, ' ')\n",
      "    HOH ('W', 3520, ' ')\n",
      "    HOH ('W', 3521, ' ')\n",
      "    HOH ('W', 3522, ' ')\n",
      "    HOH ('W', 3523, ' ')\n",
      "    HOH ('W', 3524, ' ')\n",
      "    HOH ('W', 3525, ' ')\n",
      "    HOH ('W', 3526, ' ')\n",
      "    HOH ('W', 3527, ' ')\n",
      "    HOH ('W', 3528, ' ')\n",
      "    HOH ('W', 3529, ' ')\n",
      "    HOH ('W', 3530, ' ')\n",
      "    HOH ('W', 3531, ' ')\n",
      "    HOH ('W', 3532, ' ')\n",
      "    HOH ('W', 3533, ' ')\n",
      "    HOH ('W', 3534, ' ')\n",
      "    HOH ('W', 3535, ' ')\n",
      "    HOH ('W', 3536, ' ')\n",
      "    HOH ('W', 3537, ' ')\n",
      "    HOH ('W', 3538, ' ')\n",
      "    HOH ('W', 3539, ' ')\n",
      "    HOH ('W', 3540, ' ')\n",
      "    HOH ('W', 3541, ' ')\n",
      "    HOH ('W', 3542, ' ')\n",
      "    HOH ('W', 3543, ' ')\n",
      "    HOH ('W', 3544, ' ')\n",
      "    HOH ('W', 3545, ' ')\n",
      "    HOH ('W', 3546, ' ')\n",
      "    HOH ('W', 3547, ' ')\n",
      "    HOH ('W', 3548, ' ')\n",
      "    HOH ('W', 3549, ' ')\n",
      "    HOH ('W', 3550, ' ')\n",
      "    HOH ('W', 3551, ' ')\n",
      "    HOH ('W', 3552, ' ')\n",
      "    HOH ('W', 3553, ' ')\n",
      "    HOH ('W', 3554, ' ')\n",
      "    HOH ('W', 3555, ' ')\n",
      "    HOH ('W', 3556, ' ')\n",
      "    HOH ('W', 3557, ' ')\n",
      "    HOH ('W', 3558, ' ')\n",
      "    HOH ('W', 3559, ' ')\n",
      "    HOH ('W', 3560, ' ')\n",
      "    HOH ('W', 3561, ' ')\n",
      "    HOH ('W', 3562, ' ')\n",
      "    HOH ('W', 3563, ' ')\n",
      "    HOH ('W', 3564, ' ')\n",
      "    HOH ('W', 3565, ' ')\n",
      "    HOH ('W', 3566, ' ')\n",
      "    HOH ('W', 3567, ' ')\n",
      "    HOH ('W', 3568, ' ')\n",
      "    HOH ('W', 3569, ' ')\n",
      "    HOH ('W', 3570, ' ')\n",
      "    HOH ('W', 3571, ' ')\n",
      "    HOH ('W', 3572, ' ')\n",
      "    HOH ('W', 3573, ' ')\n",
      "    HOH ('W', 3574, ' ')\n",
      "    HOH ('W', 3575, ' ')\n",
      "    HOH ('W', 3576, ' ')\n",
      "    HOH ('W', 3577, ' ')\n",
      "    HOH ('W', 3578, ' ')\n",
      "    HOH ('W', 3579, ' ')\n",
      "    HOH ('W', 3580, ' ')\n",
      "    HOH ('W', 3581, ' ')\n",
      "    HOH ('W', 3582, ' ')\n",
      "    HOH ('W', 3583, ' ')\n",
      "    HOH ('W', 3584, ' ')\n",
      "    HOH ('W', 3585, ' ')\n",
      "    HOH ('W', 3586, ' ')\n",
      "    HOH ('W', 3587, ' ')\n",
      "    HOH ('W', 3588, ' ')\n",
      "    HOH ('W', 3589, ' ')\n",
      "    HOH ('W', 3590, ' ')\n",
      "    HOH ('W', 3591, ' ')\n",
      "    HOH ('W', 3592, ' ')\n",
      "    HOH ('W', 3593, ' ')\n",
      "    HOH ('W', 3594, ' ')\n",
      "    HOH ('W', 3595, ' ')\n",
      "    HOH ('W', 3596, ' ')\n",
      "    HOH ('W', 3597, ' ')\n",
      "    HOH ('W', 3598, ' ')\n",
      "    HOH ('W', 3599, ' ')\n",
      "    HOH ('W', 3600, ' ')\n",
      "    HOH ('W', 3601, ' ')\n",
      "    HOH ('W', 3602, ' ')\n",
      "    HOH ('W', 3603, ' ')\n",
      "    HOH ('W', 3604, ' ')\n",
      "    HOH ('W', 3605, ' ')\n",
      "    HOH ('W', 3606, ' ')\n",
      "    HOH ('W', 3607, ' ')\n",
      "    HOH ('W', 3608, ' ')\n",
      "    HOH ('W', 3609, ' ')\n",
      "    HOH ('W', 3610, ' ')\n",
      "    HOH ('W', 3611, ' ')\n",
      "    HOH ('W', 3612, ' ')\n",
      "    HOH ('W', 3613, ' ')\n",
      "    HOH ('W', 3614, ' ')\n",
      "    HOH ('W', 3615, ' ')\n",
      "    HOH ('W', 3616, ' ')\n",
      "    HOH ('W', 3617, ' ')\n",
      "    HOH ('W', 3618, ' ')\n",
      "    HOH ('W', 3619, ' ')\n",
      "    HOH ('W', 3620, ' ')\n",
      "    HOH ('W', 3621, ' ')\n",
      "    HOH ('W', 3622, ' ')\n",
      "    HOH ('W', 3623, ' ')\n",
      "    HOH ('W', 3624, ' ')\n",
      "    HOH ('W', 3625, ' ')\n",
      "    HOH ('W', 3626, ' ')\n",
      "    HOH ('W', 3627, ' ')\n",
      "    HOH ('W', 3628, ' ')\n",
      "    HOH ('W', 3629, ' ')\n",
      "    HOH ('W', 3630, ' ')\n",
      "    HOH ('W', 3631, ' ')\n",
      "    HOH ('W', 3632, ' ')\n",
      "    HOH ('W', 3633, ' ')\n",
      "    HOH ('W', 3634, ' ')\n",
      "    HOH ('W', 3635, ' ')\n",
      "    HOH ('W', 3636, ' ')\n",
      "    HOH ('W', 3637, ' ')\n",
      "    HOH ('W', 3638, ' ')\n",
      "    HOH ('W', 3639, ' ')\n",
      "    HOH ('W', 3640, ' ')\n",
      "    HOH ('W', 3641, ' ')\n",
      "    HOH ('W', 3642, ' ')\n",
      "    HOH ('W', 3643, ' ')\n",
      "    HOH ('W', 3644, ' ')\n",
      "    HOH ('W', 3645, ' ')\n",
      "    HOH ('W', 3646, ' ')\n",
      "    HOH ('W', 3647, ' ')\n",
      "    HOH ('W', 3648, ' ')\n",
      "    HOH ('W', 3649, ' ')\n",
      "    HOH ('W', 3650, ' ')\n",
      "    HOH ('W', 3651, ' ')\n",
      "    HOH ('W', 3652, ' ')\n",
      "    HOH ('W', 3653, ' ')\n",
      "    HOH ('W', 3654, ' ')\n",
      "    HOH ('W', 3655, ' ')\n",
      "    HOH ('W', 3656, ' ')\n",
      "    HOH ('W', 3657, ' ')\n",
      "    HOH ('W', 3658, ' ')\n",
      "    HOH ('W', 3659, ' ')\n",
      "    HOH ('W', 3660, ' ')\n",
      "    HOH ('W', 3661, ' ')\n",
      "    HOH ('W', 3662, ' ')\n",
      "    HOH ('W', 3663, ' ')\n",
      "    HOH ('W', 3664, ' ')\n",
      "    HOH ('W', 3665, ' ')\n",
      "    HOH ('W', 3666, ' ')\n",
      "    HOH ('W', 3667, ' ')\n",
      "    HOH ('W', 3668, ' ')\n",
      "    HOH ('W', 3669, ' ')\n",
      "    HOH ('W', 3670, ' ')\n",
      "    HOH ('W', 3671, ' ')\n",
      "    HOH ('W', 3672, ' ')\n",
      "    HOH ('W', 3673, ' ')\n",
      "    HOH ('W', 3674, ' ')\n",
      "    HOH ('W', 3675, ' ')\n",
      "    HOH ('W', 3676, ' ')\n",
      "    HOH ('W', 3677, ' ')\n",
      "    HOH ('W', 3678, ' ')\n",
      "    HOH ('W', 3679, ' ')\n",
      "    HOH ('W', 3680, ' ')\n",
      "    HOH ('W', 3681, ' ')\n",
      "    HOH ('W', 3682, ' ')\n",
      "    HOH ('W', 3683, ' ')\n",
      "    HOH ('W', 3684, ' ')\n",
      "    HOH ('W', 3685, ' ')\n",
      "    HOH ('W', 3686, ' ')\n",
      "    HOH ('W', 3687, ' ')\n",
      "    HOH ('W', 3688, ' ')\n",
      "    HOH ('W', 3689, ' ')\n",
      "    HOH ('W', 3690, ' ')\n",
      "    HOH ('W', 3691, ' ')\n",
      "    HOH ('W', 3692, ' ')\n",
      "  Chain BI:\n",
      "    No hetero residues found\n",
      "  Chain BC:\n",
      "    HOH ('W', 301, ' ')\n",
      "    HOH ('W', 302, ' ')\n",
      "    HOH ('W', 303, ' ')\n",
      "    HOH ('W', 304, ' ')\n",
      "    HOH ('W', 305, ' ')\n",
      "    HOH ('W', 306, ' ')\n",
      "    HOH ('W', 307, ' ')\n",
      "  Chain BD:\n",
      "    No hetero residues found\n",
      "  Chain BK:\n",
      "    No hetero residues found\n",
      "  Chain BP:\n",
      "    No hetero residues found\n",
      "  Chain BE:\n",
      "    HOH ('W', 301, ' ')\n",
      "    HOH ('W', 302, ' ')\n",
      "    HOH ('W', 303, ' ')\n",
      "  Chain BY:\n",
      "    No hetero residues found\n",
      "  Chain B0:\n",
      "    No hetero residues found\n",
      "  Chain B4:\n",
      "    ZN ('H_ZN', 101, ' ')\n",
      "  Chain B1:\n",
      "    No hetero residues found\n",
      "  Chain B3:\n",
      "    No hetero residues found\n",
      "  Chain BV:\n",
      "    No hetero residues found\n",
      "  Chain B2:\n",
      "    HOH ('W', 101, ' ')\n",
      "  Chain BL:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "    HOH ('W', 203, ' ')\n",
      "  Chain BM:\n",
      "    No hetero residues found\n",
      "  Chain BX:\n",
      "    No hetero residues found\n",
      "  Chain BH:\n",
      "    HOH ('W', 201, ' ')\n",
      "  Chain BJ:\n",
      "    No hetero residues found\n",
      "  Chain BN:\n",
      "    No hetero residues found\n",
      "  Chain BO:\n",
      "    No hetero residues found\n",
      "  Chain BQ:\n",
      "    No hetero residues found\n",
      "  Chain BS:\n",
      "    No hetero residues found\n",
      "  Chain BU:\n",
      "    No hetero residues found\n",
      "  Chain BF:\n",
      "    No hetero residues found\n",
      "  Chain BG:\n",
      "    No hetero residues found\n",
      "  Chain BR:\n",
      "    No hetero residues found\n",
      "  Chain BT:\n",
      "    No hetero residues found\n",
      "  Chain BZ:\n",
      "    No hetero residues found\n",
      "  Chain BW:\n",
      "    No hetero residues found\n",
      "  Chain CA:\n",
      "    MG ('H_MG', 2001, ' ')\n",
      "    MG ('H_MG', 2002, ' ')\n",
      "    MG ('H_MG', 2003, ' ')\n",
      "    MG ('H_MG', 2004, ' ')\n",
      "    MG ('H_MG', 2005, ' ')\n",
      "    MG ('H_MG', 2006, ' ')\n",
      "    MG ('H_MG', 2007, ' ')\n",
      "    MG ('H_MG', 2008, ' ')\n",
      "    MG ('H_MG', 2009, ' ')\n",
      "    MG ('H_MG', 2010, ' ')\n",
      "    MG ('H_MG', 2011, ' ')\n",
      "    MG ('H_MG', 2012, ' ')\n",
      "    MG ('H_MG', 2013, ' ')\n",
      "    MG ('H_MG', 2014, ' ')\n",
      "    MG ('H_MG', 2015, ' ')\n",
      "    MG ('H_MG', 2016, ' ')\n",
      "    MG ('H_MG', 2017, ' ')\n",
      "    MG ('H_MG', 2018, ' ')\n",
      "    MG ('H_MG', 2019, ' ')\n",
      "    MG ('H_MG', 2020, ' ')\n",
      "    MG ('H_MG', 2021, ' ')\n",
      "    MG ('H_MG', 2022, ' ')\n",
      "    MG ('H_MG', 2023, ' ')\n",
      "    MG ('H_MG', 2024, ' ')\n",
      "    MG ('H_MG', 2025, ' ')\n",
      "    MG ('H_MG', 2026, ' ')\n",
      "    MG ('H_MG', 2027, ' ')\n",
      "    MG ('H_MG', 2028, ' ')\n",
      "    MG ('H_MG', 2029, ' ')\n",
      "    MG ('H_MG', 2030, ' ')\n",
      "    MG ('H_MG', 2031, ' ')\n",
      "    MG ('H_MG', 2032, ' ')\n",
      "    MG ('H_MG', 2033, ' ')\n",
      "    MG ('H_MG', 2034, ' ')\n",
      "    MG ('H_MG', 2035, ' ')\n",
      "    MG ('H_MG', 2036, ' ')\n",
      "    MG ('H_MG', 2037, ' ')\n",
      "    MG ('H_MG', 2038, ' ')\n",
      "    MG ('H_MG', 2039, ' ')\n",
      "    MG ('H_MG', 2040, ' ')\n",
      "    MG ('H_MG', 2041, ' ')\n",
      "    MG ('H_MG', 2042, ' ')\n",
      "    MG ('H_MG', 2043, ' ')\n",
      "    MG ('H_MG', 2044, ' ')\n",
      "    MG ('H_MG', 2045, ' ')\n",
      "    MG ('H_MG', 2046, ' ')\n",
      "    MG ('H_MG', 2047, ' ')\n",
      "    MG ('H_MG', 2048, ' ')\n",
      "    MG ('H_MG', 2049, ' ')\n",
      "    MG ('H_MG', 2050, ' ')\n",
      "    MG ('H_MG', 2051, ' ')\n",
      "    MG ('H_MG', 2052, ' ')\n",
      "    MG ('H_MG', 2053, ' ')\n",
      "    MG ('H_MG', 2054, ' ')\n",
      "    MG ('H_MG', 2055, ' ')\n",
      "    MG ('H_MG', 2056, ' ')\n",
      "    MG ('H_MG', 2057, ' ')\n",
      "    MG ('H_MG', 2058, ' ')\n",
      "    MG ('H_MG', 2059, ' ')\n",
      "    MG ('H_MG', 2060, ' ')\n",
      "    MG ('H_MG', 2061, ' ')\n",
      "    LLL ('H_LLL', 2062, ' ')\n",
      "    LLL ('H_LLL', 2063, ' ')\n",
      "    LLL ('H_LLL', 2064, ' ')\n",
      "    HOH ('W', 2101, ' ')\n",
      "    HOH ('W', 2102, ' ')\n",
      "    HOH ('W', 2103, ' ')\n",
      "    HOH ('W', 2104, ' ')\n",
      "    HOH ('W', 2105, ' ')\n",
      "    HOH ('W', 2106, ' ')\n",
      "    HOH ('W', 2107, ' ')\n",
      "    HOH ('W', 2108, ' ')\n",
      "    HOH ('W', 2109, ' ')\n",
      "    HOH ('W', 2110, ' ')\n",
      "    HOH ('W', 2111, ' ')\n",
      "    HOH ('W', 2112, ' ')\n",
      "    HOH ('W', 2113, ' ')\n",
      "    HOH ('W', 2114, ' ')\n",
      "    HOH ('W', 2115, ' ')\n",
      "    HOH ('W', 2116, ' ')\n",
      "    HOH ('W', 2117, ' ')\n",
      "    HOH ('W', 2118, ' ')\n",
      "    HOH ('W', 2119, ' ')\n",
      "    HOH ('W', 2120, ' ')\n",
      "    HOH ('W', 2121, ' ')\n",
      "    HOH ('W', 2122, ' ')\n",
      "    HOH ('W', 2123, ' ')\n",
      "    HOH ('W', 2124, ' ')\n",
      "    HOH ('W', 2125, ' ')\n",
      "    HOH ('W', 2126, ' ')\n",
      "    HOH ('W', 2127, ' ')\n",
      "    HOH ('W', 2128, ' ')\n",
      "    HOH ('W', 2129, ' ')\n",
      "    HOH ('W', 2130, ' ')\n",
      "    HOH ('W', 2131, ' ')\n",
      "    HOH ('W', 2132, ' ')\n",
      "    HOH ('W', 2133, ' ')\n",
      "    HOH ('W', 2134, ' ')\n",
      "    HOH ('W', 2135, ' ')\n",
      "    HOH ('W', 2136, ' ')\n",
      "    HOH ('W', 2137, ' ')\n",
      "    HOH ('W', 2138, ' ')\n",
      "    HOH ('W', 2139, ' ')\n",
      "    HOH ('W', 2140, ' ')\n",
      "    HOH ('W', 2141, ' ')\n",
      "    HOH ('W', 2142, ' ')\n",
      "    HOH ('W', 2143, ' ')\n",
      "    HOH ('W', 2144, ' ')\n",
      "    HOH ('W', 2145, ' ')\n",
      "    HOH ('W', 2146, ' ')\n",
      "    HOH ('W', 2147, ' ')\n",
      "    HOH ('W', 2148, ' ')\n",
      "    HOH ('W', 2149, ' ')\n",
      "    HOH ('W', 2150, ' ')\n",
      "    HOH ('W', 2151, ' ')\n",
      "    HOH ('W', 2152, ' ')\n",
      "    HOH ('W', 2153, ' ')\n",
      "    HOH ('W', 2154, ' ')\n",
      "    HOH ('W', 2155, ' ')\n",
      "    HOH ('W', 2156, ' ')\n",
      "    HOH ('W', 2157, ' ')\n",
      "    HOH ('W', 2158, ' ')\n",
      "    HOH ('W', 2159, ' ')\n",
      "    HOH ('W', 2160, ' ')\n",
      "    HOH ('W', 2161, ' ')\n",
      "    HOH ('W', 2162, ' ')\n",
      "    HOH ('W', 2163, ' ')\n",
      "    HOH ('W', 2164, ' ')\n",
      "    HOH ('W', 2165, ' ')\n",
      "    HOH ('W', 2166, ' ')\n",
      "    HOH ('W', 2167, ' ')\n",
      "    HOH ('W', 2168, ' ')\n",
      "    HOH ('W', 2169, ' ')\n",
      "    HOH ('W', 2170, ' ')\n",
      "    HOH ('W', 2171, ' ')\n",
      "    HOH ('W', 2172, ' ')\n",
      "    HOH ('W', 2173, ' ')\n",
      "    HOH ('W', 2174, ' ')\n",
      "    HOH ('W', 2175, ' ')\n",
      "    HOH ('W', 2176, ' ')\n",
      "    HOH ('W', 2177, ' ')\n",
      "    HOH ('W', 2178, ' ')\n",
      "    HOH ('W', 2179, ' ')\n",
      "    HOH ('W', 2180, ' ')\n",
      "    HOH ('W', 2181, ' ')\n",
      "    HOH ('W', 2182, ' ')\n",
      "    HOH ('W', 2183, ' ')\n",
      "    HOH ('W', 2184, ' ')\n",
      "    HOH ('W', 2185, ' ')\n",
      "    HOH ('W', 2186, ' ')\n",
      "    HOH ('W', 2187, ' ')\n",
      "    HOH ('W', 2188, ' ')\n",
      "    HOH ('W', 2189, ' ')\n",
      "    HOH ('W', 2190, ' ')\n",
      "    HOH ('W', 2191, ' ')\n",
      "    HOH ('W', 2192, ' ')\n",
      "    HOH ('W', 2193, ' ')\n",
      "    HOH ('W', 2194, ' ')\n",
      "    HOH ('W', 2195, ' ')\n",
      "    HOH ('W', 2196, ' ')\n",
      "    HOH ('W', 2197, ' ')\n",
      "    HOH ('W', 2198, ' ')\n",
      "    HOH ('W', 2199, ' ')\n",
      "    HOH ('W', 2200, ' ')\n",
      "    HOH ('W', 2201, ' ')\n",
      "    HOH ('W', 2202, ' ')\n",
      "    HOH ('W', 2203, ' ')\n",
      "    HOH ('W', 2204, ' ')\n",
      "    HOH ('W', 2205, ' ')\n",
      "    HOH ('W', 2206, ' ')\n",
      "    HOH ('W', 2207, ' ')\n",
      "    HOH ('W', 2208, ' ')\n",
      "    HOH ('W', 2209, ' ')\n",
      "    HOH ('W', 2210, ' ')\n",
      "    HOH ('W', 2211, ' ')\n",
      "    HOH ('W', 2212, ' ')\n",
      "    HOH ('W', 2213, ' ')\n",
      "    HOH ('W', 2214, ' ')\n",
      "    HOH ('W', 2215, ' ')\n",
      "    HOH ('W', 2216, ' ')\n",
      "    HOH ('W', 2217, ' ')\n",
      "    HOH ('W', 2218, ' ')\n",
      "    HOH ('W', 2219, ' ')\n",
      "    HOH ('W', 2220, ' ')\n",
      "    HOH ('W', 2221, ' ')\n",
      "    HOH ('W', 2222, ' ')\n",
      "    HOH ('W', 2223, ' ')\n",
      "    HOH ('W', 2224, ' ')\n",
      "    HOH ('W', 2225, ' ')\n",
      "    HOH ('W', 2226, ' ')\n",
      "    HOH ('W', 2227, ' ')\n",
      "    HOH ('W', 2228, ' ')\n",
      "    HOH ('W', 2229, ' ')\n",
      "    HOH ('W', 2230, ' ')\n",
      "    HOH ('W', 2231, ' ')\n",
      "    HOH ('W', 2232, ' ')\n",
      "    HOH ('W', 2233, ' ')\n",
      "    HOH ('W', 2234, ' ')\n",
      "    HOH ('W', 2235, ' ')\n",
      "    HOH ('W', 2236, ' ')\n",
      "    HOH ('W', 2237, ' ')\n",
      "    HOH ('W', 2238, ' ')\n",
      "    HOH ('W', 2239, ' ')\n",
      "    HOH ('W', 2240, ' ')\n",
      "    HOH ('W', 2241, ' ')\n",
      "    HOH ('W', 2242, ' ')\n",
      "    HOH ('W', 2243, ' ')\n",
      "    HOH ('W', 2244, ' ')\n",
      "    HOH ('W', 2245, ' ')\n",
      "    HOH ('W', 2246, ' ')\n",
      "    HOH ('W', 2247, ' ')\n",
      "    HOH ('W', 2248, ' ')\n",
      "    HOH ('W', 2249, ' ')\n",
      "    HOH ('W', 2250, ' ')\n",
      "    HOH ('W', 2251, ' ')\n",
      "    HOH ('W', 2252, ' ')\n",
      "    HOH ('W', 2253, ' ')\n",
      "    HOH ('W', 2254, ' ')\n",
      "    HOH ('W', 2255, ' ')\n",
      "    HOH ('W', 2256, ' ')\n",
      "    HOH ('W', 2257, ' ')\n",
      "    HOH ('W', 2258, ' ')\n",
      "    HOH ('W', 2259, ' ')\n",
      "    HOH ('W', 2260, ' ')\n",
      "    HOH ('W', 2261, ' ')\n",
      "    HOH ('W', 2262, ' ')\n",
      "    HOH ('W', 2263, ' ')\n",
      "    HOH ('W', 2264, ' ')\n",
      "    HOH ('W', 2265, ' ')\n",
      "    HOH ('W', 2266, ' ')\n",
      "    HOH ('W', 2267, ' ')\n",
      "    HOH ('W', 2268, ' ')\n",
      "    HOH ('W', 2269, ' ')\n",
      "    HOH ('W', 2270, ' ')\n",
      "    HOH ('W', 2271, ' ')\n",
      "    HOH ('W', 2272, ' ')\n",
      "    HOH ('W', 2273, ' ')\n",
      "    HOH ('W', 2274, ' ')\n",
      "    HOH ('W', 2275, ' ')\n",
      "    HOH ('W', 2276, ' ')\n",
      "    HOH ('W', 2277, ' ')\n",
      "    HOH ('W', 2278, ' ')\n",
      "    HOH ('W', 2279, ' ')\n",
      "    HOH ('W', 2280, ' ')\n",
      "    HOH ('W', 2281, ' ')\n",
      "    HOH ('W', 2282, ' ')\n",
      "    HOH ('W', 2283, ' ')\n",
      "    HOH ('W', 2284, ' ')\n",
      "    HOH ('W', 2285, ' ')\n",
      "    HOH ('W', 2286, ' ')\n",
      "    HOH ('W', 2287, ' ')\n",
      "    HOH ('W', 2288, ' ')\n",
      "    HOH ('W', 2289, ' ')\n",
      "    HOH ('W', 2290, ' ')\n",
      "    HOH ('W', 2291, ' ')\n",
      "    HOH ('W', 2292, ' ')\n",
      "    HOH ('W', 2293, ' ')\n",
      "    HOH ('W', 2294, ' ')\n",
      "    HOH ('W', 2295, ' ')\n",
      "    HOH ('W', 2296, ' ')\n",
      "    HOH ('W', 2297, ' ')\n",
      "    HOH ('W', 2298, ' ')\n",
      "    HOH ('W', 2299, ' ')\n",
      "    HOH ('W', 2300, ' ')\n",
      "    HOH ('W', 2301, ' ')\n",
      "    HOH ('W', 2302, ' ')\n",
      "    HOH ('W', 2303, ' ')\n",
      "    HOH ('W', 2304, ' ')\n",
      "    HOH ('W', 2305, ' ')\n",
      "    HOH ('W', 2306, ' ')\n",
      "    HOH ('W', 2307, ' ')\n",
      "    HOH ('W', 2308, ' ')\n",
      "    HOH ('W', 2309, ' ')\n",
      "    HOH ('W', 2310, ' ')\n",
      "    HOH ('W', 2311, ' ')\n",
      "    HOH ('W', 2312, ' ')\n",
      "    HOH ('W', 2313, ' ')\n",
      "    HOH ('W', 2314, ' ')\n",
      "    HOH ('W', 2315, ' ')\n",
      "    HOH ('W', 2316, ' ')\n",
      "    HOH ('W', 2317, ' ')\n",
      "    HOH ('W', 2318, ' ')\n",
      "    HOH ('W', 2319, ' ')\n",
      "    HOH ('W', 2320, ' ')\n",
      "    HOH ('W', 2321, ' ')\n",
      "    HOH ('W', 2322, ' ')\n",
      "    HOH ('W', 2323, ' ')\n",
      "    HOH ('W', 2324, ' ')\n",
      "    HOH ('W', 2325, ' ')\n",
      "    HOH ('W', 2326, ' ')\n",
      "    HOH ('W', 2327, ' ')\n",
      "    HOH ('W', 2328, ' ')\n",
      "    HOH ('W', 2329, ' ')\n",
      "    HOH ('W', 2330, ' ')\n",
      "    HOH ('W', 2331, ' ')\n",
      "    HOH ('W', 2332, ' ')\n",
      "    HOH ('W', 2333, ' ')\n",
      "    HOH ('W', 2334, ' ')\n",
      "    HOH ('W', 2335, ' ')\n",
      "    HOH ('W', 2336, ' ')\n",
      "    HOH ('W', 2337, ' ')\n",
      "    HOH ('W', 2338, ' ')\n",
      "    HOH ('W', 2339, ' ')\n",
      "    HOH ('W', 2340, ' ')\n",
      "    HOH ('W', 2341, ' ')\n",
      "    HOH ('W', 2342, ' ')\n",
      "    HOH ('W', 2343, ' ')\n",
      "    HOH ('W', 2344, ' ')\n",
      "    HOH ('W', 2345, ' ')\n",
      "    HOH ('W', 2346, ' ')\n",
      "    HOH ('W', 2347, ' ')\n",
      "    HOH ('W', 2348, ' ')\n",
      "    HOH ('W', 2349, ' ')\n",
      "    HOH ('W', 2350, ' ')\n",
      "    HOH ('W', 2351, ' ')\n",
      "    HOH ('W', 2352, ' ')\n",
      "    HOH ('W', 2353, ' ')\n",
      "    HOH ('W', 2354, ' ')\n",
      "    HOH ('W', 2355, ' ')\n",
      "    HOH ('W', 2356, ' ')\n",
      "    HOH ('W', 2357, ' ')\n",
      "    HOH ('W', 2358, ' ')\n",
      "    HOH ('W', 2359, ' ')\n",
      "    HOH ('W', 2360, ' ')\n",
      "    HOH ('W', 2361, ' ')\n",
      "    HOH ('W', 2362, ' ')\n",
      "    HOH ('W', 2363, ' ')\n",
      "    HOH ('W', 2364, ' ')\n",
      "    HOH ('W', 2365, ' ')\n",
      "    HOH ('W', 2366, ' ')\n",
      "    HOH ('W', 2367, ' ')\n",
      "    HOH ('W', 2368, ' ')\n",
      "    HOH ('W', 2369, ' ')\n",
      "    HOH ('W', 2370, ' ')\n",
      "    HOH ('W', 2371, ' ')\n",
      "    HOH ('W', 2372, ' ')\n",
      "    HOH ('W', 2373, ' ')\n",
      "    HOH ('W', 2374, ' ')\n",
      "    HOH ('W', 2375, ' ')\n",
      "    HOH ('W', 2376, ' ')\n",
      "    HOH ('W', 2377, ' ')\n",
      "    HOH ('W', 2378, ' ')\n",
      "    HOH ('W', 2379, ' ')\n",
      "    HOH ('W', 2380, ' ')\n",
      "    HOH ('W', 2381, ' ')\n",
      "    HOH ('W', 2382, ' ')\n",
      "    HOH ('W', 2383, ' ')\n",
      "    HOH ('W', 2384, ' ')\n",
      "    HOH ('W', 2385, ' ')\n",
      "    HOH ('W', 2386, ' ')\n",
      "    HOH ('W', 2387, ' ')\n",
      "    HOH ('W', 2388, ' ')\n",
      "    HOH ('W', 2389, ' ')\n",
      "    HOH ('W', 2390, ' ')\n",
      "    HOH ('W', 2391, ' ')\n",
      "    HOH ('W', 2392, ' ')\n",
      "    HOH ('W', 2393, ' ')\n",
      "    HOH ('W', 2394, ' ')\n",
      "    HOH ('W', 2395, ' ')\n",
      "    HOH ('W', 2396, ' ')\n",
      "    HOH ('W', 2397, ' ')\n",
      "  Chain CC:\n",
      "    No hetero residues found\n",
      "  Chain CD:\n",
      "    No hetero residues found\n",
      "  Chain CE:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "  Chain CF:\n",
      "    No hetero residues found\n",
      "  Chain CG:\n",
      "    No hetero residues found\n",
      "  Chain CH:\n",
      "    No hetero residues found\n",
      "  Chain CI:\n",
      "    No hetero residues found\n",
      "  Chain CJ:\n",
      "    No hetero residues found\n",
      "  Chain CK:\n",
      "    HOH ('W', 201, ' ')\n",
      "  Chain CL_new:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "  Chain CM:\n",
      "    No hetero residues found\n",
      "  Chain CN:\n",
      "    MG ('H_MG', 201, ' ')\n",
      "    HOH ('W', 301, ' ')\n",
      "    HOH ('W', 302, ' ')\n",
      "    HOH ('W', 303, ' ')\n",
      "    HOH ('W', 304, ' ')\n",
      "  Chain CO:\n",
      "    No hetero residues found\n",
      "  Chain CP:\n",
      "    No hetero residues found\n",
      "  Chain CQ:\n",
      "    No hetero residues found\n",
      "  Chain CR:\n",
      "    No hetero residues found\n",
      "  Chain CS:\n",
      "    No hetero residues found\n",
      "  Chain CT:\n",
      "    HOH ('W', 101, ' ')\n",
      "    HOH ('W', 102, ' ')\n",
      "  Chain CB:\n",
      "    No hetero residues found\n",
      "  Chain CU:\n",
      "    No hetero residues found\n",
      "  Chain DA:\n",
      "    No hetero residues found\n",
      "  Chain DB:\n",
      "    MG ('H_MG', 3001, ' ')\n",
      "    MG ('H_MG', 3002, ' ')\n",
      "    MG ('H_MG', 3003, ' ')\n",
      "    MG ('H_MG', 3004, ' ')\n",
      "    MG ('H_MG', 3005, ' ')\n",
      "    MG ('H_MG', 3006, ' ')\n",
      "    MG ('H_MG', 3007, ' ')\n",
      "    MG ('H_MG', 3008, ' ')\n",
      "    MG ('H_MG', 3009, ' ')\n",
      "    MG ('H_MG', 3010, ' ')\n",
      "    MG ('H_MG', 3011, ' ')\n",
      "    MG ('H_MG', 3012, ' ')\n",
      "    MG ('H_MG', 3013, ' ')\n",
      "    MG ('H_MG', 3014, ' ')\n",
      "    MG ('H_MG', 3015, ' ')\n",
      "    MG ('H_MG', 3016, ' ')\n",
      "    MG ('H_MG', 3017, ' ')\n",
      "    MG ('H_MG', 3018, ' ')\n",
      "    MG ('H_MG', 3019, ' ')\n",
      "    MG ('H_MG', 3020, ' ')\n",
      "    MG ('H_MG', 3021, ' ')\n",
      "    MG ('H_MG', 3022, ' ')\n",
      "    MG ('H_MG', 3023, ' ')\n",
      "    MG ('H_MG', 3024, ' ')\n",
      "    MG ('H_MG', 3025, ' ')\n",
      "    MG ('H_MG', 3026, ' ')\n",
      "    MG ('H_MG', 3027, ' ')\n",
      "    MG ('H_MG', 3028, ' ')\n",
      "    MG ('H_MG', 3029, ' ')\n",
      "    MG ('H_MG', 3030, ' ')\n",
      "    MG ('H_MG', 3031, ' ')\n",
      "    MG ('H_MG', 3032, ' ')\n",
      "    MG ('H_MG', 3033, ' ')\n",
      "    MG ('H_MG', 3034, ' ')\n",
      "    MG ('H_MG', 3035, ' ')\n",
      "    MG ('H_MG', 3036, ' ')\n",
      "    MG ('H_MG', 3037, ' ')\n",
      "    MG ('H_MG', 3038, ' ')\n",
      "    MG ('H_MG', 3039, ' ')\n",
      "    MG ('H_MG', 3040, ' ')\n",
      "    MG ('H_MG', 3041, ' ')\n",
      "    MG ('H_MG', 3042, ' ')\n",
      "    MG ('H_MG', 3043, ' ')\n",
      "    MG ('H_MG', 3044, ' ')\n",
      "    MG ('H_MG', 3045, ' ')\n",
      "    MG ('H_MG', 3046, ' ')\n",
      "    MG ('H_MG', 3047, ' ')\n",
      "    MG ('H_MG', 3048, ' ')\n",
      "    MG ('H_MG', 3049, ' ')\n",
      "    MG ('H_MG', 3050, ' ')\n",
      "    MG ('H_MG', 3051, ' ')\n",
      "    MG ('H_MG', 3052, ' ')\n",
      "    MG ('H_MG', 3053, ' ')\n",
      "    MG ('H_MG', 3054, ' ')\n",
      "    MG ('H_MG', 3055, ' ')\n",
      "    MG ('H_MG', 3056, ' ')\n",
      "    MG ('H_MG', 3057, ' ')\n",
      "    MG ('H_MG', 3058, ' ')\n",
      "    MG ('H_MG', 3059, ' ')\n",
      "    MG ('H_MG', 3060, ' ')\n",
      "    MG ('H_MG', 3061, ' ')\n",
      "    MG ('H_MG', 3062, ' ')\n",
      "    MG ('H_MG', 3063, ' ')\n",
      "    MG ('H_MG', 3064, ' ')\n",
      "    MG ('H_MG', 3065, ' ')\n",
      "    MG ('H_MG', 3066, ' ')\n",
      "    MG ('H_MG', 3067, ' ')\n",
      "    MG ('H_MG', 3068, ' ')\n",
      "    MG ('H_MG', 3069, ' ')\n",
      "    MG ('H_MG', 3070, ' ')\n",
      "    MG ('H_MG', 3071, ' ')\n",
      "    MG ('H_MG', 3072, ' ')\n",
      "    MG ('H_MG', 3073, ' ')\n",
      "    MG ('H_MG', 3074, ' ')\n",
      "    MG ('H_MG', 3075, ' ')\n",
      "    MG ('H_MG', 3076, ' ')\n",
      "    MG ('H_MG', 3077, ' ')\n",
      "    MG ('H_MG', 3078, ' ')\n",
      "    MG ('H_MG', 3079, ' ')\n",
      "    MG ('H_MG', 3080, ' ')\n",
      "    MG ('H_MG', 3081, ' ')\n",
      "    MG ('H_MG', 3082, ' ')\n",
      "    MG ('H_MG', 3083, ' ')\n",
      "    MG ('H_MG', 3084, ' ')\n",
      "    MG ('H_MG', 3085, ' ')\n",
      "    MG ('H_MG', 3086, ' ')\n",
      "    MG ('H_MG', 3087, ' ')\n",
      "    MG ('H_MG', 3088, ' ')\n",
      "    MG ('H_MG', 3089, ' ')\n",
      "    MG ('H_MG', 3090, ' ')\n",
      "    MG ('H_MG', 3091, ' ')\n",
      "    MG ('H_MG', 3092, ' ')\n",
      "    MG ('H_MG', 3093, ' ')\n",
      "    MG ('H_MG', 3094, ' ')\n",
      "    MG ('H_MG', 3095, ' ')\n",
      "    MG ('H_MG', 3096, ' ')\n",
      "    MG ('H_MG', 3097, ' ')\n",
      "    MG ('H_MG', 3098, ' ')\n",
      "    MG ('H_MG', 3099, ' ')\n",
      "    MG ('H_MG', 3100, ' ')\n",
      "    MG ('H_MG', 3101, ' ')\n",
      "    MG ('H_MG', 3102, ' ')\n",
      "    MG ('H_MG', 3103, ' ')\n",
      "    MG ('H_MG', 3104, ' ')\n",
      "    MG ('H_MG', 3105, ' ')\n",
      "    MG ('H_MG', 3106, ' ')\n",
      "    MG ('H_MG', 3107, ' ')\n",
      "    MG ('H_MG', 3108, ' ')\n",
      "    MG ('H_MG', 3109, ' ')\n",
      "    MG ('H_MG', 3110, ' ')\n",
      "    MG ('H_MG', 3111, ' ')\n",
      "    LLL ('H_LLL', 3112, ' ')\n",
      "    HOH ('W', 3201, ' ')\n",
      "    HOH ('W', 3202, ' ')\n",
      "    HOH ('W', 3203, ' ')\n",
      "    HOH ('W', 3204, ' ')\n",
      "    HOH ('W', 3205, ' ')\n",
      "    HOH ('W', 3206, ' ')\n",
      "    HOH ('W', 3207, ' ')\n",
      "    HOH ('W', 3208, ' ')\n",
      "    HOH ('W', 3209, ' ')\n",
      "    HOH ('W', 3210, ' ')\n",
      "    HOH ('W', 3211, ' ')\n",
      "    HOH ('W', 3212, ' ')\n",
      "    HOH ('W', 3213, ' ')\n",
      "    HOH ('W', 3214, ' ')\n",
      "    HOH ('W', 3215, ' ')\n",
      "    HOH ('W', 3216, ' ')\n",
      "    HOH ('W', 3217, ' ')\n",
      "    HOH ('W', 3218, ' ')\n",
      "    HOH ('W', 3219, ' ')\n",
      "    HOH ('W', 3220, ' ')\n",
      "    HOH ('W', 3221, ' ')\n",
      "    HOH ('W', 3222, ' ')\n",
      "    HOH ('W', 3223, ' ')\n",
      "    HOH ('W', 3224, ' ')\n",
      "    HOH ('W', 3225, ' ')\n",
      "    HOH ('W', 3226, ' ')\n",
      "    HOH ('W', 3227, ' ')\n",
      "    HOH ('W', 3228, ' ')\n",
      "    HOH ('W', 3229, ' ')\n",
      "    HOH ('W', 3230, ' ')\n",
      "    HOH ('W', 3231, ' ')\n",
      "    HOH ('W', 3232, ' ')\n",
      "    HOH ('W', 3233, ' ')\n",
      "    HOH ('W', 3234, ' ')\n",
      "    HOH ('W', 3235, ' ')\n",
      "    HOH ('W', 3236, ' ')\n",
      "    HOH ('W', 3237, ' ')\n",
      "    HOH ('W', 3238, ' ')\n",
      "    HOH ('W', 3239, ' ')\n",
      "    HOH ('W', 3240, ' ')\n",
      "    HOH ('W', 3241, ' ')\n",
      "    HOH ('W', 3242, ' ')\n",
      "    HOH ('W', 3243, ' ')\n",
      "    HOH ('W', 3244, ' ')\n",
      "    HOH ('W', 3245, ' ')\n",
      "    HOH ('W', 3246, ' ')\n",
      "    HOH ('W', 3247, ' ')\n",
      "    HOH ('W', 3248, ' ')\n",
      "    HOH ('W', 3249, ' ')\n",
      "    HOH ('W', 3250, ' ')\n",
      "    HOH ('W', 3251, ' ')\n",
      "    HOH ('W', 3252, ' ')\n",
      "    HOH ('W', 3253, ' ')\n",
      "    HOH ('W', 3254, ' ')\n",
      "    HOH ('W', 3255, ' ')\n",
      "    HOH ('W', 3256, ' ')\n",
      "    HOH ('W', 3257, ' ')\n",
      "    HOH ('W', 3258, ' ')\n",
      "    HOH ('W', 3259, ' ')\n",
      "    HOH ('W', 3260, ' ')\n",
      "    HOH ('W', 3261, ' ')\n",
      "    HOH ('W', 3262, ' ')\n",
      "    HOH ('W', 3263, ' ')\n",
      "    HOH ('W', 3264, ' ')\n",
      "    HOH ('W', 3265, ' ')\n",
      "    HOH ('W', 3266, ' ')\n",
      "    HOH ('W', 3267, ' ')\n",
      "    HOH ('W', 3268, ' ')\n",
      "    HOH ('W', 3269, ' ')\n",
      "    HOH ('W', 3270, ' ')\n",
      "    HOH ('W', 3271, ' ')\n",
      "    HOH ('W', 3272, ' ')\n",
      "    HOH ('W', 3273, ' ')\n",
      "    HOH ('W', 3274, ' ')\n",
      "    HOH ('W', 3275, ' ')\n",
      "    HOH ('W', 3276, ' ')\n",
      "    HOH ('W', 3277, ' ')\n",
      "    HOH ('W', 3278, ' ')\n",
      "    HOH ('W', 3279, ' ')\n",
      "    HOH ('W', 3280, ' ')\n",
      "    HOH ('W', 3281, ' ')\n",
      "    HOH ('W', 3282, ' ')\n",
      "    HOH ('W', 3283, ' ')\n",
      "    HOH ('W', 3284, ' ')\n",
      "    HOH ('W', 3285, ' ')\n",
      "    HOH ('W', 3286, ' ')\n",
      "    HOH ('W', 3287, ' ')\n",
      "    HOH ('W', 3288, ' ')\n",
      "    HOH ('W', 3289, ' ')\n",
      "    HOH ('W', 3290, ' ')\n",
      "    HOH ('W', 3291, ' ')\n",
      "    HOH ('W', 3292, ' ')\n",
      "    HOH ('W', 3293, ' ')\n",
      "    HOH ('W', 3294, ' ')\n",
      "    HOH ('W', 3295, ' ')\n",
      "    HOH ('W', 3296, ' ')\n",
      "    HOH ('W', 3297, ' ')\n",
      "    HOH ('W', 3298, ' ')\n",
      "    HOH ('W', 3299, ' ')\n",
      "    HOH ('W', 3300, ' ')\n",
      "    HOH ('W', 3301, ' ')\n",
      "    HOH ('W', 3302, ' ')\n",
      "    HOH ('W', 3303, ' ')\n",
      "    HOH ('W', 3304, ' ')\n",
      "    HOH ('W', 3305, ' ')\n",
      "    HOH ('W', 3306, ' ')\n",
      "    HOH ('W', 3307, ' ')\n",
      "    HOH ('W', 3308, ' ')\n",
      "    HOH ('W', 3309, ' ')\n",
      "    HOH ('W', 3310, ' ')\n",
      "    HOH ('W', 3311, ' ')\n",
      "    HOH ('W', 3312, ' ')\n",
      "    HOH ('W', 3313, ' ')\n",
      "    HOH ('W', 3314, ' ')\n",
      "    HOH ('W', 3315, ' ')\n",
      "    HOH ('W', 3316, ' ')\n",
      "    HOH ('W', 3317, ' ')\n",
      "    HOH ('W', 3318, ' ')\n",
      "    HOH ('W', 3319, ' ')\n",
      "    HOH ('W', 3320, ' ')\n",
      "    HOH ('W', 3321, ' ')\n",
      "    HOH ('W', 3322, ' ')\n",
      "    HOH ('W', 3323, ' ')\n",
      "    HOH ('W', 3324, ' ')\n",
      "    HOH ('W', 3325, ' ')\n",
      "    HOH ('W', 3326, ' ')\n",
      "    HOH ('W', 3327, ' ')\n",
      "    HOH ('W', 3328, ' ')\n",
      "    HOH ('W', 3329, ' ')\n",
      "    HOH ('W', 3330, ' ')\n",
      "    HOH ('W', 3331, ' ')\n",
      "    HOH ('W', 3332, ' ')\n",
      "    HOH ('W', 3333, ' ')\n",
      "    HOH ('W', 3334, ' ')\n",
      "    HOH ('W', 3335, ' ')\n",
      "    HOH ('W', 3336, ' ')\n",
      "    HOH ('W', 3337, ' ')\n",
      "    HOH ('W', 3338, ' ')\n",
      "    HOH ('W', 3339, ' ')\n",
      "    HOH ('W', 3340, ' ')\n",
      "    HOH ('W', 3341, ' ')\n",
      "    HOH ('W', 3342, ' ')\n",
      "    HOH ('W', 3343, ' ')\n",
      "    HOH ('W', 3344, ' ')\n",
      "    HOH ('W', 3345, ' ')\n",
      "    HOH ('W', 3346, ' ')\n",
      "    HOH ('W', 3347, ' ')\n",
      "    HOH ('W', 3348, ' ')\n",
      "    HOH ('W', 3349, ' ')\n",
      "    HOH ('W', 3350, ' ')\n",
      "    HOH ('W', 3351, ' ')\n",
      "    HOH ('W', 3352, ' ')\n",
      "    HOH ('W', 3353, ' ')\n",
      "    HOH ('W', 3354, ' ')\n",
      "    HOH ('W', 3355, ' ')\n",
      "    HOH ('W', 3356, ' ')\n",
      "    HOH ('W', 3357, ' ')\n",
      "    HOH ('W', 3358, ' ')\n",
      "    HOH ('W', 3359, ' ')\n",
      "    HOH ('W', 3360, ' ')\n",
      "    HOH ('W', 3361, ' ')\n",
      "    HOH ('W', 3362, ' ')\n",
      "    HOH ('W', 3363, ' ')\n",
      "    HOH ('W', 3364, ' ')\n",
      "    HOH ('W', 3365, ' ')\n",
      "    HOH ('W', 3366, ' ')\n",
      "    HOH ('W', 3367, ' ')\n",
      "    HOH ('W', 3368, ' ')\n",
      "    HOH ('W', 3369, ' ')\n",
      "    HOH ('W', 3370, ' ')\n",
      "    HOH ('W', 3371, ' ')\n",
      "    HOH ('W', 3372, ' ')\n",
      "    HOH ('W', 3373, ' ')\n",
      "    HOH ('W', 3374, ' ')\n",
      "    HOH ('W', 3375, ' ')\n",
      "    HOH ('W', 3376, ' ')\n",
      "    HOH ('W', 3377, ' ')\n",
      "    HOH ('W', 3378, ' ')\n",
      "    HOH ('W', 3379, ' ')\n",
      "    HOH ('W', 3380, ' ')\n",
      "    HOH ('W', 3381, ' ')\n",
      "    HOH ('W', 3382, ' ')\n",
      "    HOH ('W', 3383, ' ')\n",
      "    HOH ('W', 3384, ' ')\n",
      "    HOH ('W', 3385, ' ')\n",
      "    HOH ('W', 3386, ' ')\n",
      "    HOH ('W', 3387, ' ')\n",
      "    HOH ('W', 3388, ' ')\n",
      "    HOH ('W', 3389, ' ')\n",
      "    HOH ('W', 3390, ' ')\n",
      "    HOH ('W', 3391, ' ')\n",
      "    HOH ('W', 3392, ' ')\n",
      "    HOH ('W', 3393, ' ')\n",
      "    HOH ('W', 3394, ' ')\n",
      "    HOH ('W', 3395, ' ')\n",
      "    HOH ('W', 3396, ' ')\n",
      "    HOH ('W', 3397, ' ')\n",
      "    HOH ('W', 3398, ' ')\n",
      "    HOH ('W', 3399, ' ')\n",
      "    HOH ('W', 3400, ' ')\n",
      "    HOH ('W', 3401, ' ')\n",
      "    HOH ('W', 3402, ' ')\n",
      "    HOH ('W', 3403, ' ')\n",
      "    HOH ('W', 3404, ' ')\n",
      "    HOH ('W', 3405, ' ')\n",
      "    HOH ('W', 3406, ' ')\n",
      "    HOH ('W', 3407, ' ')\n",
      "    HOH ('W', 3408, ' ')\n",
      "    HOH ('W', 3409, ' ')\n",
      "    HOH ('W', 3410, ' ')\n",
      "    HOH ('W', 3411, ' ')\n",
      "    HOH ('W', 3412, ' ')\n",
      "    HOH ('W', 3413, ' ')\n",
      "    HOH ('W', 3414, ' ')\n",
      "    HOH ('W', 3415, ' ')\n",
      "    HOH ('W', 3416, ' ')\n",
      "    HOH ('W', 3417, ' ')\n",
      "    HOH ('W', 3418, ' ')\n",
      "    HOH ('W', 3419, ' ')\n",
      "    HOH ('W', 3420, ' ')\n",
      "    HOH ('W', 3421, ' ')\n",
      "    HOH ('W', 3422, ' ')\n",
      "    HOH ('W', 3423, ' ')\n",
      "    HOH ('W', 3424, ' ')\n",
      "    HOH ('W', 3425, ' ')\n",
      "    HOH ('W', 3426, ' ')\n",
      "    HOH ('W', 3427, ' ')\n",
      "    HOH ('W', 3428, ' ')\n",
      "    HOH ('W', 3429, ' ')\n",
      "    HOH ('W', 3430, ' ')\n",
      "    HOH ('W', 3431, ' ')\n",
      "    HOH ('W', 3432, ' ')\n",
      "    HOH ('W', 3433, ' ')\n",
      "    HOH ('W', 3434, ' ')\n",
      "    HOH ('W', 3435, ' ')\n",
      "    HOH ('W', 3436, ' ')\n",
      "    HOH ('W', 3437, ' ')\n",
      "    HOH ('W', 3438, ' ')\n",
      "    HOH ('W', 3439, ' ')\n",
      "    HOH ('W', 3440, ' ')\n",
      "    HOH ('W', 3441, ' ')\n",
      "    HOH ('W', 3442, ' ')\n",
      "    HOH ('W', 3443, ' ')\n",
      "    HOH ('W', 3444, ' ')\n",
      "    HOH ('W', 3445, ' ')\n",
      "    HOH ('W', 3446, ' ')\n",
      "    HOH ('W', 3447, ' ')\n",
      "    HOH ('W', 3448, ' ')\n",
      "    HOH ('W', 3449, ' ')\n",
      "    HOH ('W', 3450, ' ')\n",
      "    HOH ('W', 3451, ' ')\n",
      "    HOH ('W', 3452, ' ')\n",
      "    HOH ('W', 3453, ' ')\n",
      "    HOH ('W', 3454, ' ')\n",
      "    HOH ('W', 3455, ' ')\n",
      "    HOH ('W', 3456, ' ')\n",
      "    HOH ('W', 3457, ' ')\n",
      "    HOH ('W', 3458, ' ')\n",
      "    HOH ('W', 3459, ' ')\n",
      "    HOH ('W', 3460, ' ')\n",
      "    HOH ('W', 3461, ' ')\n",
      "    HOH ('W', 3462, ' ')\n",
      "    HOH ('W', 3463, ' ')\n",
      "    HOH ('W', 3464, ' ')\n",
      "    HOH ('W', 3465, ' ')\n",
      "    HOH ('W', 3466, ' ')\n",
      "    HOH ('W', 3467, ' ')\n",
      "    HOH ('W', 3468, ' ')\n",
      "    HOH ('W', 3469, ' ')\n",
      "    HOH ('W', 3470, ' ')\n",
      "    HOH ('W', 3471, ' ')\n",
      "    HOH ('W', 3472, ' ')\n",
      "    HOH ('W', 3473, ' ')\n",
      "    HOH ('W', 3474, ' ')\n",
      "    HOH ('W', 3475, ' ')\n",
      "    HOH ('W', 3476, ' ')\n",
      "    HOH ('W', 3477, ' ')\n",
      "    HOH ('W', 3478, ' ')\n",
      "    HOH ('W', 3479, ' ')\n",
      "    HOH ('W', 3480, ' ')\n",
      "    HOH ('W', 3481, ' ')\n",
      "    HOH ('W', 3482, ' ')\n",
      "    HOH ('W', 3483, ' ')\n",
      "    HOH ('W', 3484, ' ')\n",
      "    HOH ('W', 3485, ' ')\n",
      "    HOH ('W', 3486, ' ')\n",
      "    HOH ('W', 3487, ' ')\n",
      "    HOH ('W', 3488, ' ')\n",
      "    HOH ('W', 3489, ' ')\n",
      "    HOH ('W', 3490, ' ')\n",
      "    HOH ('W', 3491, ' ')\n",
      "    HOH ('W', 3492, ' ')\n",
      "    HOH ('W', 3493, ' ')\n",
      "    HOH ('W', 3494, ' ')\n",
      "    HOH ('W', 3495, ' ')\n",
      "    HOH ('W', 3496, ' ')\n",
      "    HOH ('W', 3497, ' ')\n",
      "    HOH ('W', 3498, ' ')\n",
      "    HOH ('W', 3499, ' ')\n",
      "    HOH ('W', 3500, ' ')\n",
      "    HOH ('W', 3501, ' ')\n",
      "    HOH ('W', 3502, ' ')\n",
      "    HOH ('W', 3503, ' ')\n",
      "    HOH ('W', 3504, ' ')\n",
      "    HOH ('W', 3505, ' ')\n",
      "    HOH ('W', 3506, ' ')\n",
      "    HOH ('W', 3507, ' ')\n",
      "    HOH ('W', 3508, ' ')\n",
      "    HOH ('W', 3509, ' ')\n",
      "    HOH ('W', 3510, ' ')\n",
      "    HOH ('W', 3511, ' ')\n",
      "    HOH ('W', 3512, ' ')\n",
      "    HOH ('W', 3513, ' ')\n",
      "    HOH ('W', 3514, ' ')\n",
      "    HOH ('W', 3515, ' ')\n",
      "    HOH ('W', 3516, ' ')\n",
      "    HOH ('W', 3517, ' ')\n",
      "    HOH ('W', 3518, ' ')\n",
      "    HOH ('W', 3519, ' ')\n",
      "    HOH ('W', 3520, ' ')\n",
      "    HOH ('W', 3521, ' ')\n",
      "    HOH ('W', 3522, ' ')\n",
      "    HOH ('W', 3523, ' ')\n",
      "    HOH ('W', 3524, ' ')\n",
      "    HOH ('W', 3525, ' ')\n",
      "    HOH ('W', 3526, ' ')\n",
      "    HOH ('W', 3527, ' ')\n",
      "    HOH ('W', 3528, ' ')\n",
      "    HOH ('W', 3529, ' ')\n",
      "    HOH ('W', 3530, ' ')\n",
      "    HOH ('W', 3531, ' ')\n",
      "    HOH ('W', 3532, ' ')\n",
      "    HOH ('W', 3533, ' ')\n",
      "    HOH ('W', 3534, ' ')\n",
      "    HOH ('W', 3535, ' ')\n",
      "    HOH ('W', 3536, ' ')\n",
      "    HOH ('W', 3537, ' ')\n",
      "    HOH ('W', 3538, ' ')\n",
      "    HOH ('W', 3539, ' ')\n",
      "    HOH ('W', 3540, ' ')\n",
      "    HOH ('W', 3541, ' ')\n",
      "    HOH ('W', 3542, ' ')\n",
      "    HOH ('W', 3543, ' ')\n",
      "    HOH ('W', 3544, ' ')\n",
      "    HOH ('W', 3545, ' ')\n",
      "    HOH ('W', 3546, ' ')\n",
      "    HOH ('W', 3547, ' ')\n",
      "    HOH ('W', 3548, ' ')\n",
      "    HOH ('W', 3549, ' ')\n",
      "    HOH ('W', 3550, ' ')\n",
      "    HOH ('W', 3551, ' ')\n",
      "    HOH ('W', 3552, ' ')\n",
      "    HOH ('W', 3553, ' ')\n",
      "    HOH ('W', 3554, ' ')\n",
      "    HOH ('W', 3555, ' ')\n",
      "    HOH ('W', 3556, ' ')\n",
      "    HOH ('W', 3557, ' ')\n",
      "    HOH ('W', 3558, ' ')\n",
      "    HOH ('W', 3559, ' ')\n",
      "    HOH ('W', 3560, ' ')\n",
      "    HOH ('W', 3561, ' ')\n",
      "    HOH ('W', 3562, ' ')\n",
      "    HOH ('W', 3563, ' ')\n",
      "    HOH ('W', 3564, ' ')\n",
      "    HOH ('W', 3565, ' ')\n",
      "    HOH ('W', 3566, ' ')\n",
      "    HOH ('W', 3567, ' ')\n",
      "    HOH ('W', 3568, ' ')\n",
      "    HOH ('W', 3569, ' ')\n",
      "    HOH ('W', 3570, ' ')\n",
      "    HOH ('W', 3571, ' ')\n",
      "    HOH ('W', 3572, ' ')\n",
      "    HOH ('W', 3573, ' ')\n",
      "    HOH ('W', 3574, ' ')\n",
      "    HOH ('W', 3575, ' ')\n",
      "    HOH ('W', 3576, ' ')\n",
      "    HOH ('W', 3577, ' ')\n",
      "    HOH ('W', 3578, ' ')\n",
      "    HOH ('W', 3579, ' ')\n",
      "    HOH ('W', 3580, ' ')\n",
      "    HOH ('W', 3581, ' ')\n",
      "    HOH ('W', 3582, ' ')\n",
      "    HOH ('W', 3583, ' ')\n",
      "    HOH ('W', 3584, ' ')\n",
      "    HOH ('W', 3585, ' ')\n",
      "    HOH ('W', 3586, ' ')\n",
      "    HOH ('W', 3587, ' ')\n",
      "    HOH ('W', 3588, ' ')\n",
      "    HOH ('W', 3589, ' ')\n",
      "    HOH ('W', 3590, ' ')\n",
      "    HOH ('W', 3591, ' ')\n",
      "    HOH ('W', 3592, ' ')\n",
      "    HOH ('W', 3593, ' ')\n",
      "    HOH ('W', 3594, ' ')\n",
      "    HOH ('W', 3595, ' ')\n",
      "    HOH ('W', 3596, ' ')\n",
      "    HOH ('W', 3597, ' ')\n",
      "    HOH ('W', 3598, ' ')\n",
      "    HOH ('W', 3599, ' ')\n",
      "    HOH ('W', 3600, ' ')\n",
      "    HOH ('W', 3601, ' ')\n",
      "    HOH ('W', 3602, ' ')\n",
      "    HOH ('W', 3603, ' ')\n",
      "    HOH ('W', 3604, ' ')\n",
      "    HOH ('W', 3605, ' ')\n",
      "    HOH ('W', 3606, ' ')\n",
      "    HOH ('W', 3607, ' ')\n",
      "    HOH ('W', 3608, ' ')\n",
      "    HOH ('W', 3609, ' ')\n",
      "    HOH ('W', 3610, ' ')\n",
      "    HOH ('W', 3611, ' ')\n",
      "    HOH ('W', 3612, ' ')\n",
      "    HOH ('W', 3613, ' ')\n",
      "    HOH ('W', 3614, ' ')\n",
      "    HOH ('W', 3615, ' ')\n",
      "    HOH ('W', 3616, ' ')\n",
      "    HOH ('W', 3617, ' ')\n",
      "    HOH ('W', 3618, ' ')\n",
      "    HOH ('W', 3619, ' ')\n",
      "    HOH ('W', 3620, ' ')\n",
      "    HOH ('W', 3621, ' ')\n",
      "    HOH ('W', 3622, ' ')\n",
      "    HOH ('W', 3623, ' ')\n",
      "    HOH ('W', 3624, ' ')\n",
      "    HOH ('W', 3625, ' ')\n",
      "    HOH ('W', 3626, ' ')\n",
      "    HOH ('W', 3627, ' ')\n",
      "    HOH ('W', 3628, ' ')\n",
      "    HOH ('W', 3629, ' ')\n",
      "    HOH ('W', 3630, ' ')\n",
      "    HOH ('W', 3631, ' ')\n",
      "    HOH ('W', 3632, ' ')\n",
      "    HOH ('W', 3633, ' ')\n",
      "    HOH ('W', 3634, ' ')\n",
      "    HOH ('W', 3635, ' ')\n",
      "    HOH ('W', 3636, ' ')\n",
      "    HOH ('W', 3637, ' ')\n",
      "    HOH ('W', 3638, ' ')\n",
      "    HOH ('W', 3639, ' ')\n",
      "    HOH ('W', 3640, ' ')\n",
      "    HOH ('W', 3641, ' ')\n",
      "    HOH ('W', 3642, ' ')\n",
      "    HOH ('W', 3643, ' ')\n",
      "    HOH ('W', 3644, ' ')\n",
      "    HOH ('W', 3645, ' ')\n",
      "    HOH ('W', 3646, ' ')\n",
      "    HOH ('W', 3647, ' ')\n",
      "    HOH ('W', 3648, ' ')\n",
      "    HOH ('W', 3649, ' ')\n",
      "    HOH ('W', 3650, ' ')\n",
      "    HOH ('W', 3651, ' ')\n",
      "    HOH ('W', 3652, ' ')\n",
      "    HOH ('W', 3653, ' ')\n",
      "    HOH ('W', 3654, ' ')\n",
      "    HOH ('W', 3655, ' ')\n",
      "    HOH ('W', 3656, ' ')\n",
      "    HOH ('W', 3657, ' ')\n",
      "    HOH ('W', 3658, ' ')\n",
      "    HOH ('W', 3659, ' ')\n",
      "    HOH ('W', 3660, ' ')\n",
      "    HOH ('W', 3661, ' ')\n",
      "    HOH ('W', 3662, ' ')\n",
      "    HOH ('W', 3663, ' ')\n",
      "    HOH ('W', 3664, ' ')\n",
      "    HOH ('W', 3665, ' ')\n",
      "    HOH ('W', 3666, ' ')\n",
      "    HOH ('W', 3667, ' ')\n",
      "    HOH ('W', 3668, ' ')\n",
      "    HOH ('W', 3669, ' ')\n",
      "    HOH ('W', 3670, ' ')\n",
      "    HOH ('W', 3671, ' ')\n",
      "    HOH ('W', 3672, ' ')\n",
      "    HOH ('W', 3673, ' ')\n",
      "    HOH ('W', 3674, ' ')\n",
      "    HOH ('W', 3675, ' ')\n",
      "    HOH ('W', 3676, ' ')\n",
      "    HOH ('W', 3677, ' ')\n",
      "    HOH ('W', 3678, ' ')\n",
      "    HOH ('W', 3679, ' ')\n",
      "    HOH ('W', 3680, ' ')\n",
      "    HOH ('W', 3681, ' ')\n",
      "    HOH ('W', 3682, ' ')\n",
      "    HOH ('W', 3683, ' ')\n",
      "    HOH ('W', 3684, ' ')\n",
      "    HOH ('W', 3685, ' ')\n",
      "    HOH ('W', 3686, ' ')\n",
      "    HOH ('W', 3687, ' ')\n",
      "    HOH ('W', 3688, ' ')\n",
      "    HOH ('W', 3689, ' ')\n",
      "    HOH ('W', 3690, ' ')\n",
      "    HOH ('W', 3691, ' ')\n",
      "    HOH ('W', 3692, ' ')\n",
      "    HOH ('W', 3693, ' ')\n",
      "    HOH ('W', 3694, ' ')\n",
      "    HOH ('W', 3695, ' ')\n",
      "    HOH ('W', 3696, ' ')\n",
      "    HOH ('W', 3697, ' ')\n",
      "    HOH ('W', 3698, ' ')\n",
      "    HOH ('W', 3699, ' ')\n",
      "    HOH ('W', 3700, ' ')\n",
      "    HOH ('W', 3701, ' ')\n",
      "    HOH ('W', 3702, ' ')\n",
      "  Chain DI:\n",
      "    No hetero residues found\n",
      "  Chain DC:\n",
      "    HOH ('W', 301, ' ')\n",
      "    HOH ('W', 302, ' ')\n",
      "    HOH ('W', 303, ' ')\n",
      "    HOH ('W', 304, ' ')\n",
      "  Chain DD:\n",
      "    No hetero residues found\n",
      "  Chain DK:\n",
      "    No hetero residues found\n",
      "  Chain DP:\n",
      "    No hetero residues found\n",
      "  Chain DE:\n",
      "    HOH ('W', 301, ' ')\n",
      "    HOH ('W', 302, ' ')\n",
      "  Chain DY:\n",
      "    No hetero residues found\n",
      "  Chain D0:\n",
      "    No hetero residues found\n",
      "  Chain D4:\n",
      "    ZN ('H_ZN', 101, ' ')\n",
      "  Chain D1:\n",
      "    No hetero residues found\n",
      "  Chain D3:\n",
      "    No hetero residues found\n",
      "  Chain DV:\n",
      "    No hetero residues found\n",
      "  Chain D2:\n",
      "    No hetero residues found\n",
      "  Chain DL:\n",
      "    HOH ('W', 201, ' ')\n",
      "    HOH ('W', 202, ' ')\n",
      "    HOH ('W', 203, ' ')\n",
      "    HOH ('W', 204, ' ')\n",
      "  Chain DM:\n",
      "    No hetero residues found\n",
      "  Chain DX:\n",
      "    No hetero residues found\n",
      "  Chain DH:\n",
      "    No hetero residues found\n",
      "  Chain DJ:\n",
      "    No hetero residues found\n",
      "  Chain DN:\n",
      "    No hetero residues found\n",
      "  Chain DO:\n",
      "    No hetero residues found\n",
      "  Chain DQ:\n",
      "    No hetero residues found\n",
      "  Chain DS:\n",
      "    No hetero residues found\n",
      "  Chain DU:\n",
      "    No hetero residues found\n",
      "  Chain DF:\n",
      "    No hetero residues found\n",
      "  Chain DG:\n",
      "    No hetero residues found\n",
      "  Chain DR:\n",
      "    No hetero residues found\n",
      "  Chain DT:\n",
      "    No hetero residues found\n",
      "  Chain DZ:\n",
      "    No hetero residues found\n",
      "  Chain DW:\n",
      "    No hetero residues found\n"
     ]
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from Bio.PDB import MMCIFParser\n",
    "from IPython.display import display\n",
    "\n",
    "# Parse structure and identify all hetero atoms (potential ligands)\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Find all hetero atoms/ligands\n",
    "ligands = []\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        for residue in chain:\n",
    "            if residue.get_id()[0] != ' ':  # hetero atoms\n",
    "                if residue.get_id()[0] not in ['W', 'H_']:  # exclude water and hydrogens\n",
    "                    ligands.append((model.id, chain.id, residue.get_resname(), residue.get_id()))\n",
    "\n",
    "print(\"Found ligands/hetero atoms:\")\n",
    "for model_id, chain_id, resname, res_id in ligands:\n",
    "    print(f\"Model {model_id}, Chain {chain_id}: {resname} {res_id}\")\n",
    "\n",
    "\n",
    "\n",
    "# Print detailed residue information for debugging\n",
    "print(\"\\nDetailed residue information:\")\n",
    "for model in structure:\n",
    "    print(f\"Model {model.id}:\")\n",
    "    for chain in model:\n",
    "        print(f\"  Chain {chain.id}:\")\n",
    "        hetero_residues = [res for res in chain if res.get_id()[0] != ' ']\n",
    "        if hetero_residues:\n",
    "            for res in hetero_residues:\n",
    "                print(f\"    {res.get_resname()} {res.get_id()}\")\n",
    "        else:\n",
    "            print(\"    No hetero residues found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "720b1416-a9bf-4146-b70e-21921dceb483",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available chains: ['CL', 'CY', 'CL_2', 'AA', 'AC', 'AD', 'AE', 'AF', 'AG', 'AH', 'AI', 'AJ', 'AK', 'AL', 'AM', 'AN', 'AO', 'AP', 'AQ', 'AR', 'AS', 'AT', 'AB', 'AU', 'BA', 'BB', 'BI', 'BC', 'BD', 'BK', 'BP', 'BE', 'BY', 'B0', 'B4', 'B1', 'B3', 'BV', 'B2', 'BL', 'BM', 'BX', 'BH', 'BJ', 'BN', 'BO', 'BQ', 'BS', 'BU', 'BF', 'BG', 'BR', 'BT', 'BZ', 'BW', 'CA', 'CC', 'CD', 'CE', 'CF', 'CG', 'CH', 'CI', 'CJ', 'CK', 'CL_new', 'CM', 'CN', 'CO', 'CP', 'CQ', 'CR', 'CS', 'CT', 'CB', 'CU', 'DA', 'DB', 'DI', 'DC', 'DD', 'DK', 'DP', 'DE', 'DY', 'D0', 'D4', 'D1', 'D3', 'DV', 'D2', 'DL', 'DM', 'DX', 'DH', 'DJ', 'DN', 'DO', 'DQ', 'DS', 'DU', 'DF', 'DG', 'DR', 'DT', 'DZ', 'DW']\n",
      "Added PHE:582 in chain CY as green stick\n",
      "Added gentamicin representation with selection: LLL\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39fbe4c5f7054384aaa09c64a32da258",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Structure information:\n",
      "Model 0:\n",
      "  Chain CL: 125 protein residues, 0 hetero atoms\n",
      "  Chain CY: 667 protein residues, 3 hetero atoms\n",
      "  Chain CL_2: 123 protein residues, 2 hetero atoms\n",
      "  Chain AA: 1530 protein residues, 355 hetero atoms\n",
      "  Chain AC: 206 protein residues, 0 hetero atoms\n",
      "  Chain AD: 205 protein residues, 0 hetero atoms\n",
      "  Chain AE: 150 protein residues, 1 hetero atoms\n",
      "  Chain AF: 100 protein residues, 0 hetero atoms\n",
      "  Chain AG: 150 protein residues, 0 hetero atoms\n",
      "  Chain AH: 129 protein residues, 0 hetero atoms\n",
      "  Chain AI: 127 protein residues, 0 hetero atoms\n",
      "  Chain AJ: 98 protein residues, 0 hetero atoms\n",
      "  Chain AK: 117 protein residues, 1 hetero atoms\n",
      "  Chain AL: 123 protein residues, 2 hetero atoms\n",
      "  Chain AM: 114 protein residues, 0 hetero atoms\n",
      "  Chain AN: 96 protein residues, 2 hetero atoms\n",
      "  Chain AO: 88 protein residues, 0 hetero atoms\n",
      "  Chain AP: 82 protein residues, 0 hetero atoms\n",
      "  Chain AQ: 80 protein residues, 0 hetero atoms\n",
      "  Chain AR: 55 protein residues, 0 hetero atoms\n",
      "  Chain AS: 79 protein residues, 0 hetero atoms\n",
      "  Chain AT: 85 protein residues, 2 hetero atoms\n",
      "  Chain AB: 218 protein residues, 0 hetero atoms\n",
      "  Chain AU: 51 protein residues, 0 hetero atoms\n",
      "  Chain BA: 117 protein residues, 0 hetero atoms\n",
      "  Chain BB: 2841 protein residues, 603 hetero atoms\n",
      "  Chain BI: 141 protein residues, 0 hetero atoms\n",
      "  Chain BC: 271 protein residues, 7 hetero atoms\n",
      "  Chain BD: 209 protein residues, 0 hetero atoms\n",
      "  Chain BK: 121 protein residues, 0 hetero atoms\n",
      "  Chain BP: 114 protein residues, 0 hetero atoms\n",
      "  Chain BE: 201 protein residues, 3 hetero atoms\n",
      "  Chain BY: 58 protein residues, 0 hetero atoms\n",
      "  Chain B0: 56 protein residues, 0 hetero atoms\n",
      "  Chain B4: 38 protein residues, 1 hetero atoms\n",
      "  Chain B1: 50 protein residues, 0 hetero atoms\n",
      "  Chain B3: 64 protein residues, 0 hetero atoms\n",
      "  Chain BV: 94 protein residues, 0 hetero atoms\n",
      "  Chain B2: 46 protein residues, 1 hetero atoms\n",
      "  Chain BL: 143 protein residues, 3 hetero atoms\n",
      "  Chain BM: 136 protein residues, 0 hetero atoms\n",
      "  Chain BX: 63 protein residues, 0 hetero atoms\n",
      "  Chain BH: 149 protein residues, 1 hetero atoms\n",
      "  Chain BJ: 142 protein residues, 0 hetero atoms\n",
      "  Chain BN: 120 protein residues, 0 hetero atoms\n",
      "  Chain BO: 116 protein residues, 0 hetero atoms\n",
      "  Chain BQ: 117 protein residues, 0 hetero atoms\n",
      "  Chain BS: 110 protein residues, 0 hetero atoms\n",
      "  Chain BU: 102 protein residues, 0 hetero atoms\n",
      "  Chain BF: 178 protein residues, 0 hetero atoms\n",
      "  Chain BG: 176 protein residues, 0 hetero atoms\n",
      "  Chain BR: 103 protein residues, 0 hetero atoms\n",
      "  Chain BT: 93 protein residues, 0 hetero atoms\n",
      "  Chain BZ: 77 protein residues, 0 hetero atoms\n",
      "  Chain BW: 79 protein residues, 0 hetero atoms\n",
      "  Chain CA: 1530 protein residues, 361 hetero atoms\n",
      "  Chain CC: 206 protein residues, 0 hetero atoms\n",
      "  Chain CD: 205 protein residues, 0 hetero atoms\n",
      "  Chain CE: 150 protein residues, 2 hetero atoms\n",
      "  Chain CF: 100 protein residues, 0 hetero atoms\n",
      "  Chain CG: 152 protein residues, 0 hetero atoms\n",
      "  Chain CH: 129 protein residues, 0 hetero atoms\n",
      "  Chain CI: 127 protein residues, 0 hetero atoms\n",
      "  Chain CJ: 98 protein residues, 0 hetero atoms\n",
      "  Chain CK: 117 protein residues, 1 hetero atoms\n",
      "  Chain CL_new: 123 protein residues, 2 hetero atoms\n",
      "  Chain CM: 113 protein residues, 0 hetero atoms\n",
      "  Chain CN: 96 protein residues, 5 hetero atoms\n",
      "  Chain CO: 88 protein residues, 0 hetero atoms\n",
      "  Chain CP: 80 protein residues, 0 hetero atoms\n",
      "  Chain CQ: 81 protein residues, 0 hetero atoms\n",
      "  Chain CR: 55 protein residues, 0 hetero atoms\n",
      "  Chain CS: 80 protein residues, 0 hetero atoms\n",
      "  Chain CT: 85 protein residues, 2 hetero atoms\n",
      "  Chain CB: 218 protein residues, 0 hetero atoms\n",
      "  Chain CU: 51 protein residues, 0 hetero atoms\n",
      "  Chain DA: 117 protein residues, 0 hetero atoms\n",
      "  Chain DB: 2841 protein residues, 614 hetero atoms\n",
      "  Chain DI: 141 protein residues, 0 hetero atoms\n",
      "  Chain DC: 271 protein residues, 4 hetero atoms\n",
      "  Chain DD: 209 protein residues, 0 hetero atoms\n",
      "  Chain DK: 121 protein residues, 0 hetero atoms\n",
      "  Chain DP: 114 protein residues, 0 hetero atoms\n",
      "  Chain DE: 201 protein residues, 2 hetero atoms\n",
      "  Chain DY: 58 protein residues, 0 hetero atoms\n",
      "  Chain D0: 56 protein residues, 0 hetero atoms\n",
      "  Chain D4: 38 protein residues, 1 hetero atoms\n",
      "  Chain D1: 50 protein residues, 0 hetero atoms\n",
      "  Chain D3: 64 protein residues, 0 hetero atoms\n",
      "  Chain DV: 94 protein residues, 0 hetero atoms\n",
      "  Chain D2: 46 protein residues, 0 hetero atoms\n",
      "  Chain DL: 143 protein residues, 4 hetero atoms\n",
      "  Chain DM: 136 protein residues, 0 hetero atoms\n",
      "  Chain DX: 63 protein residues, 0 hetero atoms\n",
      "  Chain DH: 149 protein residues, 0 hetero atoms\n",
      "  Chain DJ: 142 protein residues, 0 hetero atoms\n",
      "  Chain DN: 120 protein residues, 0 hetero atoms\n",
      "  Chain DO: 116 protein residues, 0 hetero atoms\n",
      "  Chain DQ: 117 protein residues, 0 hetero atoms\n",
      "  Chain DS: 110 protein residues, 0 hetero atoms\n",
      "  Chain DU: 102 protein residues, 0 hetero atoms\n",
      "  Chain DF: 178 protein residues, 0 hetero atoms\n",
      "  Chain DG: 176 protein residues, 0 hetero atoms\n",
      "  Chain DR: 103 protein residues, 0 hetero atoms\n",
      "  Chain DT: 93 protein residues, 0 hetero atoms\n",
      "  Chain DZ: 77 protein residues, 0 hetero atoms\n",
      "  Chain DW: 79 protein residues, 0 hetero atoms\n"
     ]
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from Bio.PDB import MMCIFParser\n",
    "from IPython.display import display\n",
    "\n",
    "# Parse the CIF file to get chain information\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Get all chain IDs from both models\n",
    "all_chains = []\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        all_chains.append(chain.id)\n",
    "\n",
    "print(\"Available chains:\", all_chains)\n",
    "\n",
    "# Load the structure in NGLView\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Define colors for each chain (using NGLView compatible color names)\n",
    "chain_colors = {\n",
    "   'CY': 'pink', 'CL_new': 'green',\n",
    "    # Add more chains with different colors as needed\n",
    "    'AA': 'orange',\n",
    "    'DB': 'green',\n",
    "    'CA': 'purple',\n",
    "    'BB': 'cyan',\n",
    "    'BA': 'red',\n",
    "    'BB': 'brown',\n",
    "    'BL': 'gray',\n",
    "    'DB': 'magenta'\n",
    "}\n",
    "\n",
    "\n",
    "# Add cartoon representations for protein chains\n",
    "for chain_id in all_chains:\n",
    "    if chain_id in chain_colors:\n",
    "        color = chain_colors[chain_id]\n",
    "    else:\n",
    "        # Use a default color for unlisted chains\n",
    "        color = 'green'\n",
    "    \n",
    "    view.add_representation(\"cartoon\", \n",
    "                           selection=f\":{chain_id} and polymer\", \n",
    "                           color=color,\n",
    "                           opacity=0.8)\n",
    "# Highlight PHE:582 in chain CY as red stick\n",
    "selection_string = \"582:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "print(\"Added PHE:582 in chain CY as green stick\")\n",
    "\n",
    "# Highlight gentamicin (LLL) in red\n",
    "# Try different possible selections for the ligand\n",
    "gentamicin_selections = [\n",
    "    \"LLL\",  # residue name\n",
    "    \"[LLL]\",  # alternative syntax\n",
    "    \"hetero and not water\",  # all hetero atoms except water\n",
    "    \"ligand\"  # generic ligand selection\n",
    "]\n",
    "\n",
    "for selection in gentamicin_selections:\n",
    "    try:\n",
    "        view.add_representation(\"ball+stick\", \n",
    "                               selection=selection, \n",
    "                               color=\"red\",\n",
    "                               radius=0.5)\n",
    "        print(f\"Added gentamicin representation with selection: {selection}\")\n",
    "        break\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "#Optional: Add surface representation for better visualization\n",
    "#view.add_representation(\"surface\", \n",
    "#                       selection=\"protein\", \n",
    "#                        opacity=0.3,\n",
    "#                        color=\"white\")\n",
    "\n",
    "# Set background and camera\n",
    "view.background = \"white\"\n",
    "view.camera = \"perspective\"\n",
    "\n",
    "# Add labels (optional)\n",
    "view.add_label(text=\"Gentamicin\", selection=\"LLL\", color=\"black\", size=1)\n",
    "\n",
    "#RNA visualisation\n",
    "#view.add_representation(\"tube\", selection=\"nucleic\", color=\"orange\", opacity=0.8)\n",
    "\n",
    "\n",
    "display(view)\n",
    "\n",
    "# Print information about the structure\n",
    "print(\"\\nStructure information:\")\n",
    "for i, model in enumerate(structure):\n",
    "    print(f\"Model {i}:\")\n",
    "    for chain in model:\n",
    "        residue_count = len([res for res in chain if res.get_id()[0] == ' '])  # protein residues\n",
    "        hetero_count = len([res for res in chain if res.get_id()[0] != ' '])   # hetero atoms\n",
    "        print(f\"  Chain {chain.id}: {residue_count} protein residues, {hetero_count} hetero atoms\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac5aa45-5cae-4234-8d3c-94dd83bf14e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
