{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6868296f-0d69-4e6c-96b3-5be538ed259c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Structure parsing and manipulation\n",
    "from Bio.PDB import (\n",
    "    MMCIFParser, MMCIFIO, Select, PDBList, Superimposer,\n",
    "    Model, Structure, FastMMCIFParser, PPBuilder\n",
    ")\n",
    "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
    "\n",
    "# Sequence handling\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Visualization\n",
    "import matplotlib.pyplot as plt\n",
    "import nglview as nv\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dce47ec4-8d0c-439f-8fcd-c963232600d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4WPO\", \"4wpo.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49178a24-f554-4695-b9f1-53c8910d0ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "mmcif_dict = MMCIF2Dict(\"4wpo.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a296d2c7-7918-40f4-bf0a-47b8c605c85b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4WPO\", \"4wpo.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G chains\n",
    "class EF_G_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G chains (GB and ND)\n",
    "io.save(\"4WPO_EF-G.cif\", EF_G_ChainSelect([\"BZ\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28316515-3b6a-424c-8ea4-0e17bd639d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "widgets.IntSlider()\n",
    "view = nv.show_file(\"4WPO_EF-G.cif\")\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')  \n",
    "view.add_representation('stick', selection='540 and PRO', color='black')\n",
    "display(view)\n",
    "display(slider)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f014d79c-34fc-44cc-8889-4cc386a40dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the structure\n",
    "parser = MMCIFParser(QUIET=True)\n",
    "structure = parser.get_structure(\"4WPO\", \"4WPO_EF-G.cif\")\n",
    "\n",
    "ppb = PPBuilder()\n",
    "records = []\n",
    "\n",
    "# Extract sequences for each chain\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        seq = \"\"\n",
    "        for pp in ppb.build_peptides(chain):\n",
    "            seq += str(pp.get_sequence())\n",
    "\n",
    "        if seq:\n",
    "            record = SeqRecord(Seq(seq), id=f\"Chain_{chain.id}\", description=\"\")\n",
    "            records.append(record)\n",
    "\n",
    "# Print the sequences in FASTA format\n",
    "for record in records:\n",
    "    print(f\">{record.id}\\n{record.seq}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e556671-2d7f-4a00-b660-20ca4f298452",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4499fba2-6459-4592-8730-71b672514ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mmcif_dict = MMCIF2Dict(\"4v5f.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f70cdaac-7e08-45a0-8552-1a370f519247",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G chains\n",
    "class EF_G_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G chains \n",
    "io.save(\"4v5f_EF-G.cif\", EF_G_ChainSelect([\"CY\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "346ca0e9-3d06-4408-9956-f7184f180a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for finding proline number in 4v5f structure:\n",
    "# Parse the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure('4V5F', '4v5f_EF-G.cif')\n",
    "\n",
    "# Iterate over models, chains, and residues to list amino acids and their residue numbers\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        print(f\"Chain ID: {chain.id}\")\n",
    "        for residue in chain:\n",
    "            # residue.id is a tuple like (' ', resseq, insertion_code)\n",
    "            resseq = residue.id[1]  # residue sequence number\n",
    "            resname = residue.get_resname()  # 3-letter amino acid code\n",
    "            print(f\"Residue {resseq}: {resname}\")\n",
    "        print()  # Blank line between chains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45da778-b447-4047-985e-d1a43f0563c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = nv.show_file(\"4v5f_EF-G.cif\")\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='silver')\n",
    "selection_string = \"648:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string)\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "display(view)\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14cdd56a-1fac-4d7b-a560-362c7aa53b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', '4wpo_EF-G.cif')\n",
    "structure2 = parser.get_structure('s2', '4v5f_EF-G.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'BZ'\n",
    "chain2_id = 'CY'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_aligned.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79a17878-1b21-4043-99f0-a05375912dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = nv.show_file(\"combined_aligned.cif\")\n",
    "\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "\n",
    "view.clear_representations()\n",
    "\n",
    "# Cartoon for chain CY in silver\n",
    "view.add_representation('cartoon', selection=':CY', color='silver')\n",
    "\n",
    "# Cartoon for chain BZ in pink\n",
    "view.add_representation('cartoon', selection=':BZ', color='pink')\n",
    "\n",
    "# Highlight residue 648 in chain CY as red ball+stick + label\n",
    "selection_string = \"648:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n",
    "display(slider)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c669d3f6-cef0-49fc-94f0-7d41758e8e78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli36099ef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')\n",
    "selection_string = \"659:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375acbd3-ab8f-44da-b9d8-07d6d40f1dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli3609910xmicef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='siver')\n",
    "selection_string = \"659:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e88b87b0-6885-4d91-8efa-f87b567b297c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', 'e.coli36099ef-g.cif')\n",
    "structure2 = parser.get_structure('s2', 'e.coli3609910xmicef-g.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'A'\n",
    "chain2_id = 'A'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_alignedAF3model.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a9b1778-09d0-4927-8df8-d9a2c72b5b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = nv.show_file(\"combined_alignedAF3model.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Try different model selection syntaxes\n",
    "try:\n",
    "    view.add_representation('cartoon', selection='/0', color='pink')      # Model 0 with slash\n",
    "    view.add_representation('cartoon', selection='/1', color='silver')    # Model 1 with slash\n",
    "    print(\"Using model /0 and /1 - SUCCESS\")\n",
    "except Exception as e:\n",
    "    print(f\"Model selection failed: {e}\")\n",
    "    \n",
    "    # Fallback: try chain selection\n",
    "    try:\n",
    "        view.add_representation('cartoon', selection=':A', color='pink')\n",
    "        view.add_representation('cartoon', selection=':B', color='silver')\n",
    "        print(\"Using chain A and B - SUCCESS\")\n",
    "    except Exception as e2:\n",
    "        print(f\"Chain selection failed: {e2}\")\n",
    "        \n",
    "        # Final fallback: show everything with automatic coloring\n",
    "        view.add_representation('cartoon', color='chainid')  # Colors by chain automatically\n",
    "        print(\"Using automatic chain coloring - SUCCESS\")\n",
    "\n",
    "# Highlight residue\n",
    "selection_string = \"659:A\"\n",
    "try:\n",
    "    view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "    view.add_representation('label', \n",
    "                           selection=selection_string, \n",
    "                           labelType='format', \n",
    "                           labelFormat='%(resname)s%(resno)s', \n",
    "                           color='red', \n",
    "                           xOffset=1, \n",
    "                           fixedSize=True)\n",
    "except Exception as e:\n",
    "    print(f\"Residue highlighting failed: {e}\")\n",
    "\n",
    "# Force display\n",
    "display(view)\n",
    "\n",
    "# Slider\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c688434a-fd3d-4ebc-b2d9-e38e8112eba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"e.coli36099ef-g.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='pink')\n",
    "selection_string = \"593:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c169bd5d-58fb-481b-a1ac-08bed7897dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load CIF file from AlphaFold 3 server\n",
    "view = nv.show_file(\"ef-ge.colimg165510xmic.cif\")\n",
    "view.clear_representations()\n",
    "view.add_representation('cartoon', color='siver')\n",
    "selection_string = \"593:A.CA\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "display(view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93ea78a5-34c0-4f20-ba58-6df81ed5d69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse structures\n",
    "parser = MMCIFParser()\n",
    "structure1 = parser.get_structure('s1', 'e.coli36099ef-g.cif')\n",
    "structure2 = parser.get_structure('s2', 'ef-ge.colimg165510xmic.cif')\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = 'A'\n",
    "chain2_id = 'A'\n",
    "\n",
    "atoms1 = [res['CA'] for res in model1[chain1_id] if 'CA' in res]\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"combined_alignedAF3model2.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fafcc483-58f5-42c4-8113-6d583b691845",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = nv.show_file(\"combined_alignedAF3model2.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Try different model selection syntaxes\n",
    "try:\n",
    "    view.add_representation('cartoon', selection='/0', color='pink')      # Model 0 with slash\n",
    "    view.add_representation('cartoon', selection='/1', color='silver')    # Model 1 with slash\n",
    "    print(\"Using model /0 and /1 - SUCCESS\")\n",
    "except Exception as e:\n",
    "    print(f\"Model selection failed: {e}\")\n",
    "    \n",
    "    # Fallback: try chain selection\n",
    "    try:\n",
    "        view.add_representation('cartoon', selection=':A', color='pink')\n",
    "        view.add_representation('cartoon', selection=':B', color='silver')\n",
    "        print(\"Using chain A and B - SUCCESS\")\n",
    "    except Exception as e2:\n",
    "        print(f\"Chain selection failed: {e2}\")\n",
    "        \n",
    "        # Final fallback: show everything with automatic coloring\n",
    "        view.add_representation('cartoon', color='chainid')  # Colors by chain automatically\n",
    "        print(\"Using automatic chain coloring - SUCCESS\")\n",
    "\n",
    "# Highlight residue\n",
    "selection_string = \"593:A\"\n",
    "try:\n",
    "    view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "    view.add_representation('label', \n",
    "                           selection=selection_string, \n",
    "                           labelType='format', \n",
    "                           labelFormat='%(resname)s%(resno)s', \n",
    "                           color='red', \n",
    "                           xOffset=1, \n",
    "                           fixedSize=True)\n",
    "except Exception as e:\n",
    "    print(f\"Residue highlighting failed: {e}\")\n",
    "\n",
    "# Force display\n",
    "display(view)\n",
    "# Slider\n",
    "slider = widgets.IntSlider(min=0, max=100, step=1, value=50, description='My Slider')\n",
    "display(slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89c24ca-890d-4d50-a8e7-1392385a24e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c899aeb5-d934-40d6-8ae9-0970d1ec6c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v5f\", \"4v5f.cif\")\n",
    "\n",
    "# Custom Select class to keep only EF-G_S12 chains\n",
    "class EF_G_S12_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only EF-G_S12 chains \n",
    "io.save(\"4v5f_EF-G_S12.cif\", EF_G_S12_ChainSelect([\"CY\", \"CL\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f4ef6ff-a2c0-47bf-a130-5123255cbbaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for finding proline number in 4v5f structure:\n",
    "# Parse the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure('4V5F', '4v5f_EF-G_S12.cif')\n",
    "\n",
    "# Iterate over models, chains, and residues to list amino acids and their residue numbers\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        print(f\"Chain ID: {chain.id}\")\n",
    "        for residue in chain:\n",
    "            # residue.id is a tuple like (' ', resseq, insertion_code)\n",
    "            resseq = residue.id[1]  # residue sequence number\n",
    "            resname = residue.get_resname()  # 3-letter amino acid code\n",
    "            print(f\"Residue {resseq}: {resname}\")\n",
    "        print()  # Blank line between chains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83a092dd-f2d1-4bd4-997e-b7d5989cdbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v53\", \"4v53.cif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7357c2f-d395-47d5-ae9f-06a753422cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mmcif_dict = MMCIF2Dict(\"4v53.cif\")\n",
    "\n",
    "# This will list all chain IDs and associated entity IDs\n",
    "strand_ids = mmcif_dict.get(\"_entity_poly.pdbx_strand_id\", [])\n",
    "entity_ids = mmcif_dict.get(\"_entity_poly.entity_id\", [])\n",
    "polymer_names = mmcif_dict.get(\"_entity.pdbx_description\", [])\n",
    "\n",
    "for entity, strands, name in zip(entity_ids, strand_ids, polymer_names):\n",
    "    print(f\"Entity ID: {entity}, Chains: {strands}, Molecule: {name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "416f5d92-73e9-48f9-8ea1-4752b6d0717e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the structure\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"4v53\", \"4v53.cif\")\n",
    "\n",
    "# Custom Select class to keep only S12 chains\n",
    "class S12_ChainSelect(Select):\n",
    "    def __init__(self, chain_ids):\n",
    "        self.chain_ids = chain_ids\n",
    "\n",
    "    def accept_chain(self, chain):\n",
    "        return chain.id in self.chain_ids\n",
    "\n",
    "# Set up the writer\n",
    "io = MMCIFIO()\n",
    "io.set_structure(structure)\n",
    "\n",
    "# Save only S12 chains \n",
    "io.save(\"4v53_S12.cif\", S12_ChainSelect([\"CL\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13e7dec5-3744-43b4-a802-15ae6caac1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = FastMMCIFParser(QUIET=True)\n",
    "structure1 = parser.get_structure('s1', '4v5f_EF-G_S12.cif')\n",
    "structure2 = parser.get_structure('s2', '4v53_S12.cif')\n",
    "\n",
    "\n",
    "model1 = structure1[0]\n",
    "model2 = structure2[0]\n",
    "\n",
    "chain1_id = [\"CY\", \"CL\"]  \n",
    "chain2_id = \"CL\"\n",
    "\n",
    "atoms1 = [res[\"CA\"] \n",
    "          for cid in chain1_id \n",
    "          for res in model1[cid] \n",
    "          if \"CA\" in res]\n",
    "\n",
    "atoms2 = [res['CA'] for res in model2[chain2_id] if 'CA' in res]\n",
    "\n",
    "min_len = min(len(atoms1), len(atoms2))\n",
    "atoms1 = atoms1[:min_len]\n",
    "atoms2 = atoms2[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms1, atoms2)\n",
    "sup.apply(model2.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Create new structure with 2 models\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "\n",
    "# Add first model as model 0\n",
    "combined_structure.add(model1)\n",
    "\n",
    "# Rename second model to 1 and add (Bio.PDB requires unique model ids)\n",
    "model2.id = 1\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined multi-model CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"alignedS12+EF-G.cif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157da6d9-6d33-4108-a3d5-bcb9b8b89bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename overlapping chains in model2 to avoid coloring clashes\n",
    "existing_chains = {chain.id for chain in model1}\n",
    "for chain in model2:\n",
    "    if chain.id in existing_chains:\n",
    "        chain.id = chain.id + \"_2\"  # e.g., \"CL\" -> \"CL_2\"\n",
    "\n",
    "# Create combined structure\n",
    "combined_structure = Structure.Structure(\"combined\")\n",
    "combined_structure.add(model1)\n",
    "model2.id = 1  # second model\n",
    "combined_structure.add(model2)\n",
    "\n",
    "# Save combined CIF\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"alignedS12+EF-G_renamed.cif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d6f185c-4841-4257-b59c-700e7569d150",
   "metadata": {},
   "outputs": [],
   "source": [
    "view = nv.show_file(\"alignedS12+EF-G_renamed.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "#color by chain names\n",
    "view.add_representation(\"cartoon\", selection=\":CY\", color=\"pink\")\n",
    "view.add_representation(\"cartoon\", selection=\":CL\", color=\"blue\")\n",
    "view.add_representation(\"cartoon\", selection=\":CL_2\", color=\"silver\")\n",
    "\n",
    "from IPython.display import display\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874fe6c3-47ae-4e19-926c-ef7c6ddf498f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the previously superimposed structure\n",
    "parser = MMCIFParser(QUIET=True)\n",
    "aligned_structure = parser.get_structure('aligned', 'alignedS12+EF-G_renamed.cif')\n",
    "model2 = aligned_structure[0]  # first model (contains CL_2)\n",
    "\n",
    "# Load the new structure\n",
    "new_structure = parser.get_structure('4V53', '4V53.cif')\n",
    "model_new = new_structure[0]  # first model of 4V53\n",
    "\n",
    "# Select chains to use for superimposition\n",
    "chain_old = \"CL_2\"  # chain from previously superimposed structure\n",
    "chain_new = \"CL\"    # chain in 4V53\n",
    "\n",
    "# Extract CA atoms for alignment\n",
    "atoms_old = [res[\"CA\"] for res in model2[chain_old] if \"CA\" in res]\n",
    "atoms_new = [res[\"CA\"] for res in model_new[chain_new] if \"CA\" in res]\n",
    "\n",
    "# Match lengths\n",
    "min_len = min(len(atoms_old), len(atoms_new))\n",
    "atoms_old = atoms_old[:min_len]\n",
    "atoms_new = atoms_new[:min_len]\n",
    "\n",
    "# Superimpose\n",
    "sup = Superimposer()\n",
    "sup.set_atoms(atoms_old, atoms_new)\n",
    "sup.apply(model_new.get_atoms())\n",
    "print(\"RMSD after superimposition:\", sup.rms)\n",
    "\n",
    "# Rename chain in new model if needed to avoid duplicates\n",
    "if chain_new in [c.id for c in model2]:\n",
    "    for chain in model_new:\n",
    "        if chain.id == chain_new:\n",
    "            chain.id = chain_new + \"_new\"\n",
    "\n",
    "# Assign unique model IDs before combining\n",
    "model2.id = 0\n",
    "model_new.id = 1\n",
    "\n",
    "# Combine structures into one\n",
    "combined_structure = Structure.Structure(\"super_aligned_2\")\n",
    "combined_structure.add(model2)\n",
    "combined_structure.add(model_new)\n",
    "\n",
    "# Save combined structure\n",
    "io = MMCIFIO()\n",
    "io.set_structure(combined_structure)\n",
    "io.save(\"superimposed_combined.cif\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "076eb53f-7f5b-4faa-a48a-5aeabfc396bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the CIF file directly\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "\n",
    "# Clear default representations if you want custom colors\n",
    "view.clear_representations()\n",
    "\n",
    "# Add cartoon representations for chains\n",
    "view.add_representation(\"cartoon\", selection=\":CY\", color=\"pink\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Parse the CIF file\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Extract chain IDs for model 1\n",
    "model_index = 0  # Python index\n",
    "chains_model1 = [chain.id for chain in structure[model_index]]\n",
    "\n",
    "# Assign colors\n",
    "colors = [plt.cm.tab20(i % 20) for i in range(len(chains_model1))]\n",
    "\n",
    "# Load into NGLView directly\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Add cartoon representation for each chain with colors\n",
    "for chain_id, color in zip(chains_model1, colors):\n",
    "    view.add_representation(\"cartoon\", selection=f\":{chain_id}\", color=color)\n",
    "\n",
    "from IPython.display import display\n",
    "display(view)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb9350e9-bbc7-4bd0-8c33-5113a94593f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse structure and identify all hetero atoms (potential ligands)\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Find all hetero atoms/ligands\n",
    "ligands = []\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        for residue in chain:\n",
    "            if residue.get_id()[0] != ' ':  # hetero atoms\n",
    "                if residue.get_id()[0] not in ['W', 'H_']:  # exclude water and hydrogens\n",
    "                    ligands.append((model.id, chain.id, residue.get_resname(), residue.get_id()))\n",
    "\n",
    "print(\"Found ligands/hetero atoms:\")\n",
    "for model_id, chain_id, resname, res_id in ligands:\n",
    "    print(f\"Model {model_id}, Chain {chain_id}: {resname} {res_id}\")\n",
    "\n",
    "\n",
    "\n",
    "# Print detailed residue information for debugging\n",
    "print(\"\\nDetailed residue information:\")\n",
    "for model in structure:\n",
    "    print(f\"Model {model.id}:\")\n",
    "    for chain in model:\n",
    "        print(f\"  Chain {chain.id}:\")\n",
    "        hetero_residues = [res for res in chain if res.get_id()[0] != ' ']\n",
    "        if hetero_residues:\n",
    "            for res in hetero_residues:\n",
    "                print(f\"    {res.get_resname()} {res.get_id()}\")\n",
    "        else:\n",
    "            print(\"    No hetero residues found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "720b1416-a9bf-4146-b70e-21921dceb483",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the CIF file to get chain information\n",
    "parser = MMCIFParser()\n",
    "structure = parser.get_structure(\"super\", \"superimposed_combined.cif\")\n",
    "\n",
    "# Get all chain IDs from both models\n",
    "all_chains = []\n",
    "for model in structure:\n",
    "    for chain in model:\n",
    "        all_chains.append(chain.id)\n",
    "\n",
    "print(\"Available chains:\", all_chains)\n",
    "\n",
    "# Load the structure in NGLView\n",
    "view = nv.show_file(\"superimposed_combined.cif\")\n",
    "view.clear_representations()\n",
    "\n",
    "# Define colors for each chain (using NGLView compatible color names)\n",
    "chain_colors = {\n",
    "   'CY': 'pink', 'CL_new': 'green',\n",
    "    # Add more chains with different colors as needed\n",
    "    'AA': 'orange',\n",
    "    'DB': 'green',\n",
    "    'CA': 'purple',\n",
    "    'BB': 'cyan',\n",
    "    'BA': 'red',\n",
    "    'BB': 'brown',\n",
    "    'BL': 'gray',\n",
    "    'DB': 'magenta'\n",
    "}\n",
    "\n",
    "\n",
    "# Add cartoon representations for protein chains\n",
    "for chain_id in all_chains:\n",
    "    if chain_id in chain_colors:\n",
    "        color = chain_colors[chain_id]\n",
    "    else:\n",
    "        # Use a default color for unlisted chains\n",
    "        color = 'green'\n",
    "    \n",
    "    view.add_representation(\"cartoon\", \n",
    "                           selection=f\":{chain_id} and polymer\", \n",
    "                           color=color,\n",
    "                           opacity=0.8)\n",
    "# Highlight PHE:582 in chain CY as red stick\n",
    "selection_string = \"582:CY\"\n",
    "view.add_representation('ball+stick', selection=selection_string, color='red')\n",
    "view.add_representation('label', sele=selection_string, labelType='format', labelFormat='%(resname)s%(resno)s', color='red', xOffset=1, fixedSize=True)\n",
    "\n",
    "print(\"Added PHE:582 in chain CY as green stick\")\n",
    "\n",
    "# Highlight gentamicin (LLL) in red\n",
    "# Try different possible selections for the ligand\n",
    "gentamicin_selections = [\n",
    "    \"LLL\",  # residue name\n",
    "    \"[LLL]\",  # alternative syntax\n",
    "    \"hetero and not water\",  # all hetero atoms except water\n",
    "    \"ligand\"  # generic ligand selection\n",
    "]\n",
    "\n",
    "for selection in gentamicin_selections:\n",
    "    try:\n",
    "        view.add_representation(\"ball+stick\", \n",
    "                               selection=selection, \n",
    "                               color=\"red\",\n",
    "                               radius=0.5)\n",
    "        print(f\"Added gentamicin representation with selection: {selection}\")\n",
    "        break\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "#Optional: Add surface representation for better visualization\n",
    "#view.add_representation(\"surface\", \n",
    "#                       selection=\"protein\", \n",
    "#                        opacity=0.3,\n",
    "#                        color=\"white\")\n",
    "\n",
    "# Set background and camera\n",
    "view.background = \"white\"\n",
    "view.camera = \"perspective\"\n",
    "\n",
    "# Add labels (optional)\n",
    "view.add_label(text=\"Gentamicin\", selection=\"LLL\", color=\"black\", size=1)\n",
    "\n",
    "#RNA visualisation\n",
    "#view.add_representation(\"tube\", selection=\"nucleic\", color=\"orange\", opacity=0.8)\n",
    "\n",
    "\n",
    "display(view)\n",
    "\n",
    "# Print information about the structure\n",
    "print(\"\\nStructure information:\")\n",
    "for i, model in enumerate(structure):\n",
    "    print(f\"Model {i}:\")\n",
    "    for chain in model:\n",
    "        residue_count = len([res for res in chain if res.get_id()[0] == ' '])  # protein residues\n",
    "        hetero_count = len([res for res in chain if res.get_id()[0] != ' '])   # hetero atoms\n",
    "        print(f\"  Chain {chain.id}: {residue_count} protein residues, {hetero_count} hetero atoms\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac5aa45-5cae-4234-8d3c-94dd83bf14e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
